<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.10.0" />
<title>MILWRM API documentation</title>
<meta name="description" content="Multiplex Image Labeling With Regional Morphology â€¦" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Package <code>MILWRM</code></h1>
</header>
<section id="section-intro">
<p>Multiplex Image Labeling With Regional Morphology</p>
<p>Authors: <a href="https://github.com/KSimi7">Harsimran Kaur</a> &amp; <a href="https://github.com/codyheiser">Cody N. Heiser</a></p>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python"># -*- coding: utf-8 -*-
&#34;&#34;&#34;
Multiplex Image Labeling With Regional Morphology

Authors: [Harsimran Kaur](https://github.com/KSimi7) &amp; [Cody N. Heiser](https://github.com/codyheiser)
&#34;&#34;&#34;
from .MILWRM import (
    mxif_labeler,
    st_labeler,
)
from .MxIF import img
from .ST import (
    blur_features_st,
    map_pixels,
    trim_image,
    assemble_pita,
    show_pita,
)

__all__ = [
    &#34;img&#34;,
    &#34;blur_features_st&#34;,
    &#34;map_pixels&#34;,
    &#34;trim_image&#34;,
    &#34;assemble_pita&#34;,
    &#34;show_pita&#34;,
    &#34;mxif_labeler&#34;,
    &#34;st_labeler&#34;,
]

from ._version import get_versions

__version__ = get_versions()[&#34;version&#34;]
del get_versions</code></pre>
</details>
</section>
<section>
<h2 class="section-title" id="header-submodules">Sub-modules</h2>
<dl>
<dt><code class="name"><a title="MILWRM.MILWRM" href="MILWRM.html">MILWRM.MILWRM</a></code></dt>
<dd>
<div class="desc"><p>Classes for assigning tissue domain IDs to multiplex immunofluorescence (MxIF) or 10X
Visium spatial transcriptomic (ST) and histological imaging data</p></div>
</dd>
<dt><code class="name"><a title="MILWRM.MxIF" href="MxIF.html">MILWRM.MxIF</a></code></dt>
<dd>
<div class="desc"><p>Functions and classes for analyzing multiplex imaging data</p></div>
</dd>
<dt><code class="name"><a title="MILWRM.ST" href="ST.html">MILWRM.ST</a></code></dt>
<dd>
<div class="desc"><p>Functions and classes for manipulating 10X Visium spatial transcriptomic (ST) and
histological imaging data</p></div>
</dd>
</dl>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="MILWRM.assemble_pita"><code class="name flex">
<span>def <span class="ident">assemble_pita</span></span>(<span>adata, features=None, use_rep=None, layer=None, plot_out=True, histo=None, verbose=True, **kwargs)</span>
</code></dt>
<dd>
<div class="desc"><p>Cast feature into pixel space to construct gene expression image ("pita")</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>adata</code></strong> :&ensp;<code>AnnData.anndata</code></dt>
<dd>the data</dd>
<dt><strong><code>features</code></strong> :&ensp;<code>list</code> of <code>int</code> or <code>str</code></dt>
<dd>Names or indices of features to cast onto spot image. If <code>None</code>, cast all
features. If <code>plot_out</code>, first feature in list will be plotted. If not
specified and <code>plot_out</code>, first feature (index 0) will be plotted.</dd>
<dt><strong><code>use_rep</code></strong> :&ensp;<code>str</code></dt>
<dd>Key from <code>adata.obsm</code> to use for plotting. If <code>None</code>, use <code>adata.X</code>.</dd>
<dt>layer :str</dt>
<dt>Key from <code>adata.layers</code> to use for plotting. Ignored if <code>use_rep</code> is not <code>None</code></dt>
<dt><strong><code>plot_out</code></strong> :&ensp;<code>bool</code></dt>
<dd>Show resulting image?</dd>
<dt><strong><code>histo</code></strong> :&ensp;<code>str</code> or <code>None</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>Histology image to show along with pita in gridspec (i.e. "hires",
"hires_trim", "lowres"). If <code>None</code> or if <code>plot_out</code>==<code>False</code>, ignore.</dd>
<dt><strong><code>verbose</code></strong> :&ensp;<code>bool</code>, optional <code>(default=</code>True<code>)</code></dt>
<dd>Print updates to console</dd>
<dt><strong><code>**kwargs</code></strong></dt>
<dd>Arguments to pass to <code><a title="MILWRM.show_pita" href="#MILWRM.show_pita">show_pita()</a></code> function</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>assembled</code></strong> :&ensp;<code>np.array</code></dt>
<dd>Image of desired expression in pixel space</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def assemble_pita(
    adata,
    features=None,
    use_rep=None,
    layer=None,
    plot_out=True,
    histo=None,
    verbose=True,
    **kwargs,
):
    &#34;&#34;&#34;
    Cast feature into pixel space to construct gene expression image (&#34;pita&#34;)

    Parameters
    ----------
    adata : AnnData.anndata
        the data
    features : list of int or str
        Names or indices of features to cast onto spot image. If `None`, cast all
        features. If `plot_out`, first feature in list will be plotted. If not
        specified and `plot_out`, first feature (index 0) will be plotted.
    use_rep : str
        Key from `adata.obsm` to use for plotting. If `None`, use `adata.X`.
    layer :str
        Key from `adata.layers` to use for plotting. Ignored if `use_rep` is not `None`
    plot_out : bool
        Show resulting image?
    histo : str or `None`, optional (default=`None`)
        Histology image to show along with pita in gridspec (i.e. &#34;hires&#34;,
        &#34;hires_trim&#34;, &#34;lowres&#34;). If `None` or if `plot_out`==`False`, ignore.
    verbose : bool, optional (default=`True`)
        Print updates to console
    **kwargs
        Arguments to pass to `show_pita()` function

    Returns
    -------
    assembled : np.array
        Image of desired expression in pixel space
    &#34;&#34;&#34;
    assert (
        adata.uns[&#34;pixel_map_params&#34;] is not None
    ), &#34;Pixel map not yet created. Run map_pixels() first.&#34;

    # coerce features to list if only single string
    if features and not isinstance(features, list):
        features = [features]

    if use_rep is None:
        # use all genes if no gene features specified
        if not features:
            features = adata.var_names  # [adata.var.highly_variable == 1].tolist()
        if layer is None:
            if verbose:
                print(
                    &#34;Assembling pita with {} features from adata.X&#34;.format(
                        len(features)
                    )
                )
            mapper = pd.DataFrame(
                adata.X[:, [adata.var_names.get_loc(x) for x in features]],
                index=adata.obs_names,
            )
        else:
            if verbose:
                print(
                    &#34;Assembling pita with {} features from adata.layers[&#39;{}&#39;]&#34;.format(
                        len(features), layer
                    )
                )
            mapper = pd.DataFrame(
                adata.layers[layer][:, [adata.var_names.get_loc(x) for x in features]],
                index=adata.obs_names,
            )
    elif use_rep in [&#34;.obs&#34;, &#34;obs&#34;]:
        assert features is not None, &#34;Must provide feature(s) from adata.obs&#34;
        if verbose:
            print(
                &#34;Assembling pita with {} features from adata.obs&#34;.format(len(features))
            )
        if all(isinstance(x, int) for x in features):
            mapper = adata.obs.iloc[:, features].copy()
        else:
            mapper = adata.obs[features].copy()
    else:
        if not features:
            if verbose:
                print(
                    &#34;Assembling pita with {} features from adata.obsm[&#39;{}&#39;]&#34;.format(
                        adata.obsm[use_rep].shape[1], use_rep
                    )
                )
            mapper = pd.DataFrame(adata.obsm[use_rep], index=adata.obs_names)
        else:
            assert all(
                isinstance(x, int) for x in features
            ), &#34;Features must be integer indices if using rep from adata.obsm&#34;
            if verbose:
                print(
                    &#34;Assembling pita with {} features from adata.obsm[&#39;{}&#39;]&#34;.format(
                        len(features), use_rep
                    )
                )
            mapper = pd.DataFrame(
                adata.obsm[use_rep][:, features], index=adata.obs_names
            )

    # check for categorical columns to force into discrete plots
    discrete_cols = {}
    for col in mapper.columns:
        if pd.api.types.is_categorical_dtype(mapper[col]):
            cat_max = len(mapper[col].cat.categories)
            categories = mapper[col].cat.categories  # save original categories
            mapper[col] = mapper[col].replace(
                {v: k for k, v in dict(enumerate(mapper[col].cat.categories)).items()}
            )
            discrete_cols[mapper.columns.get_loc(col)] = (cat_max, categories)
    # if no categorical columns, pass None to discrete_cols
    if bool(discrete_cols) is False:
        discrete_cols = None

    # cast barcodes into pixel dimensions for reindexing
    if verbose:
        print(
            &#34;Casting barcodes to pixel dimensions and saving to adata.uns[&#39;pixel_map&#39;]&#34;
        )
    pixel_map = (
        adata.uns[&#34;pixel_map_df&#34;].pivot(index=&#34;y&#34;, columns=&#34;x&#34;, values=&#34;barcode&#34;).values
    )

    assembled = np.array(
        [mapper.reindex(index=pixel_map[x], copy=True) for x in range(len(pixel_map))]
    ).squeeze()

    if plot_out:
        # determine where the histo image is in anndata
        if histo is not None:
            assert (
                histo
                in adata.uns[&#34;spatial&#34;][list(adata.uns[&#34;spatial&#34;].keys())[0]][
                    &#34;images&#34;
                ].keys()
            ), &#34;Must provide one of {} for histo&#34;.format(
                adata.uns[&#34;spatial&#34;][list(adata.uns[&#34;spatial&#34;].keys())[0]][
                    &#34;images&#34;
                ].keys()
            )
            histo = adata.uns[&#34;spatial&#34;][list(adata.uns[&#34;spatial&#34;].keys())[0]][
                &#34;images&#34;
            ][histo]
        show_pita(
            pita=assembled,
            features=None,
            discrete_features=discrete_cols,
            histo=histo,
            **kwargs,
        )
    if verbose:
        print(&#34;Done!&#34;)
    return assembled, discrete_cols</code></pre>
</details>
</dd>
<dt id="MILWRM.blur_features_st"><code class="name flex">
<span>def <span class="ident">blur_features_st</span></span>(<span>adata, tmp, spatial_graph_key=None, n_rings=1)</span>
</code></dt>
<dd>
<div class="desc"><p>Blur values in an <code>AnnData</code> object using spatial nearest neighbors</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>adata</code></strong> :&ensp;<code>anndata.AnnData</code></dt>
<dd>AnnData object containing Visium data</dd>
<dt><strong><code>tmp</code></strong> :&ensp;<code>pd.DataFrame</code></dt>
<dd>containing feature columns from adata.obs that will be blurred</dd>
<dt><strong><code>spatial_graph_key</code></strong> :&ensp;<code>str</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>Key in <code>adata.obsp</code> containing spatial graph connectivities (i.e.
<code>"spatial_connectivities"</code>). If <code>None</code>, compute new spatial graph using
<code>n_rings</code> in <code>squidpy</code>.</dd>
<dt><strong><code>n_rings</code></strong> :&ensp;<code>int</code>, optional <code>(default=1)</code></dt>
<dd>Number of hexagonal rings around each spatial transcriptomics spot to blur
features by for capturing regional information. Assumes 10X Genomics Visium
platform.</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>adata.obs is edited in place with new blurred columns</code></dt>
<dd>&nbsp;</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def blur_features_st(adata, tmp, spatial_graph_key=None, n_rings=1):
    &#34;&#34;&#34;
    Blur values in an `AnnData` object using spatial nearest neighbors

    Parameters
    ----------
    adata : anndata.AnnData
        AnnData object containing Visium data
    tmp : pd.DataFrame
        containing feature columns from adata.obs that will be blurred
    spatial_graph_key : str, optional (default=`None`)
        Key in `adata.obsp` containing spatial graph connectivities (i.e.
        `&#34;spatial_connectivities&#34;`). If `None`, compute new spatial graph using
        `n_rings` in `squidpy`.
    n_rings : int, optional (default=1)
        Number of hexagonal rings around each spatial transcriptomics spot to blur
        features by for capturing regional information. Assumes 10X Genomics Visium
        platform.

    Returns
    -------
    adata.obs is edited in place with new blurred columns
    &#34;&#34;&#34;
    if spatial_graph_key is not None:
        # use existing spatial graph
        assert (
            spatial_graph_key in adata.obsp.keys()
        ), &#34;Spatial connectivities key &#39;{}&#39; not found.&#34;.format(spatial_graph_key)
    else:
        # create spatial graph
        print(&#34;Computing spatial graph with {} hexagonal rings&#34;.format(n_rings))
        sq.gr.spatial_neighbors(adata, coord_type=&#34;grid&#34;, n_rings=n_rings)
        spatial_graph_key = &#34;spatial_connectivities&#34;  # set key to expected output
    tmp2 = tmp.copy()  # copy of temporary dataframe for dropping blurred features into
    cols = tmp.columns  # get column names
    # perform blurring by nearest spot neighbors
    for x in range(len(tmp)):
        vals = tmp.iloc[
            list(
                np.argwhere(
                    adata.obsp[spatial_graph_key][
                        x,
                    ]
                )[:, 1]
            )
            + [x],
            :,
        ].mean()
        tmp2.iloc[x, :] = vals.values
    # add blurred features to anndata object
    adata.obs[[x for x in cols]] = tmp.loc[:, cols].values
    adata.obs[[&#34;blur_&#34; + x for x in cols]] = tmp2.loc[:, cols].values
    return tmp2.loc[:, cols]</code></pre>
</details>
</dd>
<dt id="MILWRM.map_pixels"><code class="name flex">
<span>def <span class="ident">map_pixels</span></span>(<span>adata, filter_label='in_tissue', img_key='hires', library_id=None)</span>
</code></dt>
<dd>
<div class="desc"><p>Map spot IDs to 'pixel space' by assigning spot ID values to evenly spaced grid</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>adata</code></strong> :&ensp;<code>AnnData.anndata</code></dt>
<dd>The data</dd>
<dt><strong><code>filter_label</code></strong> :&ensp;<code>str</code> or <code>None</code></dt>
<dd>adata.obs column key that contains binary labels for filtering barcodes. If
None, do not filter.</dd>
<dt><strong><code>img_key</code></strong> :&ensp;<code>str</code></dt>
<dd>adata.uns key containing the image to use for mapping</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>adata</code></strong> :&ensp;<code>AnnData.anndata</code></dt>
<dd>with the following attributes:
adata.uns["pixel_map_df"] : pd.DataFrame
Long-form dataframe of Visium spot barcode IDs, pixel coordinates, and
.obs metadata
adata.uns["pixel_map"] : np.array
Pixel space array of Visium spot barcode IDs</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def map_pixels(adata, filter_label=&#34;in_tissue&#34;, img_key=&#34;hires&#34;, library_id=None):
    &#34;&#34;&#34;
    Map spot IDs to &#39;pixel space&#39; by assigning spot ID values to evenly spaced grid

    Parameters
    ----------
    adata : AnnData.anndata
        The data
    filter_label : str or None
        adata.obs column key that contains binary labels for filtering barcodes. If
        None, do not filter.
    img_key : str
        adata.uns key containing the image to use for mapping

    Returns
    -------
    adata : AnnData.anndata
        with the following attributes:
        adata.uns[&#34;pixel_map_df&#34;] : pd.DataFrame
            Long-form dataframe of Visium spot barcode IDs, pixel coordinates, and
            .obs metadata
        adata.uns[&#34;pixel_map&#34;] : np.array
            Pixel space array of Visium spot barcode IDs
    &#34;&#34;&#34;
    adata.uns[&#34;pixel_map_params&#34;] = {
        &#34;img_key&#34;: img_key
    }  # create params dict for future use
    # add library_id key to params
    if library_id is None:
        library_id = adata.uns[&#34;pixel_map_params&#34;][&#34;library_id&#34;] = list(
            adata.uns[&#34;spatial&#34;].keys()
        )[0]
    else:
        adata.uns[&#34;pixel_map_params&#34;][&#34;library_id&#34;] = library_id
    # first get center-to-face pixel distance of hexagonal Visium spots
    dist = euclidean_distances(adata.obsm[&#34;spatial&#34;])
    adata.uns[&#34;pixel_map_params&#34;][&#34;ctr_to_face&#34;] = (
        np.unique(dist)[np.unique(dist) != 0].min() / 2
    )
    # also save center-to-vertex pixel distance as vadata attribute
    adata.uns[&#34;pixel_map_params&#34;][&#34;ctr_to_vert&#34;] = adata.uns[&#34;pixel_map_params&#34;][
        &#34;ctr_to_face&#34;
    ] / np.cos(30 * (np.pi / 180))
    # get the spot radius from adata.uns[&#34;spatial&#34;] as well
    adata.uns[&#34;pixel_map_params&#34;][&#34;radius&#34;] = (
        adata.uns[&#34;spatial&#34;][library_id][&#34;scalefactors&#34;][&#34;spot_diameter_fullres&#34;] / 2
    )
    # get scale factor from adata.uns[&#34;spatial&#34;]
    adata.uns[&#34;pixel_map_params&#34;][&#34;scalef&#34;] = adata.uns[&#34;spatial&#34;][library_id][
        &#34;scalefactors&#34;
    ][f&#34;tissue_{img_key}_scalef&#34;]

    if filter_label is not None:
        # create frame of mock pixels to make edges look better
        # x and y deltas for moving rows and columns into a blank frame
        delta_x = (
            adata[adata.obs.array_col == 0, :].obsm[&#34;spatial&#34;]
            - adata[adata.obs.array_col == 1, :].obsm[&#34;spatial&#34;]
        )
        delta_x = np.mean(list(delta_x[:, 1])) * 2
        delta_y = (
            adata[adata.obs.array_row == 0, :].obsm[&#34;spatial&#34;]
            - adata[adata.obs.array_row == 1, :].obsm[&#34;spatial&#34;]
        )
        delta_y = np.mean(list(delta_y[:, 1])) * 2
        # left part of frame, translated
        left = adata[
            adata.obs.array_col.isin(
                [adata.obs.array_col.max() - 2, adata.obs.array_col.max() - 3]
            ),
            :,
        ].copy()
        left.obsm[&#34;spatial&#34;][..., 0] -= delta_x.astype(int)
        del left.var
        del left.uns
        left.obs[filter_label] = 0
        left.obs_names = [&#34;left&#34; + str(x) for x in range(left.n_obs)]
        # right part of frame, translated
        right = adata[adata.obs.array_col.isin([2, 3]), :].copy()
        right.obsm[&#34;spatial&#34;][..., 0] += delta_x.astype(int)
        del right.var
        del right.uns
        right.obs[filter_label] = 0
        right.obs_names = [&#34;right&#34; + str(x) for x in range(right.n_obs)]
        # add sides to orig
        a_sides = adata.concatenate(
            [left, right],
            index_unique=None,
        )
        a_sides.obs.drop(columns=&#34;batch&#34;, inplace=True)
        # bottom part of frame, translated
        bottom = a_sides[a_sides.obs.array_row == 1, :].copy()
        bottom.obsm[&#34;spatial&#34;][..., 1] += delta_y.astype(int)
        bottom.obs_names = [&#34;bottom&#34; + str(x) for x in range(bottom.n_obs)]
        del bottom.var
        del bottom.uns
        bottom.obs[filter_label] = 0
        # top part of frame, translated
        top = a_sides[
            a_sides.obs.array_row == a_sides.obs.array_row.max() - 1, :
        ].copy()
        top.obsm[&#34;spatial&#34;][..., 1] -= delta_y.astype(int)
        del top.var
        del top.uns
        top.obs[filter_label] = 0
        top.obs_names = [&#34;top&#34; + str(x) for x in range(top.n_obs)]
        # complete frame
        a_frame = a_sides.concatenate(
            [top, bottom],
            index_unique=None,
        )
        a_frame.uns = adata.uns
        a_frame.var = adata.var
        a_frame.obs.drop(columns=&#34;batch&#34;, inplace=True)
    else:
        a_frame = adata.copy()

    # determine pixel bounds from spot coords, adding center-to-face distance
    a_frame.uns[&#34;pixel_map_params&#34;][&#34;xmin_px&#34;] = int(
        np.floor(
            a_frame.uns[&#34;pixel_map_params&#34;][&#34;scalef&#34;]
            * (
                a_frame.obsm[&#34;spatial&#34;][:, 0].min()
                - a_frame.uns[&#34;pixel_map_params&#34;][&#34;radius&#34;]
            )
        )
    )
    a_frame.uns[&#34;pixel_map_params&#34;][&#34;xmax_px&#34;] = int(
        np.ceil(
            a_frame.uns[&#34;pixel_map_params&#34;][&#34;scalef&#34;]
            * (
                a_frame.obsm[&#34;spatial&#34;][:, 0].max()
                + a_frame.uns[&#34;pixel_map_params&#34;][&#34;radius&#34;]
            )
        )
    )
    a_frame.uns[&#34;pixel_map_params&#34;][&#34;ymin_px&#34;] = int(
        np.floor(
            a_frame.uns[&#34;pixel_map_params&#34;][&#34;scalef&#34;]
            * (
                a_frame.obsm[&#34;spatial&#34;][:, 1].min()
                - a_frame.uns[&#34;pixel_map_params&#34;][&#34;radius&#34;]
            )
        )
    )
    a_frame.uns[&#34;pixel_map_params&#34;][&#34;ymax_px&#34;] = int(
        np.ceil(
            a_frame.uns[&#34;pixel_map_params&#34;][&#34;scalef&#34;]
            * (
                a_frame.obsm[&#34;spatial&#34;][:, 1].max()
                + a_frame.uns[&#34;pixel_map_params&#34;][&#34;radius&#34;]
            )
        )
    )

    print(&#34;Creating pixel grid and mapping to nearest barcode coordinates&#34;)
    # define grid for pixel space
    grid_y, grid_x = np.mgrid[
        a_frame.uns[&#34;pixel_map_params&#34;][&#34;ymin_px&#34;] : a_frame.uns[&#34;pixel_map_params&#34;][
            &#34;ymax_px&#34;
        ],
        a_frame.uns[&#34;pixel_map_params&#34;][&#34;xmin_px&#34;] : a_frame.uns[&#34;pixel_map_params&#34;][
            &#34;xmax_px&#34;
        ],
    ]
    # map barcodes to pixel coordinates
    pixel_coords = np.column_stack((grid_x.ravel(order=&#34;C&#34;), grid_y.ravel(order=&#34;C&#34;)))
    barcode_list = griddata(
        np.multiply(a_frame.obsm[&#34;spatial&#34;], a_frame.uns[&#34;pixel_map_params&#34;][&#34;scalef&#34;]),
        a_frame.obs_names,
        (pixel_coords[:, 0], pixel_coords[:, 1]),
        method=&#34;nearest&#34;,
    )
    # save grid_x and grid_y to adata.uns
    a_frame.uns[&#34;grid_x&#34;], a_frame.uns[&#34;grid_y&#34;] = grid_x, grid_y

    # put results into DataFrame for filtering and reindexing
    print(&#34;Saving barcode mapping to adata.uns[&#39;pixel_map_df&#39;] and adding metadata&#34;)
    a_frame.uns[&#34;pixel_map_df&#34;] = pd.DataFrame(pixel_coords, columns=[&#34;x&#34;, &#34;y&#34;])
    # add barcodes to long-form dataframe
    a_frame.uns[&#34;pixel_map_df&#34;][&#34;barcode&#34;] = barcode_list
    # merge master df with self.adata.obs for metadata
    a_frame.uns[&#34;pixel_map_df&#34;] = a_frame.uns[&#34;pixel_map_df&#34;].merge(
        a_frame.obs, how=&#34;outer&#34;, left_on=&#34;barcode&#34;, right_index=True
    )
    # filter using label from adata.obs if desired (i.e. &#34;in_tissue&#34;)
    if filter_label is not None:
        print(
            &#34;Filtering barcodes using labels in self.adata.obs[&#39;{}&#39;]&#34;.format(
                filter_label
            )
        )
        # set empty pixels (no Visium spot) to &#34;none&#34;
        a_frame.uns[&#34;pixel_map_df&#34;].loc[
            a_frame.uns[&#34;pixel_map_df&#34;][filter_label] == 0,
            &#34;barcode&#34;,
        ] = &#34;none&#34;
        # subset the entire anndata object using filter_label
        a_frame = a_frame[a_frame.obs[filter_label] == 1, :].copy()
        print(&#34;New size: {} spots x {} genes&#34;.format(a_frame.n_obs, a_frame.n_vars))

    print(&#34;Done!&#34;)
    return a_frame</code></pre>
</details>
</dd>
<dt id="MILWRM.show_pita"><code class="name flex">
<span>def <span class="ident">show_pita</span></span>(<span>pita, features=None, discrete_features=None, RGB=False, histo=None, label='feature', ncols=4, figsize=(7, 7), cmap='plasma', save_to=None, **kwargs)</span>
</code></dt>
<dd>
<div class="desc"><p>Plot assembled pita using <code>plt.imshow()</code></p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>pita</code></strong> :&ensp;<code>np.array</code></dt>
<dd>Image of desired expression in pixel space from <code>.<a title="MILWRM.assemble_pita" href="#MILWRM.assemble_pita">assemble_pita()</a></code></dd>
<dt><strong><code>features</code></strong> :&ensp;<code>list</code> of <code>int</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>List of features by index to show in plot. If <code>None</code>, use all features.</dd>
<dt><strong><code>discrete_features</code></strong> :&ensp;<code>dict</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>Dictionary of feature indices (keys) containing discrete (categorical) values
(i.e. MILWRM domain). Values are tuple of <code>max_value</code> to pass to
<code>plot_single_image_discrete</code> for each discrete feature, and the ordered list
of categories for legend plotting. If <code>None</code>, treat all features as continuous.</dd>
<dt><strong><code>RGB</code></strong> :&ensp;<code>bool</code>, optional <code>(default=</code>False<code>)</code></dt>
<dd>Treat 3-dimensional array as RGB image</dd>
<dt><strong><code>histo</code></strong> :&ensp;<code>np.array</code> or <code>None</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>Histology image to show along with pita in gridspec. If <code>None</code>, ignore.</dd>
<dt><strong><code>label</code></strong> :&ensp;<code>str</code>, optional <code>(default="feature")</code></dt>
<dd>What to title each panel of the gridspec (i.e. "PC" or "usage") or each
channel in RGB image. Can also pass list of names e.g. ["NeuN","GFAP",
"DAPI"] corresponding to channels.</dd>
<dt><strong><code>ncols</code></strong> :&ensp;<code>int</code>, optional <code>(default=4)</code></dt>
<dd>Number of columns for gridspec</dd>
<dt><strong><code>figsize</code></strong> :&ensp;<code>tuple</code> of <code>float</code>, optional <code>(default=(7, 7))</code></dt>
<dd>Size in inches of output figure</dd>
<dt><strong><code>cmap</code></strong> :&ensp;<code>str</code>, optional <code>(default="plasma")</code></dt>
<dd>Matplotlib colormap to use</dd>
<dt><strong><code>save_to</code></strong> :&ensp;<code>str</code> or <code>None</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>Path to image file to save results. if <code>None</code>, show figure.</dd>
<dt><strong><code>**kwargs</code></strong></dt>
<dd>Arguments to pass to <code>plt.imshow()</code> function</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>Matplotlib object (if plotting one feature</code> or <code>RGB)</code> or <code>gridspec object (for</code></dt>
<dd>&nbsp;</dd>
</dl>
<p>multiple features). Saves plot to file if <code>save_to</code> is not <code>None</code>.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def show_pita(
    pita,
    features=None,
    discrete_features=None,
    RGB=False,
    histo=None,
    label=&#34;feature&#34;,
    ncols=4,
    figsize=(7, 7),
    cmap=&#34;plasma&#34;,
    save_to=None,
    **kwargs,
):
    &#34;&#34;&#34;
    Plot assembled pita using `plt.imshow()`

    Parameters
    ----------
    pita : np.array
        Image of desired expression in pixel space from `.assemble_pita()`
    features : list of int, optional (default=`None`)
        List of features by index to show in plot. If `None`, use all features.
    discrete_features : dict, optional (default=`None`)
        Dictionary of feature indices (keys) containing discrete (categorical) values
        (i.e. MILWRM domain). Values are tuple of `max_value` to pass to
        `plot_single_image_discrete` for each discrete feature, and the ordered list
        of categories for legend plotting. If `None`, treat all features as continuous.
    RGB : bool, optional (default=`False`)
        Treat 3-dimensional array as RGB image
    histo : np.array or `None`, optional (default=`None`)
        Histology image to show along with pita in gridspec. If `None`, ignore.
    label : str, optional (default=&#34;feature&#34;)
        What to title each panel of the gridspec (i.e. &#34;PC&#34; or &#34;usage&#34;) or each
        channel in RGB image. Can also pass list of names e.g. [&#34;NeuN&#34;,&#34;GFAP&#34;,
        &#34;DAPI&#34;] corresponding to channels.
    ncols : int, optional (default=4)
        Number of columns for gridspec
    figsize : tuple of float, optional (default=(7, 7))
        Size in inches of output figure
    cmap : str, optional (default=&#34;plasma&#34;)
        Matplotlib colormap to use
    save_to : str or None, optional (default=`None`)
        Path to image file to save results. if `None`, show figure.
    **kwargs
        Arguments to pass to `plt.imshow()` function

    Returns
    -------
    Matplotlib object (if plotting one feature or RGB) or gridspec object (for
    multiple features). Saves plot to file if `save_to` is not `None`.
    &#34;&#34;&#34;
    assert pita.ndim &gt; 1, &#34;Pita does not have enough dimensions: {} given&#34;.format(
        pita.ndim
    )
    assert pita.ndim &lt; 4, &#34;Pita has too many dimensions: {} given&#34;.format(pita.ndim)
    # if only one feature (2D), plot it quickly
    if (pita.ndim == 2) and histo is None:
        fig, ax = plt.subplots(1, 1, figsize=figsize)
        if discrete_features is not None:
            plot_single_image_discrete(
                image=pita,
                ax=ax,
                # use first value in dict as max
                max_val=list(discrete_features.values())[0][0],
                ticklabels=list(discrete_features.values())[0][1],
                label=label[0] if isinstance(label, list) else label,
                cmap=cmap,
                **kwargs,
            )
        else:
            plot_single_image(
                image=pita,
                ax=ax,
                label=label[0] if isinstance(label, list) else label,
                cmap=cmap,
                **kwargs,
            )
        plt.tight_layout()
        if save_to:
            plt.savefig(fname=save_to, transparent=True, bbox_inches=&#34;tight&#34;, dpi=300)
        return fig
    if (pita.ndim == 2) and histo is not None:
        n_rows, n_cols = 1, 2  # two images here, histo and RGB
        fig = plt.figure(figsize=(ncols * n_cols, ncols * n_rows))
        # arrange axes as subplots
        gs = gridspec.GridSpec(n_rows, n_cols, figure=fig)
        # add plots to axes
        ax = plt.subplot(gs[0])
        plot_single_image_rgb(
            image=histo,
            ax=ax,
            channels=None,
            label=&#34;Histology&#34;,
            **kwargs,
        )
        ax = plt.subplot(gs[1])
        if discrete_features is not None:
            plot_single_image_discrete(
                image=pita,
                ax=ax,
                # use first value in dict as max
                max_val=list(discrete_features.values())[0][0],
                ticklabels=list(discrete_features.values())[0][1],
                label=label[0] if isinstance(label, list) else label,
                cmap=cmap,
                **kwargs,
            )
        else:
            plot_single_image(
                image=pita,
                ax=ax,
                label=label[0] if isinstance(label, list) else label,
                cmap=cmap,
                **kwargs,
            )
        fig.tight_layout()
        if save_to:
            plt.savefig(fname=save_to, transparent=True, bbox_inches=&#34;tight&#34;, dpi=300)
        return fig
    if RGB:
        # if third dim has 3 features, treat as RGB and plot it quickly
        assert (pita.ndim == 3) &amp; (
            pita.shape[2] == 3
        ), &#34;Need 3 dimensions and 3 given features for an RGB image; shape = {}; features given = {}&#34;.format(
            pita.shape, len(features)
        )
        print(&#34;Plotting pita as RGB image&#34;)
        if isinstance(label, str):
            # if label is single string, name channels numerically
            channels = [&#34;{}_{}&#34;.format(label, x) for x in range(pita.shape[2])]
        else:
            assert (
                len(label) == 3
            ), &#34;Please pass 3 channel names for RGB plot; {} labels given: {}&#34;.format(
                len(label), label
            )
            channels = label
        if histo is not None:
            n_rows, n_cols = 1, 2  # two images here, histo and RGB
            fig = plt.figure(figsize=(ncols * n_cols, ncols * n_rows))
            # arrange axes as subplots
            gs = gridspec.GridSpec(n_rows, n_cols, figure=fig)
            # add plots to axes
            ax = plt.subplot(gs[0])
            plot_single_image_rgb(
                image=histo,
                ax=ax,
                channels=None,
                label=&#34;Histology&#34;,
                **kwargs,
            )
            ax = plt.subplot(gs[1])
            plot_single_image_rgb(
                image=pita,
                ax=ax,
                channels=channels,
                label=&#34;&#34;,
                **kwargs,
            )
            fig.tight_layout()
            if save_to:
                plt.savefig(
                    fname=save_to, transparent=True, bbox_inches=&#34;tight&#34;, dpi=800
                )
            return fig
        else:
            fig, ax = plt.subplots(1, 1, figsize=figsize)
            plot_single_image_rgb(
                image=pita,
                ax=ax,
                channels=channels,
                label=&#34;&#34;,
                **kwargs,
            )
            if save_to:
                plt.savefig(
                    fname=save_to, transparent=True, bbox_inches=&#34;tight&#34;, dpi=300
                )
            return fig
    # if pita has multiple features, plot them in gridspec
    if isinstance(features, int):  # force features into list if single integer
        features = [features]
    # if no features are given, use all of them
    if features is None:
        features = [x for x in range(pita.shape[2])]
    else:
        assert (
            pita.ndim &gt; 2
        ), &#34;Not enough features in pita: shape {}, expecting 3rd dim with length {}&#34;.format(
            pita.shape, len(features)
        )
        assert (
            len(features) &lt;= pita.shape[2]
        ), &#34;Too many features given: pita has {}, expected {}&#34;.format(
            pita.shape[2], len(features)
        )
    if isinstance(label, str):
        # if label is single string, name channels numerically
        labels = [&#34;{}_{}&#34;.format(label, x) for x in features]
    else:
        assert len(label) == len(
            features
        ), &#34;Please provide the same number of labels as features; {} labels given, {} features given.&#34;.format(
            len(label), len(features)
        )
        labels = label
    # calculate gridspec dimensions
    if histo is not None:
        labels = [&#34;Histology&#34;] + labels  # append histo to front of labels
        if len(features) + 1 &lt;= ncols:
            n_rows, n_cols = 1, len(features) + 1
        else:
            n_rows, n_cols = ceil((len(features) + 1) / ncols), ncols
    else:
        if len(features) &lt;= ncols:
            n_rows, n_cols = 1, len(features)
        else:
            n_rows, n_cols = ceil(len(features) / ncols), ncols
    fig = plt.figure(figsize=(ncols * n_cols, ncols * n_rows))
    # arrange axes as subplots
    gs = gridspec.GridSpec(n_rows, n_cols, figure=fig)
    # add plots to axes
    i = 0
    if histo is not None:
        # add histology plot to first axes
        ax = plt.subplot(gs[i])
        plot_single_image_rgb(
            image=histo,
            ax=ax,
            channels=None,
            label=labels[i],
            **kwargs,
        )
        i = i + 1
    for feature in features:
        ax = plt.subplot(gs[i])
        if discrete_features is not None:
            if feature in discrete_features:
                plot_single_image_discrete(
                    image=pita[:, :, feature],
                    ax=ax,
                    # use corresponding value in dict as max
                    max_val=discrete_features[feature][0],
                    ticklabels=discrete_features[feature][1],
                    label=labels[i],
                    cmap=cmap,
                    **kwargs,
                )
            else:
                plot_single_image(
                    image=pita[:, :, feature],
                    ax=ax,
                    label=labels[i],
                    cmap=cmap,
                    **kwargs,
                )
        else:
            plot_single_image(
                image=pita[:, :, feature],
                ax=ax,
                label=labels[i],
                cmap=cmap,
                **kwargs,
            )
        i = i + 1
    fig.tight_layout()
    if save_to:
        plt.savefig(fname=save_to, transparent=True, bbox_inches=&#34;tight&#34;, dpi=300)
    return fig</code></pre>
</details>
</dd>
<dt id="MILWRM.trim_image"><code class="name flex">
<span>def <span class="ident">trim_image</span></span>(<span>adata, distance_trim=False, threshold=None, channels=None, plot_out=True, **kwargs)</span>
</code></dt>
<dd>
<div class="desc"><p>Trim pixels in image using pixel map output from Visium barcodes</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>adata</code></strong> :&ensp;<code>AnnData.anndata</code></dt>
<dd>The data</dd>
<dt><strong><code>distance_trim</code></strong> :&ensp;<code>bool</code></dt>
<dd>Manually trim pixels by distance to nearest Visium spot center</dd>
<dt><strong><code>threshold</code></strong> :&ensp;<code>int</code> or <code>None</code></dt>
<dd>Number of pixels from nearest Visium spot center to call barcode ID. Ignored
if <code>distance_trim==False</code>.</dd>
<dt><strong><code>channels</code></strong> :&ensp;<code>list</code> of <code>str</code> or <code>None</code></dt>
<dd>Names of image channels in axis order. If None, channels are named "ch_0",
"ch_1", etc.</dd>
<dt><strong><code>plot_out</code></strong> :&ensp;<code>bool</code></dt>
<dd>Plot final trimmed image</dd>
<dt><strong><code>**kwargs</code></strong></dt>
<dd>Arguments to pass to <code><a title="MILWRM.show_pita" href="#MILWRM.show_pita">show_pita()</a></code> function if <code>plot_out==True</code></dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>adata.uns["pixel_map_trim"] : np.array</code></dt>
<dd>Contains image with unused pixels set to <code>np.nan</code></dd>
<dt><code>adata.obsm["spatial_trim"] : np.array</code></dt>
<dd>Contains spatial coords with adjusted pixel values after image cropping</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def trim_image(
    adata, distance_trim=False, threshold=None, channels=None, plot_out=True, **kwargs
):
    &#34;&#34;&#34;
    Trim pixels in image using pixel map output from Visium barcodes

    Parameters
    ----------
    adata : AnnData.anndata
        The data
    distance_trim : bool
        Manually trim pixels by distance to nearest Visium spot center
    threshold : int or None
        Number of pixels from nearest Visium spot center to call barcode ID. Ignored
        if `distance_trim==False`.
    channels : list of str or None
        Names of image channels in axis order. If None, channels are named &#34;ch_0&#34;,
        &#34;ch_1&#34;, etc.
    plot_out : bool
        Plot final trimmed image
    **kwargs
        Arguments to pass to `show_pita()` function if `plot_out==True`

    Returns
    -------
    adata.uns[&#34;pixel_map_trim&#34;] : np.array
        Contains image with unused pixels set to `np.nan`
    adata.obsm[&#34;spatial_trim&#34;] : np.array
        Contains spatial coords with adjusted pixel values after image cropping
    &#34;&#34;&#34;
    assert (
        adata.uns[&#34;pixel_map_params&#34;] is not None
    ), &#34;Pixel map not yet created. Run map_pixels() first.&#34;

    print(
        &#34;Cropping image to pixel dimensions and adding values to adata.uns[&#39;pixel_map_df&#39;]&#34;
    )
    cropped = adata.uns[&#34;spatial&#34;][adata.uns[&#34;pixel_map_params&#34;][&#34;library_id&#34;]][
        &#34;images&#34;
    ][adata.uns[&#34;pixel_map_params&#34;][&#34;img_key&#34;]].transpose(1, 0, 2)[
        int(adata.uns[&#34;pixel_map_params&#34;][&#34;xmin_px&#34;]) : int(
            (adata.uns[&#34;pixel_map_params&#34;][&#34;xmax_px&#34;])
        ),
        int(adata.uns[&#34;pixel_map_params&#34;][&#34;ymin_px&#34;]) : int(
            (adata.uns[&#34;pixel_map_params&#34;][&#34;ymax_px&#34;])
        ),
    ]
    # crop x,y coords and save to .obsm as well
    print(&#34;Cropping Visium spot coordinates and saving to adata.obsm[&#39;spatial_trim&#39;]&#34;)
    adata.obsm[&#34;spatial_trim&#34;] = adata.obsm[&#34;spatial&#34;] - np.repeat(
        [
            [
                adata.uns[&#34;pixel_map_params&#34;][&#34;xmin_px&#34;],
                adata.uns[&#34;pixel_map_params&#34;][&#34;ymin_px&#34;],
            ]
        ],
        adata.obsm[&#34;spatial&#34;].shape[0],
        axis=0,
    )

    # manual trimming of pixels by distance if desired
    if distance_trim:
        print(&#34;Calculating pixel distances from spot centers for thresholding&#34;)
        tree = cKDTree(adata.obsm[&#34;spatial&#34;])
        xi = interpnd._ndim_coords_from_arrays(
            (adata.uns[&#34;grid_x&#34;], adata.uns[&#34;grid_y&#34;]),
            ndim=adata.obsm[&#34;spatial&#34;].shape[1],
        )
        dists, _ = tree.query(xi)

        # determine distance threshold
        if threshold is None:
            threshold = int(adata.uns[&#34;pixel_map_params&#34;][&#34;ctr_to_vert&#34;] + 1)
            print(
                &#34;Using distance threshold of {} pixels from adata.uns[&#39;pixel_map_params&#39;][&#39;ctr_to_vert&#39;]&#34;.format(
                    threshold
                )
            )

        dist_mask = bin_threshold(dists, threshmax=threshold)
        if plot_out:
            # plot pixel distances from spot centers on image
            show_pita(pita=dists, figsize=(4, 4))
            # plot binary thresholded image
            show_pita(pita=dist_mask, figsize=(4, 4))

        print(
            &#34;Trimming pixels by spot distance and adjusting labels in adata.uns[&#39;pixel_map_df&#39;]&#34;
        )
        mask_df = pd.DataFrame(dist_mask.T.ravel(order=&#34;F&#34;), columns=[&#34;manual_trim&#34;])
        adata.uns[&#34;pixel_map_df&#34;] = adata.uns[&#34;pixel_map_df&#34;].merge(
            mask_df, left_index=True, right_index=True
        )
        adata.uns[&#34;pixel_map_df&#34;].loc[
            adata.uns[&#34;pixel_map_df&#34;][&#34;manual_trim&#34;] == 1, [&#34;barcode&#34;]
        ] = &#34;none&#34;  # set empty pixels to empty barcode
        adata.uns[&#34;pixel_map_df&#34;].drop(
            columns=&#34;manual_trim&#34;, inplace=True
        )  # remove unneeded label

    if channels is None:
        # if channel names not specified, name them numerically
        channels = [&#34;ch_{}&#34;.format(x) for x in range(cropped.shape[2])]
    # cast image intensity values to long-form and add to adata.uns[&#34;pixel_map_df&#34;]
    rgb = pd.DataFrame(
        np.column_stack(
            [cropped[:, :, x].ravel(order=&#34;F&#34;) for x in range(cropped.shape[2])]
        ),
        columns=channels,
    )
    adata.uns[&#34;pixel_map_df&#34;] = adata.uns[&#34;pixel_map_df&#34;].merge(
        rgb, left_index=True, right_index=True
    )
    adata.uns[&#34;pixel_map_df&#34;].loc[
        adata.uns[&#34;pixel_map_df&#34;][&#34;barcode&#34;] == &#34;none&#34;, channels
    ] = np.nan  # set empty pixels to invalid image intensity value

    # calculate mean image values for each channel and create .obsm key
    adata.obsm[&#34;image_means&#34;] = (
        adata.uns[&#34;pixel_map_df&#34;]
        .loc[adata.uns[&#34;pixel_map_df&#34;][&#34;barcode&#34;] != &#34;none&#34;, [&#34;barcode&#34;] + channels]
        .groupby(&#34;barcode&#34;)
        .mean()
        .values
    )

    print(
        &#34;Saving cropped and trimmed image to adata.uns[&#39;spatial&#39;][&#39;{}&#39;][&#39;images&#39;][&#39;{}_trim&#39;]&#34;.format(
            adata.uns[&#34;pixel_map_params&#34;][&#34;library_id&#34;],
            adata.uns[&#34;pixel_map_params&#34;][&#34;img_key&#34;],
        )
    )
    adata.uns[&#34;spatial&#34;][adata.uns[&#34;pixel_map_params&#34;][&#34;library_id&#34;]][&#34;images&#34;][
        &#34;{}_trim&#34;.format(adata.uns[&#34;pixel_map_params&#34;][&#34;img_key&#34;])
    ] = np.dstack(
        [
            adata.uns[&#34;pixel_map_df&#34;]
            .pivot(index=&#34;y&#34;, columns=&#34;x&#34;, values=[channels[x]])
            .values
            for x in range(len(channels))
        ]
    )
    # save scale factor as well
    adata.uns[&#34;spatial&#34;][adata.uns[&#34;pixel_map_params&#34;][&#34;library_id&#34;]][&#34;scalefactors&#34;][
        &#34;tissue_{}_trim_scalef&#34;.format(adata.uns[&#34;pixel_map_params&#34;][&#34;img_key&#34;])
    ] = adata.uns[&#34;spatial&#34;][adata.uns[&#34;pixel_map_params&#34;][&#34;library_id&#34;]][
        &#34;scalefactors&#34;
    ][
        &#34;tissue_{}_scalef&#34;.format(adata.uns[&#34;pixel_map_params&#34;][&#34;img_key&#34;])
    ]
    # plot results if desired
    if plot_out:
        if len(channels) == 3:
            show_pita(
                pita=adata.uns[&#34;spatial&#34;][adata.uns[&#34;pixel_map_params&#34;][&#34;library_id&#34;]][
                    &#34;images&#34;
                ][&#34;{}_trim&#34;.format(adata.uns[&#34;pixel_map_params&#34;][&#34;img_key&#34;])],
                RGB=True,
                label=channels,
                **kwargs,
            )
        else:
            show_pita(
                pita=adata.uns[&#34;spatial&#34;][adata.uns[&#34;pixel_map_params&#34;][&#34;library_id&#34;]][
                    &#34;images&#34;
                ][&#34;{}_trim&#34;.format(adata.uns[&#34;pixel_map_params&#34;][&#34;img_key&#34;])],
                RGB=False,
                label=channels,
                **kwargs,
            )
    print(&#34;Done!&#34;)</code></pre>
</details>
</dd>
</dl>
</section>
<section>
<h2 class="section-title" id="header-classes">Classes</h2>
<dl>
<dt id="MILWRM.img"><code class="flex name class">
<span>class <span class="ident">img</span></span>
<span>(</span><span>img_arr, channels=None, mask=None)</span>
</code></dt>
<dd>
<div class="desc"><p>Initialize img class</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>img_arr</code></strong> :&ensp;<code>np.ndarray</code></dt>
<dd>The image as a numpy array</dd>
<dt><strong><code>channels</code></strong> :&ensp;<code>list</code> of <code>str</code> or <code>None</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>List of channel names corresponding to img.shape[2]. i.e. <code>("DAPI","GFAP",
"NeuH")&lt;code&gt;. If &lt;/code&gt;None</code>, channels are named "ch_0", "ch_1", etc.</dd>
<dt><strong><code>mask</code></strong> :&ensp;<code>np.ndarray</code></dt>
<dd>Mask defining pixels containing tissue in the image</dd>
</dl>
<h2 id="returns">Returns</h2>
<p><code><a title="MILWRM.img" href="#MILWRM.img">img</a></code> object</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class img:
    def __init__(self, img_arr, channels=None, mask=None):
        &#34;&#34;&#34;
        Initialize img class

        Parameters
        ----------
        img_arr : np.ndarray
            The image as a numpy array
        channels : list of str or None, optional (default=`None`)
            List of channel names corresponding to img.shape[2]. i.e. `(&#34;DAPI&#34;,&#34;GFAP&#34;,
            &#34;NeuH&#34;)`. If `None`, channels are named &#34;ch_0&#34;, &#34;ch_1&#34;, etc.
        mask : np.ndarray
            Mask defining pixels containing tissue in the image

        Returns
        -------
        `img` object
        &#34;&#34;&#34;
        assert (
            img_arr.ndim &gt; 1
        ), &#34;Image does not have enough dimensions: {} given&#34;.format(img_arr.ndim)
        self.img = img_arr.astype(&#34;float64&#34;)  # save image array to .img attribute
        if img_arr.ndim &gt; 2:
            self.n_ch = img_arr.shape[2]  # save number of channels to attribute
        else:
            self.n_ch = 1
        if channels is None:
            # if channel names not specified, name them numerically
            self.ch = [&#34;ch_{}&#34;.format(x) for x in range(self.n_ch)]
        else:
            if not isinstance(channels, list):
                raise Exception(&#34;Channels must be given in a list&#34;)
            assert (
                len(channels) == self.n_ch
            ), &#34;Number of channels must match img_arr.shape[2]&#34;
            self.ch = channels
        if mask is not None:
            # validate that mask matches img_arr
            assert (
                mask.shape == img_arr.shape[:2]
            ), &#34;Shape of mask must match the first two dimensions of img_arr&#34;
        self.mask = mask  # set mask attribute, regardless of value given

    def __repr__(self) -&gt; str:
        &#34;&#34;&#34;Representation of contents of img object&#34;&#34;&#34;
        descr = &#34;img object with {} of {} and shape {}px x {}px\n&#34;.format(
            type(self.img),
            self.img.dtype,
            self.img.shape[0],
            self.img.shape[1],
        ) + &#34;{} image channels:\n\t{}&#34;.format(self.n_ch, self.ch)
        if self.mask is not None:
            descr += &#34;\n\ntissue mask {} of {} and shape {}px x {}px&#34;.format(
                type(self.mask),
                self.mask.dtype,
                self.mask.shape[0],
                self.mask.shape[1],
            )
        return descr

    def copy(self) -&gt; &#34;img&#34;:
        &#34;&#34;&#34;Full copy of img object&#34;&#34;&#34;
        new = {}
        for key in [&#34;img&#34;, &#34;ch&#34;, &#34;mask&#34;]:
            attr = self.__getattribute__(key)
            if attr is not None:
                new[key] = attr.copy()
            else:
                new[key] = None
        return img(img_arr=new[&#34;img&#34;], channels=new[&#34;ch&#34;], mask=new[&#34;mask&#34;])

    def __getitem__(self, channels):
        &#34;&#34;&#34;Slice img object with channel name(s)&#34;&#34;&#34;
        if isinstance(channels, int):  # force channels into list if single integer
            channels = [channels]
        if isinstance(channels, str):  # force channels into int if single string
            channels = [self.ch.index(channels)]
        if checktype(channels):  # force channels into list of int if list of strings
            channels = [self.ch.index(x) for x in channels]

        if channels is None:  # if no channels are given, use all of them
            channels = [x for x in range(self.n_ch)]

        return self.img[:, :, channels]

    @classmethod
    def from_tiffs(cls, tiffdir, channels, common_strings=None, mask=None):
        &#34;&#34;&#34;
        Initialize img class from `.tif` files

        Parameters
        ----------
        tiffdir : str
            Path to directory containing `.tif` files for a multiplexed image
        channels : list of str
            List of channels present in `.tif` file names (case-sensitive)
            corresponding to img.shape[2] e.g. `(&#34;ACTG1&#34;,&#34;BCATENIN&#34;,&#34;DAPI&#34;,...)`
        common_strings : str, list of str, or `None`, optional (default=None)
            Strings to look for in all `.tif` files in `tiffdir` corresponding to
            `channels` e.g. `(&#34;WD86055_&#34;, &#34;_region_001.tif&#34;)` for files named
            &#34;WD86055_[MARKERNAME]_region_001.tif&#34;. If `None`, assume that only 1 image
            for each marker in `channels` is present in `tiffdir`.
        mask : str, optional (default=None)
            Name of mask defining pixels containing tissue in the image, present in
            `.tif` file names (case-sensitive) e.g. &#34;_01_TISSUE_MASK.tif&#34;

        Returns
        -------
        `img` object
        &#34;&#34;&#34;
        if common_strings is not None:
            # coerce single string to list
            if isinstance(common_strings, str):
                common_strings = [common_strings]
        A = []  # list for dumping numpy arrays
        for channel in channels:
            if common_strings is None:
                # find file matching all common_strings and channel name
                f = [f for f in os.listdir(tiffdir) if channel in f]
            else:
                # find file matching all common_strings and channel name
                f = [
                    f
                    for f in os.listdir(tiffdir)
                    if all(x in f for x in common_strings + [channel])
                ]
            # assertions so we only get one file per channel
            assert len(f) != 0, &#34;No file found with channel {}&#34;.format(channel)
            assert (
                len(f) == 1
            ), &#34;More than one match found for file with channel {}&#34;.format(channel)
            f = os.path.join(tiffdir, f[0])  # get full path to file for reading
            print(&#34;Reading marker {} from {}&#34;.format(channel, f))
            tmp = imread(f)  # read in .tif file
            A.append(tmp)  # append numpy array to list
        A_arr = np.dstack(
            A
        )  # stack numpy arrays in new dimension (third dim is channel)
        print(&#34;Final image array of shape: {}&#34;.format(A_arr.shape))
        # read in tissue mask if available
        if mask is not None:
            f = [f for f in os.listdir(tiffdir) if mask in f]
            # assertions so we only get one mask file
            assert len(f) != 0, &#34;No tissue mask file found&#34;
            assert len(f) == 1, &#34;More than one match found for tissue mask file&#34;
            f = os.path.join(tiffdir, f[0])  # get full path to file for reading
            print(&#34;Reading tissue mask from {}&#34;.format(f))
            A_mask = imread(f)  # read in .tif file
            assert (
                A_mask.shape == A_arr.shape[:2]
            ), &#34;Mask (shape: {}) is not the same shape as marker images (shape: {})&#34;.format(
                A_mask.shape, A_arr.shape[:2]
            )
            print(&#34;Final mask array of shape: {}&#34;.format(A_mask.shape))
        else:
            A_mask = None
        # generate img object
        return cls(img_arr=A_arr, channels=channels, mask=A_mask)

    @classmethod
    def from_npz(cls, file):
        &#34;&#34;&#34;
        Initialize img class from `.npz` file

        Parameters
        ----------
        file : str
            Path to `.npz` file containing saved img object and metadata

        Returns
        -------
        `img` object
        &#34;&#34;&#34;
        print(&#34;Loading img object from {}...&#34;.format(file))
        tmp = np.load(file)  # load from .npz compressed file
        assert (
            &#34;img&#34; in tmp.files
        ), &#34;Unexpected files in .npz: {}, expected [&#39;img&#39;,&#39;mask&#39;,&#39;ch&#39;].&#34;.format(
            tmp.files
        )
        A_mask = tmp[&#34;mask&#34;] if &#34;mask&#34; in tmp.files else None
        A_ch = list(tmp[&#34;ch&#34;]) if &#34;ch&#34; in tmp.files else None
        # generate img object
        return cls(img_arr=tmp[&#34;img&#34;], channels=A_ch, mask=A_mask)

    def to_npz(self, file):
        &#34;&#34;&#34;
        Save img object to compressed `.npz` file

        Parameters
        ----------
        file : str
            Path to `.npz` file in which to save img object and metadata

        Returns
        -------
        Writes object to `file`
        &#34;&#34;&#34;
        print(&#34;Saving img object to {}...&#34;.format(file))
        if self.mask is None:
            np.savez_compressed(file, img=self.img, ch=self.ch)
        else:
            np.savez_compressed(file, img=self.img, ch=self.ch, mask=self.mask)

    def clip(self, **kwargs):
        &#34;&#34;&#34;
        Clips outlier values and rescales intensities

        Parameters
        ----------
        **kwargs
            Keyword args to pass to `clip_values()` function

        Returns
        -------
        Clips outlier values from `self.img`
        &#34;&#34;&#34;
        self.img = clip_values(self.img, **kwargs)

    def scale(self, **kwargs):
        &#34;&#34;&#34;
        Scales intensities to [0.0, 1.0]

        Parameters
        ----------
        **kwargs
            Keyword args to pass to `scale_rgb()` function

        Returns
        -------
        Scales intensities of `self.img`
        &#34;&#34;&#34;
        self.img = scale_rgb(self.img, **kwargs)

    def equalize_hist(self, **kwargs):
        &#34;&#34;&#34;
        Contrast Limited Adaptive Histogram Equalization (CLAHE)

        Parameters
        ----------
        **kwargs
            Keyword args to pass to `CLAHE()` function

        Returns
        -------
        `self.img` is updated with exposure-adjusted values
        &#34;&#34;&#34;
        self.img = CLAHE(self.img, **kwargs)

    def blurring(self, filter_name=&#34;gaussian&#34;, sigma=2, **kwargs):
        &#34;&#34;&#34;
        Aplying a filter on the images

        Parameters
        ----------
        filter : str
            str to define which type of filter to apply

        sigma : int
            parameter controlling extent of smoothening
        &#34;&#34;&#34;
        if filter_name == &#34;gaussian&#34;:
            print(&#34;Applying gaussian filter&#34;)
            self.img = filters.gaussian(
                self.img,
                sigma=sigma,
                channel_axis=2,
                **kwargs,
            )
        elif filter_name == &#34;median&#34;:
            print(&#34;Applying median filter&#34;)
            if isinstance(sigma, float):
                sigma = int(sigma)
            for i in range(self.img.shape[2]):
                image_array = self.img[:, :, i]
                image_array_blurred = filters.median(
                    image_array,
                    np.ones(sigma, sigma),
                )
                self.img[:, :, i] = image_array_blurred
        elif filter_name == &#34;bilateral&#34;:
            print(&#34;Applying biltaral filter&#34;)
            self.img = denoise_bilateral(
                self.img, sigma_spatial=sigma, channel_axis=2, **kwargs
            )
        else:
            raise Exception(
                &#34;filter name should be either gaussian, median or bilateral&#34;
            )

    def log_normalize(self, pseudoval=1, mean=None, mask=True):
        &#34;&#34;&#34;
        Log-normalizes values for each marker with `log10(arr/arr.mean() + pseudoval)`

        Parameters
        ----------
        pseudoval : float
            Value to add to image values prior to log-transforming to avoid issues
            with zeros
        mask : bool, optional (default=True)
            Use tissue mask to determine marker mean factor for normalization. Default
            `True`.

        Returns
        -------
        Log-normalizes values in each channel of `self.img`
        &#34;&#34;&#34;
        if mean is not None:
            if mask:
                assert self.mask is not None, &#34;No tissue mask available&#34;
                for i in range(self.img.shape[2]):
                    fact = mean[i]
                    self.img[:, :, i] = np.log10(self.img[:, :, i] / fact + pseudoval)
            else:
                print(&#34;WARNING: Performing normalization without a tissue mask.&#34;)
                for i in range(self.img.shape[2]):
                    fact = mean[i]
                    self.img[:, :, i] = np.log10(self.img[:, :, i] / fact + pseudoval)
        else:
            print(&#34;mean calculated to perform log normalization&#34;)
            if mask:
                assert self.mask is not None, &#34;No tissue mask available&#34;
                for i in range(self.img.shape[2]):
                    fact = self.img[:, :, i].mean()
                    self.img[:, :, i] = np.log10(self.img[:, :, i] / fact + pseudoval)
            else:
                print(&#34;WARNING: Performing normalization without a tissue mask.&#34;)
                for i in range(self.img.shape[2]):
                    fact = self.img[:, :, i].mean()
                    self.img[:, :, i] = np.log10(self.img[:, :, i] / fact + pseudoval)

    def subsample_pixels(self, features, fract=0.2, random_state=16):
        &#34;&#34;&#34;
        Sub-samples fraction of pixels from the image randomly for each channel

        Parameters
        ----------
        features : list of int or str
            Indices or names of MxIF channels to use for tissue labeling
        fract : float, optional (default=0.2)
            Fraction of cluster data from each image to randomly select
            for model building

        Returns
        -------
        tmp : np.array
            Clustering data from `image`

        &#34;&#34;&#34;
        if isinstance(features, int):  # force features into list if single integer
            features = [features]
        if isinstance(features, str):  # force features into int if single string
            features = [self.ch.index(features)]
        if checktype(features):  # force features into list of int if list of strings
            features = [self.ch.index(x) for x in features]
        if features is None:  # if no features are given, use all of them
            features = [x for x in range(self.n_ch)]
        # subsample data for given image
        np.random.seed(random_state)
        tmp = []
        for i in range(self.img.shape[2]):
            tmp.append(self.img[:, :, i][self.mask != 0])
        tmp = np.column_stack(tmp)
        # select cluster data
        i = np.random.choice(tmp.shape[0], int(tmp.shape[0] * fract))
        tmp = tmp[np.ix_(i, features)]
        return tmp

    def downsample(self, fact, func=np.mean):
        &#34;&#34;&#34;
        Downsamples image by applying `func` to `fact` pixels in both directions from
        each pixel

        Parameters
        ----------
        fact : int
            Number of pixels in each direction (x &amp; y) to downsample with
        func : function
            Numpy function to apply to squares of size (fact, fact, :) for downsampling
            (e.g. `np.mean`, `np.max`, `np.sum`)

        Returns
        -------
        self.img and self.mask are downsampled accordingly in place
        &#34;&#34;&#34;
        # downsample mask if mask available
        if self.mask is not None:
            self.mask = block_reduce(
                self.mask, block_size=(fact, fact), func=func, cval=0
            )
        # downsample image
        self.img = block_reduce(self.img, block_size=(fact, fact, 1), func=func, cval=0)

    def calculate_non_zero_mean(self):
        &#34;&#34;&#34;
        Calculate mean estimator for the given image array avoiding mask pixels or
        pixels with value 0

        Parameters
        ----------

        Returns
        -------
        mean_estimator : list of float
            List of mean estimator (mean*pixel) values for each channel
        pixels : int
            pixel count for the image excluding masked pixels
        &#34;&#34;&#34;
        image = self.img
        pixels = np.count_nonzero(image != 0)
        mean_estimator = []
        for i in range(image.shape[2]):
            ar = image[:, :, i]
            mean = ar[ar != 0].mean()
            mean_estimator.append(mean * pixels)
        return mean_estimator, pixels

    def create_tissue_mask(self, features=None, fract=0.2):
        &#34;&#34;&#34;
        Create tissue mask

        Parameters
        ----------
        features : list of int or str
            Indices or names of MxIF channels to use for tissue labeling
        fract : float, optional (default=0.2)
            Fraction of cluster data from each image to randomly select
            for model building

        Returns
        -------
        a numpy array as tissue mask set to self.mask
        &#34;&#34;&#34;
        # create a copy of the image
        image_cp = self.copy()
        # create a temporary tissue mask that covers no region
        w, h, d = image_cp.img.shape
        image_cp.mask = np.ones((w, h))
        # log normalization on image
        image_cp.log_normalize()
        # apply gaussian filter
        image_cp.img = filters.gaussian(image_cp.img, sigma=2, channel_axis=2)
        # subsample data to build kmeans model
        subsampled_data = image_cp.subsample_pixels(features, fract=fract)
        cluster_data = np.row_stack(subsampled_data)
        # reshape image for prediction
        image_ar_reshape = image_cp.img.reshape((w * h, d))
        # build kmeans model with 2 clusters
        kmeans = KMeans(n_clusters=2, random_state=18).fit(cluster_data)
        labels = kmeans.predict(image_ar_reshape).astype(float)
        tID = labels.reshape((w, h))
        # check if the background is labelled as 0 or 1
        scores = kmeans.cluster_centers_
        mean = scores.mean()
        std = scores.std()
        z_scores = (scores - mean) / std
        if z_scores[0].mean() &gt; 0:
            where_0 = np.where(tID == 0.0)
            tID[where_0] = 0.5
            where_1 = np.where(tID == 1.0)
            tID[where_1] = 0.0
            where_05 = np.where(tID == 0.5)
            tID[where_05] = 1.0
        self.mask = tID

    def show(
        self,
        channels=None,
        RGB=False,
        cbar=False,
        mask_out=True,
        ncols=4,
        figsize=(7, 7),
        save_to=None,
        **kwargs,
    ):
        &#34;&#34;&#34;
        Plot image

        Parameters
        ----------
        channels : tuple of int or None, optional (default=`None`)
            List of channels by index or name to show
        RGB : bool
            Treat 3- or 4-dimensional array as RGB image. If `False`, plot channels
            individually.
        cbar : bool
            Show colorbar for scale of image intensities if plotting individual
            channels.
        mask_out : bool, optional (default=`True`)
            Mask out non-tissue pixels prior to showing
        ncols : int
            Number of columns for gridspec if plotting individual channels.
        figsize : tuple of float
            Size in inches of output figure.
        save_to : str or None
            Path to image file to save results. If `None`, show figure.
        **kwargs
            Arguments to pass to `plt.imshow()` function.

        Returns
        -------
        Matplotlib object (if plotting one feature or RGB) or gridspec object (for
        multiple features). Saves plot to file if `save_to` is not `None`.
        &#34;&#34;&#34;
        # if only one feature (2D), plot it quickly
        if self.img.ndim == 2:
            fig = plt.figure(figsize=figsize)
            if self.mask is not None and mask_out:
                im_tmp = self.img.copy()  # make copy for masking
                im_tmp[:, :][self.mask == 0] = np.nan  # area outside mask to NaN
                plt.imshow(im_tmp, **kwargs)
            else:
                plt.imshow(self.img, **kwargs)
            plt.tick_params(labelbottom=False, labelleft=False)
            sns.despine(bottom=True, left=True)
            if cbar:
                plt.colorbar(shrink=0.8)
            plt.tight_layout()
            if save_to:
                plt.savefig(
                    fname=save_to, transparent=True, bbox_inches=&#34;tight&#34;, dpi=800
                )
            return fig
        # if image has multiple channels, plot them in gridspec
        if isinstance(channels, int):  # force channels into list if single integer
            channels = [channels]
        if isinstance(channels, str):  # force channels into int if single string
            channels = [self.ch.index(channels)]
        if checktype(channels):  # force channels into list of int if list of strings
            channels = [self.ch.index(x) for x in channels]
        if channels is None:  # if no channels are given, use all of them
            channels = [x for x in range(self.n_ch)]
        assert (
            len(channels) &lt;= self.n_ch
        ), &#34;Too many channels given: image has {}, expected {}&#34;.format(
            self.n_ch, len(channels)
        )
        if RGB:
            # if third dim has 3 or 4 features, treat as RGB and plot it quickly
            assert (self.img.ndim == 3) &amp; (
                len(channels) == 3
            ), &#34;Need 3 dimensions and 3 given channels for an RGB image; shape = {}; channels given = {}&#34;.format(
                self.img.shape, len(channels)
            )
            fig = plt.figure(figsize=figsize)
            # rearrange channels to specified order
            im_tmp = np.dstack(
                [
                    self.img[:, :, channels[0]],
                    self.img[:, :, channels[1]],
                    self.img[:, :, channels[2]],
                ]
            )
            if self.mask is not None and mask_out:
                for i in [0, 1, 2]:  # for 3-channel image
                    im_tmp[:, :, i][self.mask == 0] = np.nan  # area outside mask NaN
            plt.imshow(im_tmp, **kwargs)
            # add legend for channel IDs
            custom_lines = [
                Line2D([0], [0], color=(1, 0, 0), lw=5),
                Line2D([0], [0], color=(0, 1, 0), lw=5),
                Line2D([0], [0], color=(0, 0, 1), lw=5),
            ]
            plt.legend(custom_lines, [self.ch[x] for x in channels], fontsize=&#34;medium&#34;)
            plt.tick_params(labelbottom=False, labelleft=False)
            sns.despine(bottom=True, left=True)
            plt.tight_layout()
            if save_to:
                plt.savefig(
                    fname=save_to, transparent=True, bbox_inches=&#34;tight&#34;, dpi=300
                )
            return fig
        # calculate gridspec dimensions
        if len(channels) &lt;= ncols:
            n_rows, n_cols = 1, len(channels)
        else:
            n_rows, n_cols = ceil(len(channels) / ncols), ncols
        fig = plt.figure(figsize=(ncols * n_cols, ncols * n_rows))
        # arrange axes as subplots
        gs = gridspec.GridSpec(n_rows, n_cols, figure=fig)
        # add plots to axes
        i = 0
        for channel in channels:
            ax = plt.subplot(gs[i])
            if self.mask is not None and mask_out:
                im_tmp = self.img[:, :, channel].copy()  # make copy for masking
                im_tmp[self.mask == 0] = np.nan  # area outside mask NaN
                im = ax.imshow(im_tmp, **kwargs)
            else:
                im = ax.imshow(self.img[:, :, channel], **kwargs)
            ax.tick_params(labelbottom=False, labelleft=False)
            sns.despine(bottom=True, left=True)
            ax.set_title(
                label=self.ch[channel],
                loc=&#34;left&#34;,
                fontweight=&#34;bold&#34;,
                fontsize=16,
            )
            if cbar:
                _ = plt.colorbar(im, shrink=0.8)
            i = i + 1
        fig.tight_layout()
        if save_to:
            plt.savefig(fname=save_to, transparent=True, bbox_inches=&#34;tight&#34;, dpi=300)
        return fig

    def plot_image_histogram(self, channels=None, ncols=4, save_to=None, **kwargs):

        &#34;&#34;&#34;
        Plot image histogram

        Parameters
        ----------
        channels : tuple of int or None, optional (default=`None`)
            List of channels by index or name to show
        ncols : int
            Number sof columns for gridspec if plotting individual channels.
        save_to : str or None
            Path to image file to save results. If `None`, show figure.
        **kwargs
            Arguments to pass to `plt.imshow()` function.

        Returns
        -------
        Gridspec object (for multiple features). Saves plot to file if `save_to` is
        not `None`.
        &#34;&#34;&#34;
        # calculate gridspec dimensions
        if len(channels) &lt;= ncols:
            n_rows, n_cols = 1, len(channels)
        else:
            n_rows, n_cols = ceil(len(channels) / ncols), ncols
        fig = plt.figure(figsize=(ncols * n_cols, ncols * n_rows))
        # arrange axes as subplots
        gs = gridspec.GridSpec(n_rows, n_cols, figure=fig)
        # add plots to axes
        i = 0
        for channel in channels:
            ax = plt.subplot(gs[i])
            data = self[channel].copy()
            ax.hist(data.ravel(), bins=100, **kwargs)
            ax.set_title(channel, fontweight=&#34;bold&#34;, fontsize=16)
            i = i + 1
        fig.tight_layout()
        plt.show()
        if save_to:
            plt.savefig(fname=save_to, transparent=True, bbox_inches=&#34;tight&#34;, dpi=300)
        return gs</code></pre>
</details>
<h3>Static methods</h3>
<dl>
<dt id="MILWRM.img.from_npz"><code class="name flex">
<span>def <span class="ident">from_npz</span></span>(<span>file)</span>
</code></dt>
<dd>
<div class="desc"><p>Initialize img class from <code>.npz</code> file</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>file</code></strong> :&ensp;<code>str</code></dt>
<dd>Path to <code>.npz</code> file containing saved img object and metadata</dd>
</dl>
<h2 id="returns">Returns</h2>
<p><code><a title="MILWRM.img" href="#MILWRM.img">img</a></code> object</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">@classmethod
def from_npz(cls, file):
    &#34;&#34;&#34;
    Initialize img class from `.npz` file

    Parameters
    ----------
    file : str
        Path to `.npz` file containing saved img object and metadata

    Returns
    -------
    `img` object
    &#34;&#34;&#34;
    print(&#34;Loading img object from {}...&#34;.format(file))
    tmp = np.load(file)  # load from .npz compressed file
    assert (
        &#34;img&#34; in tmp.files
    ), &#34;Unexpected files in .npz: {}, expected [&#39;img&#39;,&#39;mask&#39;,&#39;ch&#39;].&#34;.format(
        tmp.files
    )
    A_mask = tmp[&#34;mask&#34;] if &#34;mask&#34; in tmp.files else None
    A_ch = list(tmp[&#34;ch&#34;]) if &#34;ch&#34; in tmp.files else None
    # generate img object
    return cls(img_arr=tmp[&#34;img&#34;], channels=A_ch, mask=A_mask)</code></pre>
</details>
</dd>
<dt id="MILWRM.img.from_tiffs"><code class="name flex">
<span>def <span class="ident">from_tiffs</span></span>(<span>tiffdir, channels, common_strings=None, mask=None)</span>
</code></dt>
<dd>
<div class="desc"><p>Initialize img class from <code>.tif</code> files</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>tiffdir</code></strong> :&ensp;<code>str</code></dt>
<dd>Path to directory containing <code>.tif</code> files for a multiplexed image</dd>
<dt><strong><code>channels</code></strong> :&ensp;<code>list</code> of <code>str</code></dt>
<dd>List of channels present in <code>.tif</code> file names (case-sensitive)
corresponding to img.shape[2] e.g. <code>("ACTG1","BCATENIN","DAPI",...)</code></dd>
<dt><strong><code>common_strings</code></strong> :&ensp;<code>str, list</code> of <code>str,</code> or <code>None</code>, optional <code>(default=None)</code></dt>
<dd>Strings to look for in all <code>.tif</code> files in <code>tiffdir</code> corresponding to
<code>channels</code> e.g. <code>("WD86055_", "_region_001.tif")</code> for files named
"WD86055_[MARKERNAME]_region_001.tif". If <code>None</code>, assume that only 1 image
for each marker in <code>channels</code> is present in <code>tiffdir</code>.</dd>
<dt><strong><code>mask</code></strong> :&ensp;<code>str</code>, optional <code>(default=None)</code></dt>
<dd>Name of mask defining pixels containing tissue in the image, present in
<code>.tif</code> file names (case-sensitive) e.g. "_01_TISSUE_MASK.tif"</dd>
</dl>
<h2 id="returns">Returns</h2>
<p><code><a title="MILWRM.img" href="#MILWRM.img">img</a></code> object</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">@classmethod
def from_tiffs(cls, tiffdir, channels, common_strings=None, mask=None):
    &#34;&#34;&#34;
    Initialize img class from `.tif` files

    Parameters
    ----------
    tiffdir : str
        Path to directory containing `.tif` files for a multiplexed image
    channels : list of str
        List of channels present in `.tif` file names (case-sensitive)
        corresponding to img.shape[2] e.g. `(&#34;ACTG1&#34;,&#34;BCATENIN&#34;,&#34;DAPI&#34;,...)`
    common_strings : str, list of str, or `None`, optional (default=None)
        Strings to look for in all `.tif` files in `tiffdir` corresponding to
        `channels` e.g. `(&#34;WD86055_&#34;, &#34;_region_001.tif&#34;)` for files named
        &#34;WD86055_[MARKERNAME]_region_001.tif&#34;. If `None`, assume that only 1 image
        for each marker in `channels` is present in `tiffdir`.
    mask : str, optional (default=None)
        Name of mask defining pixels containing tissue in the image, present in
        `.tif` file names (case-sensitive) e.g. &#34;_01_TISSUE_MASK.tif&#34;

    Returns
    -------
    `img` object
    &#34;&#34;&#34;
    if common_strings is not None:
        # coerce single string to list
        if isinstance(common_strings, str):
            common_strings = [common_strings]
    A = []  # list for dumping numpy arrays
    for channel in channels:
        if common_strings is None:
            # find file matching all common_strings and channel name
            f = [f for f in os.listdir(tiffdir) if channel in f]
        else:
            # find file matching all common_strings and channel name
            f = [
                f
                for f in os.listdir(tiffdir)
                if all(x in f for x in common_strings + [channel])
            ]
        # assertions so we only get one file per channel
        assert len(f) != 0, &#34;No file found with channel {}&#34;.format(channel)
        assert (
            len(f) == 1
        ), &#34;More than one match found for file with channel {}&#34;.format(channel)
        f = os.path.join(tiffdir, f[0])  # get full path to file for reading
        print(&#34;Reading marker {} from {}&#34;.format(channel, f))
        tmp = imread(f)  # read in .tif file
        A.append(tmp)  # append numpy array to list
    A_arr = np.dstack(
        A
    )  # stack numpy arrays in new dimension (third dim is channel)
    print(&#34;Final image array of shape: {}&#34;.format(A_arr.shape))
    # read in tissue mask if available
    if mask is not None:
        f = [f for f in os.listdir(tiffdir) if mask in f]
        # assertions so we only get one mask file
        assert len(f) != 0, &#34;No tissue mask file found&#34;
        assert len(f) == 1, &#34;More than one match found for tissue mask file&#34;
        f = os.path.join(tiffdir, f[0])  # get full path to file for reading
        print(&#34;Reading tissue mask from {}&#34;.format(f))
        A_mask = imread(f)  # read in .tif file
        assert (
            A_mask.shape == A_arr.shape[:2]
        ), &#34;Mask (shape: {}) is not the same shape as marker images (shape: {})&#34;.format(
            A_mask.shape, A_arr.shape[:2]
        )
        print(&#34;Final mask array of shape: {}&#34;.format(A_mask.shape))
    else:
        A_mask = None
    # generate img object
    return cls(img_arr=A_arr, channels=channels, mask=A_mask)</code></pre>
</details>
</dd>
</dl>
<h3>Methods</h3>
<dl>
<dt id="MILWRM.img.blurring"><code class="name flex">
<span>def <span class="ident">blurring</span></span>(<span>self, filter_name='gaussian', sigma=2, **kwargs)</span>
</code></dt>
<dd>
<div class="desc"><p>Aplying a filter on the images</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>filter</code></strong> :&ensp;<code>str</code></dt>
<dd>str to define which type of filter to apply</dd>
<dt><strong><code>sigma</code></strong> :&ensp;<code>int</code></dt>
<dd>parameter controlling extent of smoothening</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def blurring(self, filter_name=&#34;gaussian&#34;, sigma=2, **kwargs):
    &#34;&#34;&#34;
    Aplying a filter on the images

    Parameters
    ----------
    filter : str
        str to define which type of filter to apply

    sigma : int
        parameter controlling extent of smoothening
    &#34;&#34;&#34;
    if filter_name == &#34;gaussian&#34;:
        print(&#34;Applying gaussian filter&#34;)
        self.img = filters.gaussian(
            self.img,
            sigma=sigma,
            channel_axis=2,
            **kwargs,
        )
    elif filter_name == &#34;median&#34;:
        print(&#34;Applying median filter&#34;)
        if isinstance(sigma, float):
            sigma = int(sigma)
        for i in range(self.img.shape[2]):
            image_array = self.img[:, :, i]
            image_array_blurred = filters.median(
                image_array,
                np.ones(sigma, sigma),
            )
            self.img[:, :, i] = image_array_blurred
    elif filter_name == &#34;bilateral&#34;:
        print(&#34;Applying biltaral filter&#34;)
        self.img = denoise_bilateral(
            self.img, sigma_spatial=sigma, channel_axis=2, **kwargs
        )
    else:
        raise Exception(
            &#34;filter name should be either gaussian, median or bilateral&#34;
        )</code></pre>
</details>
</dd>
<dt id="MILWRM.img.calculate_non_zero_mean"><code class="name flex">
<span>def <span class="ident">calculate_non_zero_mean</span></span>(<span>self)</span>
</code></dt>
<dd>
<div class="desc"><p>Calculate mean estimator for the given image array avoiding mask pixels or
pixels with value 0</p>
<h2 id="parameters">Parameters</h2>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>mean_estimator</code></strong> :&ensp;<code>list</code> of <code>float</code></dt>
<dd>List of mean estimator (mean*pixel) values for each channel</dd>
<dt><strong><code>pixels</code></strong> :&ensp;<code>int</code></dt>
<dd>pixel count for the image excluding masked pixels</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def calculate_non_zero_mean(self):
    &#34;&#34;&#34;
    Calculate mean estimator for the given image array avoiding mask pixels or
    pixels with value 0

    Parameters
    ----------

    Returns
    -------
    mean_estimator : list of float
        List of mean estimator (mean*pixel) values for each channel
    pixels : int
        pixel count for the image excluding masked pixels
    &#34;&#34;&#34;
    image = self.img
    pixels = np.count_nonzero(image != 0)
    mean_estimator = []
    for i in range(image.shape[2]):
        ar = image[:, :, i]
        mean = ar[ar != 0].mean()
        mean_estimator.append(mean * pixels)
    return mean_estimator, pixels</code></pre>
</details>
</dd>
<dt id="MILWRM.img.clip"><code class="name flex">
<span>def <span class="ident">clip</span></span>(<span>self, **kwargs)</span>
</code></dt>
<dd>
<div class="desc"><p>Clips outlier values and rescales intensities</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>**kwargs</code></strong></dt>
<dd>Keyword args to pass to <code>clip_values()</code> function</dd>
</dl>
<h2 id="returns">Returns</h2>
<p>Clips outlier values from <code>self.img</code></p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def clip(self, **kwargs):
    &#34;&#34;&#34;
    Clips outlier values and rescales intensities

    Parameters
    ----------
    **kwargs
        Keyword args to pass to `clip_values()` function

    Returns
    -------
    Clips outlier values from `self.img`
    &#34;&#34;&#34;
    self.img = clip_values(self.img, **kwargs)</code></pre>
</details>
</dd>
<dt id="MILWRM.img.copy"><code class="name flex">
<span>def <span class="ident">copy</span></span>(<span>self) â€‘>Â <a title="MILWRM.MxIF.img" href="MxIF.html#MILWRM.MxIF.img">img</a></span>
</code></dt>
<dd>
<div class="desc"><p>Full copy of img object</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def copy(self) -&gt; &#34;img&#34;:
    &#34;&#34;&#34;Full copy of img object&#34;&#34;&#34;
    new = {}
    for key in [&#34;img&#34;, &#34;ch&#34;, &#34;mask&#34;]:
        attr = self.__getattribute__(key)
        if attr is not None:
            new[key] = attr.copy()
        else:
            new[key] = None
    return img(img_arr=new[&#34;img&#34;], channels=new[&#34;ch&#34;], mask=new[&#34;mask&#34;])</code></pre>
</details>
</dd>
<dt id="MILWRM.img.create_tissue_mask"><code class="name flex">
<span>def <span class="ident">create_tissue_mask</span></span>(<span>self, features=None, fract=0.2)</span>
</code></dt>
<dd>
<div class="desc"><p>Create tissue mask</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>features</code></strong> :&ensp;<code>list</code> of <code>int</code> or <code>str</code></dt>
<dd>Indices or names of MxIF channels to use for tissue labeling</dd>
<dt><strong><code>fract</code></strong> :&ensp;<code>float</code>, optional <code>(default=0.2)</code></dt>
<dd>Fraction of cluster data from each image to randomly select
for model building</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>a numpy array as tissue mask set to self.mask</code></dt>
<dd>&nbsp;</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def create_tissue_mask(self, features=None, fract=0.2):
    &#34;&#34;&#34;
    Create tissue mask

    Parameters
    ----------
    features : list of int or str
        Indices or names of MxIF channels to use for tissue labeling
    fract : float, optional (default=0.2)
        Fraction of cluster data from each image to randomly select
        for model building

    Returns
    -------
    a numpy array as tissue mask set to self.mask
    &#34;&#34;&#34;
    # create a copy of the image
    image_cp = self.copy()
    # create a temporary tissue mask that covers no region
    w, h, d = image_cp.img.shape
    image_cp.mask = np.ones((w, h))
    # log normalization on image
    image_cp.log_normalize()
    # apply gaussian filter
    image_cp.img = filters.gaussian(image_cp.img, sigma=2, channel_axis=2)
    # subsample data to build kmeans model
    subsampled_data = image_cp.subsample_pixels(features, fract=fract)
    cluster_data = np.row_stack(subsampled_data)
    # reshape image for prediction
    image_ar_reshape = image_cp.img.reshape((w * h, d))
    # build kmeans model with 2 clusters
    kmeans = KMeans(n_clusters=2, random_state=18).fit(cluster_data)
    labels = kmeans.predict(image_ar_reshape).astype(float)
    tID = labels.reshape((w, h))
    # check if the background is labelled as 0 or 1
    scores = kmeans.cluster_centers_
    mean = scores.mean()
    std = scores.std()
    z_scores = (scores - mean) / std
    if z_scores[0].mean() &gt; 0:
        where_0 = np.where(tID == 0.0)
        tID[where_0] = 0.5
        where_1 = np.where(tID == 1.0)
        tID[where_1] = 0.0
        where_05 = np.where(tID == 0.5)
        tID[where_05] = 1.0
    self.mask = tID</code></pre>
</details>
</dd>
<dt id="MILWRM.img.downsample"><code class="name flex">
<span>def <span class="ident">downsample</span></span>(<span>self, fact, func=&lt;function mean&gt;)</span>
</code></dt>
<dd>
<div class="desc"><p>Downsamples image by applying <code>func</code> to <code>fact</code> pixels in both directions from
each pixel</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>fact</code></strong> :&ensp;<code>int</code></dt>
<dd>Number of pixels in each direction (x &amp; y) to downsample with</dd>
<dt><strong><code>func</code></strong> :&ensp;<code>function</code></dt>
<dd>Numpy function to apply to squares of size (fact, fact, :) for downsampling
(e.g. <code>np.mean</code>, <code>np.max</code>, <code>np.sum</code>)</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>self.img and self.mask are downsampled accordingly in place</code></dt>
<dd>&nbsp;</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def downsample(self, fact, func=np.mean):
    &#34;&#34;&#34;
    Downsamples image by applying `func` to `fact` pixels in both directions from
    each pixel

    Parameters
    ----------
    fact : int
        Number of pixels in each direction (x &amp; y) to downsample with
    func : function
        Numpy function to apply to squares of size (fact, fact, :) for downsampling
        (e.g. `np.mean`, `np.max`, `np.sum`)

    Returns
    -------
    self.img and self.mask are downsampled accordingly in place
    &#34;&#34;&#34;
    # downsample mask if mask available
    if self.mask is not None:
        self.mask = block_reduce(
            self.mask, block_size=(fact, fact), func=func, cval=0
        )
    # downsample image
    self.img = block_reduce(self.img, block_size=(fact, fact, 1), func=func, cval=0)</code></pre>
</details>
</dd>
<dt id="MILWRM.img.equalize_hist"><code class="name flex">
<span>def <span class="ident">equalize_hist</span></span>(<span>self, **kwargs)</span>
</code></dt>
<dd>
<div class="desc"><p>Contrast Limited Adaptive Histogram Equalization (CLAHE)</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>**kwargs</code></strong></dt>
<dd>Keyword args to pass to <code>CLAHE()</code> function</dd>
</dl>
<h2 id="returns">Returns</h2>
<p><code>self.img</code> is updated with exposure-adjusted values</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def equalize_hist(self, **kwargs):
    &#34;&#34;&#34;
    Contrast Limited Adaptive Histogram Equalization (CLAHE)

    Parameters
    ----------
    **kwargs
        Keyword args to pass to `CLAHE()` function

    Returns
    -------
    `self.img` is updated with exposure-adjusted values
    &#34;&#34;&#34;
    self.img = CLAHE(self.img, **kwargs)</code></pre>
</details>
</dd>
<dt id="MILWRM.img.log_normalize"><code class="name flex">
<span>def <span class="ident">log_normalize</span></span>(<span>self, pseudoval=1, mean=None, mask=True)</span>
</code></dt>
<dd>
<div class="desc"><p>Log-normalizes values for each marker with <code>log10(arr/arr.mean() + pseudoval)</code></p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>pseudoval</code></strong> :&ensp;<code>float</code></dt>
<dd>Value to add to image values prior to log-transforming to avoid issues
with zeros</dd>
<dt><strong><code>mask</code></strong> :&ensp;<code>bool</code>, optional <code>(default=True)</code></dt>
<dd>Use tissue mask to determine marker mean factor for normalization. Default
<code>True</code>.</dd>
</dl>
<h2 id="returns">Returns</h2>
<p>Log-normalizes values in each channel of <code>self.img</code></p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def log_normalize(self, pseudoval=1, mean=None, mask=True):
    &#34;&#34;&#34;
    Log-normalizes values for each marker with `log10(arr/arr.mean() + pseudoval)`

    Parameters
    ----------
    pseudoval : float
        Value to add to image values prior to log-transforming to avoid issues
        with zeros
    mask : bool, optional (default=True)
        Use tissue mask to determine marker mean factor for normalization. Default
        `True`.

    Returns
    -------
    Log-normalizes values in each channel of `self.img`
    &#34;&#34;&#34;
    if mean is not None:
        if mask:
            assert self.mask is not None, &#34;No tissue mask available&#34;
            for i in range(self.img.shape[2]):
                fact = mean[i]
                self.img[:, :, i] = np.log10(self.img[:, :, i] / fact + pseudoval)
        else:
            print(&#34;WARNING: Performing normalization without a tissue mask.&#34;)
            for i in range(self.img.shape[2]):
                fact = mean[i]
                self.img[:, :, i] = np.log10(self.img[:, :, i] / fact + pseudoval)
    else:
        print(&#34;mean calculated to perform log normalization&#34;)
        if mask:
            assert self.mask is not None, &#34;No tissue mask available&#34;
            for i in range(self.img.shape[2]):
                fact = self.img[:, :, i].mean()
                self.img[:, :, i] = np.log10(self.img[:, :, i] / fact + pseudoval)
        else:
            print(&#34;WARNING: Performing normalization without a tissue mask.&#34;)
            for i in range(self.img.shape[2]):
                fact = self.img[:, :, i].mean()
                self.img[:, :, i] = np.log10(self.img[:, :, i] / fact + pseudoval)</code></pre>
</details>
</dd>
<dt id="MILWRM.img.plot_image_histogram"><code class="name flex">
<span>def <span class="ident">plot_image_histogram</span></span>(<span>self, channels=None, ncols=4, save_to=None, **kwargs)</span>
</code></dt>
<dd>
<div class="desc"><p>Plot image histogram</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>channels</code></strong> :&ensp;<code>tuple</code> of <code>int</code> or <code>None</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>List of channels by index or name to show</dd>
<dt><strong><code>ncols</code></strong> :&ensp;<code>int</code></dt>
<dd>Number sof columns for gridspec if plotting individual channels.</dd>
<dt><strong><code>save_to</code></strong> :&ensp;<code>str</code> or <code>None</code></dt>
<dd>Path to image file to save results. If <code>None</code>, show figure.</dd>
<dt><strong><code>**kwargs</code></strong></dt>
<dd>Arguments to pass to <code>plt.imshow()</code> function.</dd>
</dl>
<h2 id="returns">Returns</h2>
<p>Gridspec object (for multiple features). Saves plot to file if <code>save_to</code> is
not <code>None</code>.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def plot_image_histogram(self, channels=None, ncols=4, save_to=None, **kwargs):

    &#34;&#34;&#34;
    Plot image histogram

    Parameters
    ----------
    channels : tuple of int or None, optional (default=`None`)
        List of channels by index or name to show
    ncols : int
        Number sof columns for gridspec if plotting individual channels.
    save_to : str or None
        Path to image file to save results. If `None`, show figure.
    **kwargs
        Arguments to pass to `plt.imshow()` function.

    Returns
    -------
    Gridspec object (for multiple features). Saves plot to file if `save_to` is
    not `None`.
    &#34;&#34;&#34;
    # calculate gridspec dimensions
    if len(channels) &lt;= ncols:
        n_rows, n_cols = 1, len(channels)
    else:
        n_rows, n_cols = ceil(len(channels) / ncols), ncols
    fig = plt.figure(figsize=(ncols * n_cols, ncols * n_rows))
    # arrange axes as subplots
    gs = gridspec.GridSpec(n_rows, n_cols, figure=fig)
    # add plots to axes
    i = 0
    for channel in channels:
        ax = plt.subplot(gs[i])
        data = self[channel].copy()
        ax.hist(data.ravel(), bins=100, **kwargs)
        ax.set_title(channel, fontweight=&#34;bold&#34;, fontsize=16)
        i = i + 1
    fig.tight_layout()
    plt.show()
    if save_to:
        plt.savefig(fname=save_to, transparent=True, bbox_inches=&#34;tight&#34;, dpi=300)
    return gs</code></pre>
</details>
</dd>
<dt id="MILWRM.img.scale"><code class="name flex">
<span>def <span class="ident">scale</span></span>(<span>self, **kwargs)</span>
</code></dt>
<dd>
<div class="desc"><p>Scales intensities to [0.0, 1.0]</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>**kwargs</code></strong></dt>
<dd>Keyword args to pass to <code>scale_rgb()</code> function</dd>
</dl>
<h2 id="returns">Returns</h2>
<p>Scales intensities of <code>self.img</code></p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def scale(self, **kwargs):
    &#34;&#34;&#34;
    Scales intensities to [0.0, 1.0]

    Parameters
    ----------
    **kwargs
        Keyword args to pass to `scale_rgb()` function

    Returns
    -------
    Scales intensities of `self.img`
    &#34;&#34;&#34;
    self.img = scale_rgb(self.img, **kwargs)</code></pre>
</details>
</dd>
<dt id="MILWRM.img.show"><code class="name flex">
<span>def <span class="ident">show</span></span>(<span>self, channels=None, RGB=False, cbar=False, mask_out=True, ncols=4, figsize=(7, 7), save_to=None, **kwargs)</span>
</code></dt>
<dd>
<div class="desc"><p>Plot image</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>channels</code></strong> :&ensp;<code>tuple</code> of <code>int</code> or <code>None</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>List of channels by index or name to show</dd>
<dt><strong><code>RGB</code></strong> :&ensp;<code>bool</code></dt>
<dd>Treat 3- or 4-dimensional array as RGB image. If <code>False</code>, plot channels
individually.</dd>
<dt><strong><code>cbar</code></strong> :&ensp;<code>bool</code></dt>
<dd>Show colorbar for scale of image intensities if plotting individual
channels.</dd>
<dt><strong><code>mask_out</code></strong> :&ensp;<code>bool</code>, optional <code>(default=</code>True<code>)</code></dt>
<dd>Mask out non-tissue pixels prior to showing</dd>
<dt><strong><code>ncols</code></strong> :&ensp;<code>int</code></dt>
<dd>Number of columns for gridspec if plotting individual channels.</dd>
<dt><strong><code>figsize</code></strong> :&ensp;<code>tuple</code> of <code>float</code></dt>
<dd>Size in inches of output figure.</dd>
<dt><strong><code>save_to</code></strong> :&ensp;<code>str</code> or <code>None</code></dt>
<dd>Path to image file to save results. If <code>None</code>, show figure.</dd>
<dt><strong><code>**kwargs</code></strong></dt>
<dd>Arguments to pass to <code>plt.imshow()</code> function.</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>Matplotlib object (if plotting one feature</code> or <code>RGB)</code> or <code>gridspec object (for</code></dt>
<dd>&nbsp;</dd>
</dl>
<p>multiple features). Saves plot to file if <code>save_to</code> is not <code>None</code>.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def show(
    self,
    channels=None,
    RGB=False,
    cbar=False,
    mask_out=True,
    ncols=4,
    figsize=(7, 7),
    save_to=None,
    **kwargs,
):
    &#34;&#34;&#34;
    Plot image

    Parameters
    ----------
    channels : tuple of int or None, optional (default=`None`)
        List of channels by index or name to show
    RGB : bool
        Treat 3- or 4-dimensional array as RGB image. If `False`, plot channels
        individually.
    cbar : bool
        Show colorbar for scale of image intensities if plotting individual
        channels.
    mask_out : bool, optional (default=`True`)
        Mask out non-tissue pixels prior to showing
    ncols : int
        Number of columns for gridspec if plotting individual channels.
    figsize : tuple of float
        Size in inches of output figure.
    save_to : str or None
        Path to image file to save results. If `None`, show figure.
    **kwargs
        Arguments to pass to `plt.imshow()` function.

    Returns
    -------
    Matplotlib object (if plotting one feature or RGB) or gridspec object (for
    multiple features). Saves plot to file if `save_to` is not `None`.
    &#34;&#34;&#34;
    # if only one feature (2D), plot it quickly
    if self.img.ndim == 2:
        fig = plt.figure(figsize=figsize)
        if self.mask is not None and mask_out:
            im_tmp = self.img.copy()  # make copy for masking
            im_tmp[:, :][self.mask == 0] = np.nan  # area outside mask to NaN
            plt.imshow(im_tmp, **kwargs)
        else:
            plt.imshow(self.img, **kwargs)
        plt.tick_params(labelbottom=False, labelleft=False)
        sns.despine(bottom=True, left=True)
        if cbar:
            plt.colorbar(shrink=0.8)
        plt.tight_layout()
        if save_to:
            plt.savefig(
                fname=save_to, transparent=True, bbox_inches=&#34;tight&#34;, dpi=800
            )
        return fig
    # if image has multiple channels, plot them in gridspec
    if isinstance(channels, int):  # force channels into list if single integer
        channels = [channels]
    if isinstance(channels, str):  # force channels into int if single string
        channels = [self.ch.index(channels)]
    if checktype(channels):  # force channels into list of int if list of strings
        channels = [self.ch.index(x) for x in channels]
    if channels is None:  # if no channels are given, use all of them
        channels = [x for x in range(self.n_ch)]
    assert (
        len(channels) &lt;= self.n_ch
    ), &#34;Too many channels given: image has {}, expected {}&#34;.format(
        self.n_ch, len(channels)
    )
    if RGB:
        # if third dim has 3 or 4 features, treat as RGB and plot it quickly
        assert (self.img.ndim == 3) &amp; (
            len(channels) == 3
        ), &#34;Need 3 dimensions and 3 given channels for an RGB image; shape = {}; channels given = {}&#34;.format(
            self.img.shape, len(channels)
        )
        fig = plt.figure(figsize=figsize)
        # rearrange channels to specified order
        im_tmp = np.dstack(
            [
                self.img[:, :, channels[0]],
                self.img[:, :, channels[1]],
                self.img[:, :, channels[2]],
            ]
        )
        if self.mask is not None and mask_out:
            for i in [0, 1, 2]:  # for 3-channel image
                im_tmp[:, :, i][self.mask == 0] = np.nan  # area outside mask NaN
        plt.imshow(im_tmp, **kwargs)
        # add legend for channel IDs
        custom_lines = [
            Line2D([0], [0], color=(1, 0, 0), lw=5),
            Line2D([0], [0], color=(0, 1, 0), lw=5),
            Line2D([0], [0], color=(0, 0, 1), lw=5),
        ]
        plt.legend(custom_lines, [self.ch[x] for x in channels], fontsize=&#34;medium&#34;)
        plt.tick_params(labelbottom=False, labelleft=False)
        sns.despine(bottom=True, left=True)
        plt.tight_layout()
        if save_to:
            plt.savefig(
                fname=save_to, transparent=True, bbox_inches=&#34;tight&#34;, dpi=300
            )
        return fig
    # calculate gridspec dimensions
    if len(channels) &lt;= ncols:
        n_rows, n_cols = 1, len(channels)
    else:
        n_rows, n_cols = ceil(len(channels) / ncols), ncols
    fig = plt.figure(figsize=(ncols * n_cols, ncols * n_rows))
    # arrange axes as subplots
    gs = gridspec.GridSpec(n_rows, n_cols, figure=fig)
    # add plots to axes
    i = 0
    for channel in channels:
        ax = plt.subplot(gs[i])
        if self.mask is not None and mask_out:
            im_tmp = self.img[:, :, channel].copy()  # make copy for masking
            im_tmp[self.mask == 0] = np.nan  # area outside mask NaN
            im = ax.imshow(im_tmp, **kwargs)
        else:
            im = ax.imshow(self.img[:, :, channel], **kwargs)
        ax.tick_params(labelbottom=False, labelleft=False)
        sns.despine(bottom=True, left=True)
        ax.set_title(
            label=self.ch[channel],
            loc=&#34;left&#34;,
            fontweight=&#34;bold&#34;,
            fontsize=16,
        )
        if cbar:
            _ = plt.colorbar(im, shrink=0.8)
        i = i + 1
    fig.tight_layout()
    if save_to:
        plt.savefig(fname=save_to, transparent=True, bbox_inches=&#34;tight&#34;, dpi=300)
    return fig</code></pre>
</details>
</dd>
<dt id="MILWRM.img.subsample_pixels"><code class="name flex">
<span>def <span class="ident">subsample_pixels</span></span>(<span>self, features, fract=0.2, random_state=16)</span>
</code></dt>
<dd>
<div class="desc"><p>Sub-samples fraction of pixels from the image randomly for each channel</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>features</code></strong> :&ensp;<code>list</code> of <code>int</code> or <code>str</code></dt>
<dd>Indices or names of MxIF channels to use for tissue labeling</dd>
<dt><strong><code>fract</code></strong> :&ensp;<code>float</code>, optional <code>(default=0.2)</code></dt>
<dd>Fraction of cluster data from each image to randomly select
for model building</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>tmp</code></strong> :&ensp;<code>np.array</code></dt>
<dd>Clustering data from <code>image</code></dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def subsample_pixels(self, features, fract=0.2, random_state=16):
    &#34;&#34;&#34;
    Sub-samples fraction of pixels from the image randomly for each channel

    Parameters
    ----------
    features : list of int or str
        Indices or names of MxIF channels to use for tissue labeling
    fract : float, optional (default=0.2)
        Fraction of cluster data from each image to randomly select
        for model building

    Returns
    -------
    tmp : np.array
        Clustering data from `image`

    &#34;&#34;&#34;
    if isinstance(features, int):  # force features into list if single integer
        features = [features]
    if isinstance(features, str):  # force features into int if single string
        features = [self.ch.index(features)]
    if checktype(features):  # force features into list of int if list of strings
        features = [self.ch.index(x) for x in features]
    if features is None:  # if no features are given, use all of them
        features = [x for x in range(self.n_ch)]
    # subsample data for given image
    np.random.seed(random_state)
    tmp = []
    for i in range(self.img.shape[2]):
        tmp.append(self.img[:, :, i][self.mask != 0])
    tmp = np.column_stack(tmp)
    # select cluster data
    i = np.random.choice(tmp.shape[0], int(tmp.shape[0] * fract))
    tmp = tmp[np.ix_(i, features)]
    return tmp</code></pre>
</details>
</dd>
<dt id="MILWRM.img.to_npz"><code class="name flex">
<span>def <span class="ident">to_npz</span></span>(<span>self, file)</span>
</code></dt>
<dd>
<div class="desc"><p>Save img object to compressed <code>.npz</code> file</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>file</code></strong> :&ensp;<code>str</code></dt>
<dd>Path to <code>.npz</code> file in which to save img object and metadata</dd>
</dl>
<h2 id="returns">Returns</h2>
<p>Writes object to <code>file</code></p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def to_npz(self, file):
    &#34;&#34;&#34;
    Save img object to compressed `.npz` file

    Parameters
    ----------
    file : str
        Path to `.npz` file in which to save img object and metadata

    Returns
    -------
    Writes object to `file`
    &#34;&#34;&#34;
    print(&#34;Saving img object to {}...&#34;.format(file))
    if self.mask is None:
        np.savez_compressed(file, img=self.img, ch=self.ch)
    else:
        np.savez_compressed(file, img=self.img, ch=self.ch, mask=self.mask)</code></pre>
</details>
</dd>
</dl>
</dd>
<dt id="MILWRM.mxif_labeler"><code class="flex name class">
<span>class <span class="ident">mxif_labeler</span></span>
<span>(</span><span>image_df)</span>
</code></dt>
<dd>
<div class="desc"><p>Tissue domain labeling class for multiplex immunofluorescence (MxIF) data</p>
<p>Initialize MxIF tissue labeler class</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>image_df</code></strong> :&ensp;<code>pd.DataFrame object</code></dt>
<dd>Containing MILWRM.MxIF.img objects or str path to compressed npz files,
batch names, mean estimator and pixel count for each image in the
following column order ['Img', 'batch_names', 'mean estimators', 'pixels']</dd>
</dl>
<h2 id="returns">Returns</h2>
<p>Does not return anything. <code>self.images</code> attribute is updated,
<code>self.cluster_data</code> attribute is initiated as <code>None</code>.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class mxif_labeler(tissue_labeler):
    &#34;&#34;&#34;
    Tissue domain labeling class for multiplex immunofluorescence (MxIF) data
    &#34;&#34;&#34;

    def __init__(self, image_df):
        &#34;&#34;&#34;
        Initialize MxIF tissue labeler class

        Parameters
        ----------
        image_df : pd.DataFrame object
            Containing MILWRM.MxIF.img objects or str path to compressed npz files,
            batch names, mean estimator and pixel count for each image in the
            following column order [&#39;Img&#39;, &#39;batch_names&#39;, &#39;mean estimators&#39;, &#39;pixels&#39;]

        Returns
        -------
        Does not return anything. `self.images` attribute is updated,
        `self.cluster_data` attribute is initiated as `None`.
        &#34;&#34;&#34;
        tissue_labeler.__init__(self)  # initialize parent class
        # validate the format of the image_df dataframe
        if np.all(
            image_df.columns == [&#34;Img&#34;, &#34;batch_names&#34;, &#34;mean estimators&#34;, &#34;pixels&#34;]
        ):
            self.image_df = image_df
        else:
            raise Exception(
                &#34;Image_df must be given with these columns in this format [&#39;Img&#39;, &#39;batch_names&#39;, &#39;mean estimators&#39;, &#39;pixels&#39;]&#34;
            )
        if self.image_df[&#34;Img&#34;].apply(isinstance, args=[img]).all():
            self.use_paths = False
        elif self.image_df[&#34;Img&#34;].apply(isinstance, args=[str]).all():
            self.use_paths = True
        else:
            raise Exception(
                &#34;Img column in the dataframe should be either str for paths to the files or mxif.img object&#34;
            )

    def prep_cluster_data(
        self, features, filter_name=&#34;gaussian&#34;, sigma=2, fract=0.2, path_save=None
    ):
        &#34;&#34;&#34;
        Prepare master array for tissue level clustering

        Parameters
        ----------
        features : list of int or str
            Indices or names of MxIF channels to use for tissue labeling
        filter_name : str
            Name of the filter to use - gaussian, median or bilateral
        sigma : float, optional (default=2)
            Standard deviation of Gaussian kernel for blurring
        fract : float, optional (default=0.2)
            Fraction of cluster data from each image to randomly select for model
            building
        path_save : str (default = None)
            Path to save final preprocessed files, if self.use_path is True
            default path_save will raise Exception

        Returns
        -------
        Does not return anything. `self.images` are normalized, blurred and scaled
        according to user parameters. `self.cluster_data` becomes master `np.array`
        for cluster training. Parameters are also captured as attributes for posterity.

        &#34;&#34;&#34;
        if self.cluster_data is not None:
            print(&#34;WARNING: overwriting existing cluster data&#34;)
            self.cluster_data = None
        # save the hyperparams as object attributes
        self.model_features = features
        use_path = self.use_paths
        # calculate the batch wise means
        mean_for_each_batch = {}
        for batch in self.image_df[&#34;batch_names&#34;].unique():
            list_mean_estimators = list(
                self.image_df[self.image_df[&#34;batch_names&#34;] == batch][&#34;mean estimators&#34;]
            )
            mean_estimator_batch = sum(map(np.array, list_mean_estimators))
            pixels = sum(self.image_df[self.image_df[&#34;batch_names&#34;] == batch][&#34;pixels&#34;])
            mean_for_each_batch[batch] = mean_estimator_batch / pixels
        # log_normalize, apply blurring filter, minmax scale each channel and subsample
        subsampled_data = []
        path_to_blurred_npz = []
        for image, batch in zip(self.image_df[&#34;Img&#34;], self.image_df[&#34;batch_names&#34;]):
            tmp = prep_data_single_sample_mxif(
                image,
                use_path=use_path,
                mean=mean_for_each_batch[batch],
                filter_name=filter_name,
                sigma=sigma,
                features=self.model_features,
                fract=fract,
                path_save=path_save,
            )
            if self.use_paths == True:
                subsampled_data.append(tmp[0])
                path_to_blurred_npz.append(tmp[1])
            else:
                subsampled_data.append(tmp)
        batch_labels = [
            [x] * len(subsampled_data[x]) for x in range(len(subsampled_data))
        ]  # batch labels for umap
        self.merged_batch_labels = list(itertools.chain(*batch_labels))
        if self.use_paths == True:
            self.image_df[&#34;Img&#34;] = path_to_blurred_npz
        cluster_data = np.row_stack(subsampled_data)
        # perform z-score normalization on cluster_Data
        scaler = StandardScaler()
        self.scaler = scaler.fit(cluster_data)
        scaled_data = scaler.transform(cluster_data)
        self.cluster_data = scaled_data

    def label_tissue_regions(
        self, k=None, alpha=0.05, plot_out=True, random_state=18, n_jobs=-1
    ):
        &#34;&#34;&#34;
        Perform tissue-level clustering and label pixels in the corresponding
        images.

        Parameters
        ----------
        k : int, optional (default=None)
            Number of tissue regions to define
        alpha: float
            Manually tuned factor on [0.0, 1.0] that penalizes the number of clusters
        plot_out : boolean, optional (default=True)
            Determines if scaled inertia plot should be output
        random_state : int, optional (default=18)
            Seed for k-means clustering model
        n_jobs : int
            Number of cores to parallelize k-choosing and tissue domain assignment across.
            Default all available cores.

        Returns
        -------
        Does not return anything. `self.tissue_ID` is added, containing image with
        final tissue region IDs. `self.kmeans` contains trained `sklearn` clustering
        model. Parameters are also captured as attributes for posterity.
        &#34;&#34;&#34;
        # save the hyperparams as object attributes
        use_path = self.use_paths
        # find optimal k with parent class
        if k is None:
            print(&#34;Determining optimal cluster number k via scaled inertia&#34;)
            self.find_optimal_k(
                alpha=alpha,
                plot_out=plot_out,
                random_state=random_state,
                n_jobs=n_jobs,
            )
        # call k-means model from parent class
        self.find_tissue_regions(k=k, random_state=random_state)
        # loop through image objects and create tissue label images
        print(&#34;Creating tissue_ID images for image objects...&#34;)
        self.tissue_IDs = Parallel(n_jobs=n_jobs, verbose=10)(
            delayed(add_tissue_ID_single_sample_mxif)(
                image, use_path, self.model_features, self.kmeans, self.scaler
            )
            for image in self.image_df[&#34;Img&#34;]
        )

    def plot_percentage_variance_explained(
        self, fig_size=(5, 5), R_square=False, save_to=None
    ):
        &#34;&#34;&#34;
        plot percentage variance_explained or not explained by clustering

        Parameters
        ----------
        fig_size : Tuple
            size for the bar plot
        R_square : Boolean
            Decides if R_square is plotted or S_square
        save_to : str or None
            Path to image file to save results. If `None`, show figure.

        Returns
        -------
        Matplotlib object
        &#34;&#34;&#34;
        scaler = self.scaler
        centroids = self.kmeans.cluster_centers_
        features = self.model_features
        use_path = self.use_paths
        S_squre_for_each_image = []
        R_squre_for_each_image = []
        for image, tissue_ID in zip(self.image_df[&#34;Img&#34;], self.tissue_IDs):
            S_square = estimate_percentage_variance_mxif(
                image, use_path, scaler, centroids, features, tissue_ID
            )
            S_squre_for_each_image.append(S_square)
            R_squre_for_each_image.append(100 - S_square)

        if R_square == True:
            fig = plt.figure(figsize=fig_size)
            fig = plt.figure(figsize=(5, 5))
            plt.scatter(
                range(len(R_squre_for_each_image)),
                R_squre_for_each_image,
                color=&#34;black&#34;,
            )
            plt.xlabel(&#34;images&#34;)
            plt.ylabel(&#34;percentage variance explained by Kmeans&#34;)
            plt.ylim((0, 100))
            plt.axhline(
                y=np.mean(R_squre_for_each_image),
                linestyle=&#34;dashed&#34;,
                linewidth=1,
                color=&#34;black&#34;,
            )

        else:
            fig = plt.figure(figsize=fig_size)
            plt.scatter(
                range(len(S_squre_for_each_image)),
                S_squre_for_each_image,
                color=&#34;black&#34;,
            )
            plt.xlabel(&#34;images&#34;)
            plt.ylabel(&#34;percentage variance explained by Kmeans&#34;)
            plt.ylim((0, 100))
            plt.axhline(
                y=np.mean(S_squre_for_each_image),
                linestyle=&#34;dashed&#34;,
                linewidth=1,
                color=&#34;black&#34;,
            )

        fig.tight_layout()
        if save_to:
            plt.savefig(fname=save_to, transparent=True, bbox_inches=&#34;tight&#34;, dpi=300)
        return fig

    def confidence_score_images(self):
        &#34;&#34;&#34;
        estimate confidence score for each image

        Parameters
        ----------

        Returns
        -------
        self.confidence_IDs and self.confidence_score_df is added containing
        confidence score for each tissue domain assignment and mean confidence score for
        each tissue domain within each image
        &#34;&#34;&#34;
        scaler = self.scaler
        centroids = self.kmeans.cluster_centers_
        features = self.model_features
        tissue_IDs = self.tissue_IDs
        use_path = self.use_paths
        # confidence score estimation for each image
        confidence_IDs = []
        confidence_score_df = pd.DataFrame()
        for i, image in enumerate(self.image_df[&#34;Img&#34;]):
            cID, scores_dict = estimate_confidence_score_mxif(
                image, use_path, scaler, centroids, features, tissue_IDs[i]
            )
            confidence_IDs.append(cID)
            df = pd.DataFrame(scores_dict.values(), columns=[i])
            confidence_score_df = pd.concat(
                [confidence_score_df, df.T], ignore_index=True
            )
        # adding confidence_IDs and confidence_score_df to tissue labeller object
        self.confidence_IDs = confidence_IDs
        self.confidence_score_df = confidence_score_df

    def plot_mse_mxif(
        self,
        figsize=(5, 5),
        ncols=None,
        labels=None,
        legend_cols=2,
        titles=None,
        loc=&#34;lower right&#34;,
        bbox_coordinates=(0, 0, 1.5, 1.5),
        save_to=None,
    ):
        &#34;&#34;&#34;
        estimate mean square error within each tissue domain

        Parameters
        ----------
        fig_size : Tuple
            size for the bar plot
        ncols : int, optional (default=`None`)
            Number of columns for gridspec. If `None`, uses number of tissue domains k.
        labels : list of str, optional (default=`None`)
            Labels corresponding to each image in legend. If `None`, numeric index is
            used for each imaage
        legend_cols : int, optional (default = `2`)
            n_cols for legend
        titles : list of str, optional (default=`None`)
            Titles of plots corresponding to each MILWRM domain. If `None`, titles
            will be numbers 0 through k.
        loc : str, optional (default = &#39;lower right&#39;)
            str for legend position
        bbox_coordinates : Tuple, optional (default = (0,0,1.5,1.5))
            coordinates for the legend box
        save_to : str, optional (default=`None`)
            Path to image file to save plot

        Returns
        -------
        Matplotlib object
        &#34;&#34;&#34;
        assert (
            self.kmeans is not None
        ), &#34;No cluster results found. Run \
        label_tissue_regions() first.&#34;
        images = self.image_df[&#34;Img&#34;]
        use_path = self.use_paths
        scaler = self.scaler
        centroids = self.kmeans.cluster_centers_
        features = self.model_features
        k = self.k
        features = self.model_features
        tissue_IDs = self.tissue_IDs
        mse_id = estimate_mse_mxif(
            images, use_path, tissue_IDs, scaler, centroids, features, k
        )
        if labels is None:
            labels = range(len(images))
        if titles is None:
            titles = [&#34;tissue_ID &#34; + str(x) for x in range(self.k)]
        n_panels = len(mse_id.keys())
        if ncols is None:
            ncols = len(titles)
        if n_panels &lt;= ncols:
            n_rows, n_cols = 1, n_panels
        else:
            n_rows, n_cols = ceil(n_panels / ncols), ncols
        colors = plt.cm.tab20(np.linspace(0, 1, len(images)))
        fig = plt.figure(figsize=(n_cols * figsize[0], n_rows * figsize[1]))
        left, bottom = 0.1 / n_cols, 0.1 / n_rows
        gs = gridspec.GridSpec(
            nrows=n_rows,
            ncols=n_cols,
            left=left,
            bottom=bottom,
            right=1 - (n_cols - 1) * left - 0.01 / n_cols,
            top=1 - (n_rows - 1) * bottom - 0.1 / n_rows,
        )
        for i in mse_id.keys():
            plt.subplot(gs[i])
            df = pd.DataFrame.from_dict(mse_id[i])
            plt.boxplot(df, positions=range(len(features)), showfliers=False)
            plt.xticks(
                ticks=range(len(features)),
                labels=self.model_features,
                rotation=60,
                fontsize=8,
            )
            for col in df:
                for k in range(len(images)):
                    dots = plt.scatter(
                        col,
                        df[col][k],
                        s=k + 1,
                        color=colors[k],
                        label=labels[k] if col == 0 else &#34;&#34;,
                    )
                    offsets = dots.get_offsets()
                    jittered_offsets = offsets
                    # only jitter in the x-direction
                    jittered_offsets[:, 0] += np.random.uniform(
                        -0.3, 0.3, offsets.shape[0]
                    )
                    dots.set_offsets(jittered_offsets)
            plt.xlabel(&#34;marker&#34;)
            plt.ylabel(&#34;mean square error&#34;)
            plt.title(titles[i])
        plt.legend(loc=loc, bbox_to_anchor=bbox_coordinates, ncol=legend_cols)
        gs.tight_layout(fig)
        if save_to:
            plt.savefig(fname=save_to, transparent=True, dpi=300)
        return fig

    def plot_tissue_ID_proportions_mxif(
        self,
        tID_labels=None,
        slide_labels=None,
        figsize=(5, 5),
        cmap=&#34;tab20&#34;,
        save_to=None,
    ):
        &#34;&#34;&#34;
        Plot proportion of each tissue domain within each slide

        Parameters
        ----------
        tID_labels : list of str, optional (default=`None`)
            List of labels corresponding to MILWRM tissue domains for plotting legend
        slide_labels : list of str, optional (default=`None`)
            List of labels for each slide batch for labeling x-axis
        figsize : tuple of float, optional (default=(5,5))
            Size of matplotlib figure
        cmap : str, optional (default = `&#34;tab20&#34;`)
        save_to : str, optional (default=`None`)
            Path to image file to save plot

        Returns
        -------
        `gridspec.GridSpec` if `save_to` is `None`, else saves plot to file
        &#34;&#34;&#34;
        df_count = pd.DataFrame()
        for i in range(len(self.tissue_IDs)):
            unique, counts = np.unique(self.tissue_IDs[i], return_counts=True)
            dict_ = dict(zip(unique, counts))
            n_counts = []
            for k in range(self.k):
                if k not in dict_.keys():
                    n_counts.append(0)
                else:
                    n_counts.append(dict_[k])
            df = pd.DataFrame(n_counts, columns=[i])
            df_count = pd.concat([df_count, df], axis=1)
        df_count = df_count / df_count.sum()
        if tID_labels:
            assert (
                len(tID_labels) == df_count.shape[1]
            ), &#34;Length of given tissue domain labels does not match number of tissue domains!&#34;
            df_count.columns = tID_labels
        if slide_labels:
            assert (
                len(slide_labels) == df_count.shape[0]
            ), &#34;Length of given slide labels does not match number of slides!&#34;
            df_count.index = slide_labels
        self.tissue_ID_proportion = df_count
        ax = df_count.T.plot.bar(stacked=True, cmap=cmap, figsize=figsize)
        ax.legend(loc=&#34;best&#34;, bbox_to_anchor=(1, 1))
        ax.set_xlabel(&#34;images&#34;)
        ax.set_ylabel(&#34;tissue domain proportion&#34;)
        ax.set_ylim((0, 1))
        plt.tight_layout()
        if save_to is not None:
            ax.figure.savefig(save_to)
        else:
            return ax

    def make_umap(self, frac=None, cmap=&#34;tab20&#34;, save_to=None, alpha=0.8, dot_size_batch = 0.1):
        &#34;&#34;&#34;
        plot umap for the cluster data

        Parameters
        ----------
        frac : None or float
            if None entire cluster data is used for the computation of umap
            else that percentage of cluster data is used.
        cmap : str
            str for cmap used for plotting. Default `&#34;tab20&#34;`.
        save_to : str or None
            Path to image file to save results. if `None`, show figure.
        alpha : float
            opaqueness of umap scatter plot (default=`0.8`)
        dot_size_batch = float
            scatter plot dot size (default=`0.1`)

        Returns
        -------
        Matplotlib object
        &#34;&#34;&#34;
        cluster_data = self.cluster_data
        centroids = self.kmeans.cluster_centers_
        batch_labels = self.merged_batch_labels
        kmeans_labels = self.kmeans.labels_
        k = self.k
        # perform umap on the cluster data
        umap_centroid_data, standard_embedding_1 = perform_umap(
            cluster_data=cluster_data,
            centroids=centroids,
            batch_labels=batch_labels,
            kmeans_labels=kmeans_labels,
            frac=frac,
        )
        # defining a size of datapoints for scatter plot and tick labels
        size = [0.01] * len(umap_centroid_data.index)
        size[-k:] = [10] * k
        ticks = np.unique(np.array(umap_centroid_data[&#34;Kmeans_labels&#34;]))
        tick_label = list(np.unique(np.array(umap_centroid_data[&#34;Kmeans_labels&#34;])))
        tick_label[-1] = &#34;centroids&#34;
        # plotting a fig with two subplots
        fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(20, 10))
        # defining color_map
        disc_cmap_1 = plt.cm.get_cmap(
            cmap, len(np.unique(np.array(umap_centroid_data.index)))
        )
        disc_cmap_2 = plt.cm.get_cmap(
            cmap, len(np.unique(np.array(umap_centroid_data[&#34;Kmeans_labels&#34;])))
        )
        plot_1 = ax1.scatter(
            standard_embedding_1[:, 0],
            standard_embedding_1[:, 1],
            s=dot_size_batch,
            c=umap_centroid_data.index,
            cmap=disc_cmap_1,
            alpha=alpha,
        )
        ax1.set_title(&#34;UMAP with batch labels&#34;, fontsize = 24)
        ax1.set_xlabel(&#34;UMAP 2&#34;)
        ax1.set_ylabel(&#34;UMAP 1&#34;)
        ax1.set_xticks([])
        ax1.set_yticks([])
        cbar_1 = plt.colorbar(plot_1, ax=ax1)

        plot_2 = ax2.scatter(
            standard_embedding_1[:, 0],
            standard_embedding_1[:, 1],
            s=size,
            c=umap_centroid_data[&#34;Kmeans_labels&#34;],
            cmap=disc_cmap_2,
            alpha=alpha,
        )
        ax2.set_title(&#34;UMAP with tissue domains&#34;, fontsize = 24)
        ax2.set_xticks([])
        ax2.set_yticks([])
        ax2.set_xlabel(&#34;UMAP 2&#34;)
        ax2.set_ylabel(&#34;UMAP 1&#34;)
        cbar_2 = plt.colorbar(plot_2, ax=ax2, ticks=ticks)
        cbar_2.ax.set_yticklabels(tick_label)
        fig.tight_layout()
        if save_to:
            plt.savefig(fname=save_to, transparent=True, bbox_inches=&#34;tight&#34;, dpi=300)
        return fig

    def show_marker_overlay(
        self,
        image_index,
        channels=None,
        cmap=&#34;Set1&#34;,
        mask_out=True,
        ncols=4,
        save_to=None,
        **kwargs,
    ):
        &#34;&#34;&#34;
        Plot tissue_ID with individual markers as alpha values to distinguish
        expression in identified tissue domains

        Parameters
        ----------
        image_index : int
            Index of image from `self.images` to plot overlays for (e.g. 0 for first
            image)
        channels : tuple of int or None, optional (default=`None`)
            List of channels by index or name to show
        cmap : str, optional (default=&#34;plasma&#34;)
            Matplotlib colormap to use for plotting tissue domains
        mask_out : bool, optional (default=`True`)
            Mask out non-tissue pixels prior to showing
        ncols : int
            Number of columns for gridspec if plotting individual channels.
        save_to : str or None
            Path to image file to save results. If `None`, show figure.
        **kwargs
            Arguments to pass to `plt.imshow()` function.

        Returns
        -------
        Matplotlib object (if plotting one feature or RGB) or gridspec object (for
        multiple features). Saves plot to file if `save_to` is not `None`.
        &#34;&#34;&#34;
        # if image has multiple channels, plot them in gridspec
        if isinstance(channels, int):  # force channels into list if single integer
            channels = [channels]
        if isinstance(channels, str):  # force channels into int if single string
            channels = [self[image_index].ch.index(channels)]
        if checktype(channels):  # force channels into list of int if list of strings
            channels = [self[image_index].ch.index(x) for x in channels]
        if channels is None:  # if no channels are given, use all of them
            channels = [x for x in range(self[image_index].n_ch)]
        assert (
            len(channels) &lt;= self[image_index].n_ch
        ), &#34;Too many channels given: image has {}, expected {}&#34;.format(
            self[image_index].n_ch, len(channels)
        )
        # creating a copy of the image
        image_cp = self[image_index].copy()
        # re-scaling to set pixel value range between 0 to 1
        image_cp.scale()
        # defining cmap for discrete color bar
        cmap = plt.cm.get_cmap(cmap, self.k)
        # calculate gridspec dimensions
        if len(channels) + 1 &lt;= ncols:
            n_rows, n_cols = 1, len(channels) + 1
        else:
            n_rows, n_cols = ceil(len(channels) + 1 / ncols), ncols
        fig = plt.figure(figsize=(ncols * n_cols, ncols * n_rows))
        # arrange axes as subplots
        gs = gridspec.GridSpec(n_rows, n_cols, figure=fig)
        # plot tissue_ID first with colorbar
        ax = plt.subplot(gs[0])
        im = ax.imshow(self.tissue_IDs[image_index], cmap=cmap, **kwargs)
        ax.set_title(
            label=&#34;tissue_ID&#34;,
            loc=&#34;left&#34;,
            fontweight=&#34;bold&#34;,
            fontsize=16,
        )
        ax.tick_params(labelbottom=False, labelleft=False)
        sns.despine(bottom=True, left=True)
        # colorbar scale for tissue_IDs
        _ = plt.colorbar(im, ticks=range(self.k), shrink=0.7)
        # add plots to axes
        i = 1
        for channel in channels:
            ax = plt.subplot(gs[i])
            # make copy for alpha
            im_tmp = image_cp.img[:, :, channel].copy()
            if self[image_index].mask is not None and mask_out:
                # area outside mask NaN
                self.tissue_IDs[image_index][self[image_index].mask == 0] = np.nan
                im = ax.imshow(
                    self.tissue_IDs[image_index], cmap=cmap, alpha=im_tmp, **kwargs
                )
            else:
                ax.imshow(self.tissue_IDs[image_index], alpha=im_tmp, **kwargs)
            ax.tick_params(labelbottom=False, labelleft=False)
            sns.despine(bottom=True, left=True)
            ax.set_title(
                label=self[image_index].ch[channel],
                loc=&#34;left&#34;,
                fontweight=&#34;bold&#34;,
                fontsize=16,
            )
            i = i + 1
        fig.tight_layout()
        if save_to:
            plt.savefig(fname=save_to, transparent=True, bbox_inches=&#34;tight&#34;, dpi=300)
        return fig</code></pre>
</details>
<h3>Ancestors</h3>
<ul class="hlist">
<li><a title="MILWRM.MILWRM.tissue_labeler" href="MILWRM.html#MILWRM.MILWRM.tissue_labeler">tissue_labeler</a></li>
</ul>
<h3>Methods</h3>
<dl>
<dt id="MILWRM.mxif_labeler.confidence_score_images"><code class="name flex">
<span>def <span class="ident">confidence_score_images</span></span>(<span>self)</span>
</code></dt>
<dd>
<div class="desc"><p>estimate confidence score for each image</p>
<h2 id="parameters">Parameters</h2>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>self.confidence_IDs and self.confidence_score_df is added containing</code></dt>
<dd>&nbsp;</dd>
<dt><code>confidence score for each tissue domain assignment and mean confidence score for</code></dt>
<dd>&nbsp;</dd>
<dt><code>each tissue domain within each image</code></dt>
<dd>&nbsp;</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def confidence_score_images(self):
    &#34;&#34;&#34;
    estimate confidence score for each image

    Parameters
    ----------

    Returns
    -------
    self.confidence_IDs and self.confidence_score_df is added containing
    confidence score for each tissue domain assignment and mean confidence score for
    each tissue domain within each image
    &#34;&#34;&#34;
    scaler = self.scaler
    centroids = self.kmeans.cluster_centers_
    features = self.model_features
    tissue_IDs = self.tissue_IDs
    use_path = self.use_paths
    # confidence score estimation for each image
    confidence_IDs = []
    confidence_score_df = pd.DataFrame()
    for i, image in enumerate(self.image_df[&#34;Img&#34;]):
        cID, scores_dict = estimate_confidence_score_mxif(
            image, use_path, scaler, centroids, features, tissue_IDs[i]
        )
        confidence_IDs.append(cID)
        df = pd.DataFrame(scores_dict.values(), columns=[i])
        confidence_score_df = pd.concat(
            [confidence_score_df, df.T], ignore_index=True
        )
    # adding confidence_IDs and confidence_score_df to tissue labeller object
    self.confidence_IDs = confidence_IDs
    self.confidence_score_df = confidence_score_df</code></pre>
</details>
</dd>
<dt id="MILWRM.mxif_labeler.label_tissue_regions"><code class="name flex">
<span>def <span class="ident">label_tissue_regions</span></span>(<span>self, k=None, alpha=0.05, plot_out=True, random_state=18, n_jobs=-1)</span>
</code></dt>
<dd>
<div class="desc"><p>Perform tissue-level clustering and label pixels in the corresponding
images.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>k</code></strong> :&ensp;<code>int</code>, optional <code>(default=None)</code></dt>
<dd>Number of tissue regions to define</dd>
<dt><strong><code>alpha</code></strong> :&ensp;<code>float</code></dt>
<dd>Manually tuned factor on [0.0, 1.0] that penalizes the number of clusters</dd>
<dt><strong><code>plot_out</code></strong> :&ensp;<code>boolean</code>, optional <code>(default=True)</code></dt>
<dd>Determines if scaled inertia plot should be output</dd>
<dt><strong><code>random_state</code></strong> :&ensp;<code>int</code>, optional <code>(default=18)</code></dt>
<dd>Seed for k-means clustering model</dd>
<dt><strong><code>n_jobs</code></strong> :&ensp;<code>int</code></dt>
<dd>Number of cores to parallelize k-choosing and tissue domain assignment across.
Default all available cores.</dd>
</dl>
<h2 id="returns">Returns</h2>
<p>Does not return anything. <code>self.tissue_ID</code> is added, containing image with
final tissue region IDs. <code>self.kmeans</code> contains trained <code>sklearn</code> clustering
model. Parameters are also captured as attributes for posterity.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def label_tissue_regions(
    self, k=None, alpha=0.05, plot_out=True, random_state=18, n_jobs=-1
):
    &#34;&#34;&#34;
    Perform tissue-level clustering and label pixels in the corresponding
    images.

    Parameters
    ----------
    k : int, optional (default=None)
        Number of tissue regions to define
    alpha: float
        Manually tuned factor on [0.0, 1.0] that penalizes the number of clusters
    plot_out : boolean, optional (default=True)
        Determines if scaled inertia plot should be output
    random_state : int, optional (default=18)
        Seed for k-means clustering model
    n_jobs : int
        Number of cores to parallelize k-choosing and tissue domain assignment across.
        Default all available cores.

    Returns
    -------
    Does not return anything. `self.tissue_ID` is added, containing image with
    final tissue region IDs. `self.kmeans` contains trained `sklearn` clustering
    model. Parameters are also captured as attributes for posterity.
    &#34;&#34;&#34;
    # save the hyperparams as object attributes
    use_path = self.use_paths
    # find optimal k with parent class
    if k is None:
        print(&#34;Determining optimal cluster number k via scaled inertia&#34;)
        self.find_optimal_k(
            alpha=alpha,
            plot_out=plot_out,
            random_state=random_state,
            n_jobs=n_jobs,
        )
    # call k-means model from parent class
    self.find_tissue_regions(k=k, random_state=random_state)
    # loop through image objects and create tissue label images
    print(&#34;Creating tissue_ID images for image objects...&#34;)
    self.tissue_IDs = Parallel(n_jobs=n_jobs, verbose=10)(
        delayed(add_tissue_ID_single_sample_mxif)(
            image, use_path, self.model_features, self.kmeans, self.scaler
        )
        for image in self.image_df[&#34;Img&#34;]
    )</code></pre>
</details>
</dd>
<dt id="MILWRM.mxif_labeler.make_umap"><code class="name flex">
<span>def <span class="ident">make_umap</span></span>(<span>self, frac=None, cmap='tab20', save_to=None, alpha=0.8, dot_size_batch=0.1)</span>
</code></dt>
<dd>
<div class="desc"><p>plot umap for the cluster data</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>frac</code></strong> :&ensp;<code>None</code> or <code>float</code></dt>
<dd>if None entire cluster data is used for the computation of umap
else that percentage of cluster data is used.</dd>
<dt><strong><code>cmap</code></strong> :&ensp;<code>str</code></dt>
<dd>str for cmap used for plotting. Default <code>"tab20"</code>.</dd>
<dt><strong><code>save_to</code></strong> :&ensp;<code>str</code> or <code>None</code></dt>
<dd>Path to image file to save results. if <code>None</code>, show figure.</dd>
<dt><strong><code>alpha</code></strong> :&ensp;<code>float</code></dt>
<dd>opaqueness of umap scatter plot (default=<code>0.8</code>)</dd>
</dl>
<p>dot_size_batch = float
scatter plot dot size (default=<code>0.1</code>)</p>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>Matplotlib object</code></dt>
<dd>&nbsp;</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def make_umap(self, frac=None, cmap=&#34;tab20&#34;, save_to=None, alpha=0.8, dot_size_batch = 0.1):
    &#34;&#34;&#34;
    plot umap for the cluster data

    Parameters
    ----------
    frac : None or float
        if None entire cluster data is used for the computation of umap
        else that percentage of cluster data is used.
    cmap : str
        str for cmap used for plotting. Default `&#34;tab20&#34;`.
    save_to : str or None
        Path to image file to save results. if `None`, show figure.
    alpha : float
        opaqueness of umap scatter plot (default=`0.8`)
    dot_size_batch = float
        scatter plot dot size (default=`0.1`)

    Returns
    -------
    Matplotlib object
    &#34;&#34;&#34;
    cluster_data = self.cluster_data
    centroids = self.kmeans.cluster_centers_
    batch_labels = self.merged_batch_labels
    kmeans_labels = self.kmeans.labels_
    k = self.k
    # perform umap on the cluster data
    umap_centroid_data, standard_embedding_1 = perform_umap(
        cluster_data=cluster_data,
        centroids=centroids,
        batch_labels=batch_labels,
        kmeans_labels=kmeans_labels,
        frac=frac,
    )
    # defining a size of datapoints for scatter plot and tick labels
    size = [0.01] * len(umap_centroid_data.index)
    size[-k:] = [10] * k
    ticks = np.unique(np.array(umap_centroid_data[&#34;Kmeans_labels&#34;]))
    tick_label = list(np.unique(np.array(umap_centroid_data[&#34;Kmeans_labels&#34;])))
    tick_label[-1] = &#34;centroids&#34;
    # plotting a fig with two subplots
    fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(20, 10))
    # defining color_map
    disc_cmap_1 = plt.cm.get_cmap(
        cmap, len(np.unique(np.array(umap_centroid_data.index)))
    )
    disc_cmap_2 = plt.cm.get_cmap(
        cmap, len(np.unique(np.array(umap_centroid_data[&#34;Kmeans_labels&#34;])))
    )
    plot_1 = ax1.scatter(
        standard_embedding_1[:, 0],
        standard_embedding_1[:, 1],
        s=dot_size_batch,
        c=umap_centroid_data.index,
        cmap=disc_cmap_1,
        alpha=alpha,
    )
    ax1.set_title(&#34;UMAP with batch labels&#34;, fontsize = 24)
    ax1.set_xlabel(&#34;UMAP 2&#34;)
    ax1.set_ylabel(&#34;UMAP 1&#34;)
    ax1.set_xticks([])
    ax1.set_yticks([])
    cbar_1 = plt.colorbar(plot_1, ax=ax1)

    plot_2 = ax2.scatter(
        standard_embedding_1[:, 0],
        standard_embedding_1[:, 1],
        s=size,
        c=umap_centroid_data[&#34;Kmeans_labels&#34;],
        cmap=disc_cmap_2,
        alpha=alpha,
    )
    ax2.set_title(&#34;UMAP with tissue domains&#34;, fontsize = 24)
    ax2.set_xticks([])
    ax2.set_yticks([])
    ax2.set_xlabel(&#34;UMAP 2&#34;)
    ax2.set_ylabel(&#34;UMAP 1&#34;)
    cbar_2 = plt.colorbar(plot_2, ax=ax2, ticks=ticks)
    cbar_2.ax.set_yticklabels(tick_label)
    fig.tight_layout()
    if save_to:
        plt.savefig(fname=save_to, transparent=True, bbox_inches=&#34;tight&#34;, dpi=300)
    return fig</code></pre>
</details>
</dd>
<dt id="MILWRM.mxif_labeler.plot_mse_mxif"><code class="name flex">
<span>def <span class="ident">plot_mse_mxif</span></span>(<span>self, figsize=(5, 5), ncols=None, labels=None, legend_cols=2, titles=None, loc='lower right', bbox_coordinates=(0, 0, 1.5, 1.5), save_to=None)</span>
</code></dt>
<dd>
<div class="desc"><p>estimate mean square error within each tissue domain</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>fig_size</code></strong> :&ensp;<code>Tuple</code></dt>
<dd>size for the bar plot</dd>
<dt><strong><code>ncols</code></strong> :&ensp;<code>int</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>Number of columns for gridspec. If <code>None</code>, uses number of tissue domains k.</dd>
<dt><strong><code>labels</code></strong> :&ensp;<code>list</code> of <code>str</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>Labels corresponding to each image in legend. If <code>None</code>, numeric index is
used for each imaage</dd>
<dt><strong><code>legend_cols</code></strong> :&ensp;<code>int</code>, optional <code>(default =</code>2<code>)</code></dt>
<dd>n_cols for legend</dd>
<dt><strong><code>titles</code></strong> :&ensp;<code>list</code> of <code>str</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>Titles of plots corresponding to each MILWRM domain. If <code>None</code>, titles
will be numbers 0 through k.</dd>
<dt><strong><code>loc</code></strong> :&ensp;<code>str</code>, optional <code>(default = 'lower right')</code></dt>
<dd>str for legend position</dd>
<dt><strong><code>bbox_coordinates</code></strong> :&ensp;<code>Tuple</code>, optional <code>(default = (0,0,1.5,1.5))</code></dt>
<dd>coordinates for the legend box</dd>
<dt><strong><code>save_to</code></strong> :&ensp;<code>str</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>Path to image file to save plot</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>Matplotlib object</code></dt>
<dd>&nbsp;</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def plot_mse_mxif(
    self,
    figsize=(5, 5),
    ncols=None,
    labels=None,
    legend_cols=2,
    titles=None,
    loc=&#34;lower right&#34;,
    bbox_coordinates=(0, 0, 1.5, 1.5),
    save_to=None,
):
    &#34;&#34;&#34;
    estimate mean square error within each tissue domain

    Parameters
    ----------
    fig_size : Tuple
        size for the bar plot
    ncols : int, optional (default=`None`)
        Number of columns for gridspec. If `None`, uses number of tissue domains k.
    labels : list of str, optional (default=`None`)
        Labels corresponding to each image in legend. If `None`, numeric index is
        used for each imaage
    legend_cols : int, optional (default = `2`)
        n_cols for legend
    titles : list of str, optional (default=`None`)
        Titles of plots corresponding to each MILWRM domain. If `None`, titles
        will be numbers 0 through k.
    loc : str, optional (default = &#39;lower right&#39;)
        str for legend position
    bbox_coordinates : Tuple, optional (default = (0,0,1.5,1.5))
        coordinates for the legend box
    save_to : str, optional (default=`None`)
        Path to image file to save plot

    Returns
    -------
    Matplotlib object
    &#34;&#34;&#34;
    assert (
        self.kmeans is not None
    ), &#34;No cluster results found. Run \
    label_tissue_regions() first.&#34;
    images = self.image_df[&#34;Img&#34;]
    use_path = self.use_paths
    scaler = self.scaler
    centroids = self.kmeans.cluster_centers_
    features = self.model_features
    k = self.k
    features = self.model_features
    tissue_IDs = self.tissue_IDs
    mse_id = estimate_mse_mxif(
        images, use_path, tissue_IDs, scaler, centroids, features, k
    )
    if labels is None:
        labels = range(len(images))
    if titles is None:
        titles = [&#34;tissue_ID &#34; + str(x) for x in range(self.k)]
    n_panels = len(mse_id.keys())
    if ncols is None:
        ncols = len(titles)
    if n_panels &lt;= ncols:
        n_rows, n_cols = 1, n_panels
    else:
        n_rows, n_cols = ceil(n_panels / ncols), ncols
    colors = plt.cm.tab20(np.linspace(0, 1, len(images)))
    fig = plt.figure(figsize=(n_cols * figsize[0], n_rows * figsize[1]))
    left, bottom = 0.1 / n_cols, 0.1 / n_rows
    gs = gridspec.GridSpec(
        nrows=n_rows,
        ncols=n_cols,
        left=left,
        bottom=bottom,
        right=1 - (n_cols - 1) * left - 0.01 / n_cols,
        top=1 - (n_rows - 1) * bottom - 0.1 / n_rows,
    )
    for i in mse_id.keys():
        plt.subplot(gs[i])
        df = pd.DataFrame.from_dict(mse_id[i])
        plt.boxplot(df, positions=range(len(features)), showfliers=False)
        plt.xticks(
            ticks=range(len(features)),
            labels=self.model_features,
            rotation=60,
            fontsize=8,
        )
        for col in df:
            for k in range(len(images)):
                dots = plt.scatter(
                    col,
                    df[col][k],
                    s=k + 1,
                    color=colors[k],
                    label=labels[k] if col == 0 else &#34;&#34;,
                )
                offsets = dots.get_offsets()
                jittered_offsets = offsets
                # only jitter in the x-direction
                jittered_offsets[:, 0] += np.random.uniform(
                    -0.3, 0.3, offsets.shape[0]
                )
                dots.set_offsets(jittered_offsets)
        plt.xlabel(&#34;marker&#34;)
        plt.ylabel(&#34;mean square error&#34;)
        plt.title(titles[i])
    plt.legend(loc=loc, bbox_to_anchor=bbox_coordinates, ncol=legend_cols)
    gs.tight_layout(fig)
    if save_to:
        plt.savefig(fname=save_to, transparent=True, dpi=300)
    return fig</code></pre>
</details>
</dd>
<dt id="MILWRM.mxif_labeler.plot_percentage_variance_explained"><code class="name flex">
<span>def <span class="ident">plot_percentage_variance_explained</span></span>(<span>self, fig_size=(5, 5), R_square=False, save_to=None)</span>
</code></dt>
<dd>
<div class="desc"><p>plot percentage variance_explained or not explained by clustering</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>fig_size</code></strong> :&ensp;<code>Tuple</code></dt>
<dd>size for the bar plot</dd>
<dt><strong><code>R_square</code></strong> :&ensp;<code>Boolean</code></dt>
<dd>Decides if R_square is plotted or S_square</dd>
<dt><strong><code>save_to</code></strong> :&ensp;<code>str</code> or <code>None</code></dt>
<dd>Path to image file to save results. If <code>None</code>, show figure.</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>Matplotlib object</code></dt>
<dd>&nbsp;</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def plot_percentage_variance_explained(
    self, fig_size=(5, 5), R_square=False, save_to=None
):
    &#34;&#34;&#34;
    plot percentage variance_explained or not explained by clustering

    Parameters
    ----------
    fig_size : Tuple
        size for the bar plot
    R_square : Boolean
        Decides if R_square is plotted or S_square
    save_to : str or None
        Path to image file to save results. If `None`, show figure.

    Returns
    -------
    Matplotlib object
    &#34;&#34;&#34;
    scaler = self.scaler
    centroids = self.kmeans.cluster_centers_
    features = self.model_features
    use_path = self.use_paths
    S_squre_for_each_image = []
    R_squre_for_each_image = []
    for image, tissue_ID in zip(self.image_df[&#34;Img&#34;], self.tissue_IDs):
        S_square = estimate_percentage_variance_mxif(
            image, use_path, scaler, centroids, features, tissue_ID
        )
        S_squre_for_each_image.append(S_square)
        R_squre_for_each_image.append(100 - S_square)

    if R_square == True:
        fig = plt.figure(figsize=fig_size)
        fig = plt.figure(figsize=(5, 5))
        plt.scatter(
            range(len(R_squre_for_each_image)),
            R_squre_for_each_image,
            color=&#34;black&#34;,
        )
        plt.xlabel(&#34;images&#34;)
        plt.ylabel(&#34;percentage variance explained by Kmeans&#34;)
        plt.ylim((0, 100))
        plt.axhline(
            y=np.mean(R_squre_for_each_image),
            linestyle=&#34;dashed&#34;,
            linewidth=1,
            color=&#34;black&#34;,
        )

    else:
        fig = plt.figure(figsize=fig_size)
        plt.scatter(
            range(len(S_squre_for_each_image)),
            S_squre_for_each_image,
            color=&#34;black&#34;,
        )
        plt.xlabel(&#34;images&#34;)
        plt.ylabel(&#34;percentage variance explained by Kmeans&#34;)
        plt.ylim((0, 100))
        plt.axhline(
            y=np.mean(S_squre_for_each_image),
            linestyle=&#34;dashed&#34;,
            linewidth=1,
            color=&#34;black&#34;,
        )

    fig.tight_layout()
    if save_to:
        plt.savefig(fname=save_to, transparent=True, bbox_inches=&#34;tight&#34;, dpi=300)
    return fig</code></pre>
</details>
</dd>
<dt id="MILWRM.mxif_labeler.plot_tissue_ID_proportions_mxif"><code class="name flex">
<span>def <span class="ident">plot_tissue_ID_proportions_mxif</span></span>(<span>self, tID_labels=None, slide_labels=None, figsize=(5, 5), cmap='tab20', save_to=None)</span>
</code></dt>
<dd>
<div class="desc"><p>Plot proportion of each tissue domain within each slide</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>tID_labels</code></strong> :&ensp;<code>list</code> of <code>str</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>List of labels corresponding to MILWRM tissue domains for plotting legend</dd>
<dt><strong><code>slide_labels</code></strong> :&ensp;<code>list</code> of <code>str</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>List of labels for each slide batch for labeling x-axis</dd>
<dt><strong><code>figsize</code></strong> :&ensp;<code>tuple</code> of <code>float</code>, optional <code>(default=(5,5))</code></dt>
<dd>Size of matplotlib figure</dd>
<dt><strong><code>cmap</code></strong> :&ensp;<code>str</code>, optional <code>(default =</code>"tab20"<code>)</code></dt>
<dd>&nbsp;</dd>
<dt><strong><code>save_to</code></strong> :&ensp;<code>str</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>Path to image file to save plot</dd>
</dl>
<h2 id="returns">Returns</h2>
<p><code>gridspec.GridSpec</code> if <code>save_to</code> is <code>None</code>, else saves plot to file</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def plot_tissue_ID_proportions_mxif(
    self,
    tID_labels=None,
    slide_labels=None,
    figsize=(5, 5),
    cmap=&#34;tab20&#34;,
    save_to=None,
):
    &#34;&#34;&#34;
    Plot proportion of each tissue domain within each slide

    Parameters
    ----------
    tID_labels : list of str, optional (default=`None`)
        List of labels corresponding to MILWRM tissue domains for plotting legend
    slide_labels : list of str, optional (default=`None`)
        List of labels for each slide batch for labeling x-axis
    figsize : tuple of float, optional (default=(5,5))
        Size of matplotlib figure
    cmap : str, optional (default = `&#34;tab20&#34;`)
    save_to : str, optional (default=`None`)
        Path to image file to save plot

    Returns
    -------
    `gridspec.GridSpec` if `save_to` is `None`, else saves plot to file
    &#34;&#34;&#34;
    df_count = pd.DataFrame()
    for i in range(len(self.tissue_IDs)):
        unique, counts = np.unique(self.tissue_IDs[i], return_counts=True)
        dict_ = dict(zip(unique, counts))
        n_counts = []
        for k in range(self.k):
            if k not in dict_.keys():
                n_counts.append(0)
            else:
                n_counts.append(dict_[k])
        df = pd.DataFrame(n_counts, columns=[i])
        df_count = pd.concat([df_count, df], axis=1)
    df_count = df_count / df_count.sum()
    if tID_labels:
        assert (
            len(tID_labels) == df_count.shape[1]
        ), &#34;Length of given tissue domain labels does not match number of tissue domains!&#34;
        df_count.columns = tID_labels
    if slide_labels:
        assert (
            len(slide_labels) == df_count.shape[0]
        ), &#34;Length of given slide labels does not match number of slides!&#34;
        df_count.index = slide_labels
    self.tissue_ID_proportion = df_count
    ax = df_count.T.plot.bar(stacked=True, cmap=cmap, figsize=figsize)
    ax.legend(loc=&#34;best&#34;, bbox_to_anchor=(1, 1))
    ax.set_xlabel(&#34;images&#34;)
    ax.set_ylabel(&#34;tissue domain proportion&#34;)
    ax.set_ylim((0, 1))
    plt.tight_layout()
    if save_to is not None:
        ax.figure.savefig(save_to)
    else:
        return ax</code></pre>
</details>
</dd>
<dt id="MILWRM.mxif_labeler.prep_cluster_data"><code class="name flex">
<span>def <span class="ident">prep_cluster_data</span></span>(<span>self, features, filter_name='gaussian', sigma=2, fract=0.2, path_save=None)</span>
</code></dt>
<dd>
<div class="desc"><p>Prepare master array for tissue level clustering</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>features</code></strong> :&ensp;<code>list</code> of <code>int</code> or <code>str</code></dt>
<dd>Indices or names of MxIF channels to use for tissue labeling</dd>
<dt><strong><code>filter_name</code></strong> :&ensp;<code>str</code></dt>
<dd>Name of the filter to use - gaussian, median or bilateral</dd>
<dt><strong><code>sigma</code></strong> :&ensp;<code>float</code>, optional <code>(default=2)</code></dt>
<dd>Standard deviation of Gaussian kernel for blurring</dd>
<dt><strong><code>fract</code></strong> :&ensp;<code>float</code>, optional <code>(default=0.2)</code></dt>
<dd>Fraction of cluster data from each image to randomly select for model
building</dd>
<dt><strong><code>path_save</code></strong> :&ensp;<code>str (default = None)</code></dt>
<dd>Path to save final preprocessed files, if self.use_path is True
default path_save will raise Exception</dd>
</dl>
<h2 id="returns">Returns</h2>
<p>Does not return anything. <code>self.images</code> are normalized, blurred and scaled
according to user parameters. <code>self.cluster_data</code> becomes master <code>np.array</code>
for cluster training. Parameters are also captured as attributes for posterity.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def prep_cluster_data(
    self, features, filter_name=&#34;gaussian&#34;, sigma=2, fract=0.2, path_save=None
):
    &#34;&#34;&#34;
    Prepare master array for tissue level clustering

    Parameters
    ----------
    features : list of int or str
        Indices or names of MxIF channels to use for tissue labeling
    filter_name : str
        Name of the filter to use - gaussian, median or bilateral
    sigma : float, optional (default=2)
        Standard deviation of Gaussian kernel for blurring
    fract : float, optional (default=0.2)
        Fraction of cluster data from each image to randomly select for model
        building
    path_save : str (default = None)
        Path to save final preprocessed files, if self.use_path is True
        default path_save will raise Exception

    Returns
    -------
    Does not return anything. `self.images` are normalized, blurred and scaled
    according to user parameters. `self.cluster_data` becomes master `np.array`
    for cluster training. Parameters are also captured as attributes for posterity.

    &#34;&#34;&#34;
    if self.cluster_data is not None:
        print(&#34;WARNING: overwriting existing cluster data&#34;)
        self.cluster_data = None
    # save the hyperparams as object attributes
    self.model_features = features
    use_path = self.use_paths
    # calculate the batch wise means
    mean_for_each_batch = {}
    for batch in self.image_df[&#34;batch_names&#34;].unique():
        list_mean_estimators = list(
            self.image_df[self.image_df[&#34;batch_names&#34;] == batch][&#34;mean estimators&#34;]
        )
        mean_estimator_batch = sum(map(np.array, list_mean_estimators))
        pixels = sum(self.image_df[self.image_df[&#34;batch_names&#34;] == batch][&#34;pixels&#34;])
        mean_for_each_batch[batch] = mean_estimator_batch / pixels
    # log_normalize, apply blurring filter, minmax scale each channel and subsample
    subsampled_data = []
    path_to_blurred_npz = []
    for image, batch in zip(self.image_df[&#34;Img&#34;], self.image_df[&#34;batch_names&#34;]):
        tmp = prep_data_single_sample_mxif(
            image,
            use_path=use_path,
            mean=mean_for_each_batch[batch],
            filter_name=filter_name,
            sigma=sigma,
            features=self.model_features,
            fract=fract,
            path_save=path_save,
        )
        if self.use_paths == True:
            subsampled_data.append(tmp[0])
            path_to_blurred_npz.append(tmp[1])
        else:
            subsampled_data.append(tmp)
    batch_labels = [
        [x] * len(subsampled_data[x]) for x in range(len(subsampled_data))
    ]  # batch labels for umap
    self.merged_batch_labels = list(itertools.chain(*batch_labels))
    if self.use_paths == True:
        self.image_df[&#34;Img&#34;] = path_to_blurred_npz
    cluster_data = np.row_stack(subsampled_data)
    # perform z-score normalization on cluster_Data
    scaler = StandardScaler()
    self.scaler = scaler.fit(cluster_data)
    scaled_data = scaler.transform(cluster_data)
    self.cluster_data = scaled_data</code></pre>
</details>
</dd>
<dt id="MILWRM.mxif_labeler.show_marker_overlay"><code class="name flex">
<span>def <span class="ident">show_marker_overlay</span></span>(<span>self, image_index, channels=None, cmap='Set1', mask_out=True, ncols=4, save_to=None, **kwargs)</span>
</code></dt>
<dd>
<div class="desc"><p>Plot tissue_ID with individual markers as alpha values to distinguish
expression in identified tissue domains</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>image_index</code></strong> :&ensp;<code>int</code></dt>
<dd>Index of image from <code>self.images</code> to plot overlays for (e.g. 0 for first
image)</dd>
<dt><strong><code>channels</code></strong> :&ensp;<code>tuple</code> of <code>int</code> or <code>None</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>List of channels by index or name to show</dd>
<dt><strong><code>cmap</code></strong> :&ensp;<code>str</code>, optional <code>(default="plasma")</code></dt>
<dd>Matplotlib colormap to use for plotting tissue domains</dd>
<dt><strong><code>mask_out</code></strong> :&ensp;<code>bool</code>, optional <code>(default=</code>True<code>)</code></dt>
<dd>Mask out non-tissue pixels prior to showing</dd>
<dt><strong><code>ncols</code></strong> :&ensp;<code>int</code></dt>
<dd>Number of columns for gridspec if plotting individual channels.</dd>
<dt><strong><code>save_to</code></strong> :&ensp;<code>str</code> or <code>None</code></dt>
<dd>Path to image file to save results. If <code>None</code>, show figure.</dd>
<dt><strong><code>**kwargs</code></strong></dt>
<dd>Arguments to pass to <code>plt.imshow()</code> function.</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>Matplotlib object (if plotting one feature</code> or <code>RGB)</code> or <code>gridspec object (for</code></dt>
<dd>&nbsp;</dd>
</dl>
<p>multiple features). Saves plot to file if <code>save_to</code> is not <code>None</code>.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def show_marker_overlay(
    self,
    image_index,
    channels=None,
    cmap=&#34;Set1&#34;,
    mask_out=True,
    ncols=4,
    save_to=None,
    **kwargs,
):
    &#34;&#34;&#34;
    Plot tissue_ID with individual markers as alpha values to distinguish
    expression in identified tissue domains

    Parameters
    ----------
    image_index : int
        Index of image from `self.images` to plot overlays for (e.g. 0 for first
        image)
    channels : tuple of int or None, optional (default=`None`)
        List of channels by index or name to show
    cmap : str, optional (default=&#34;plasma&#34;)
        Matplotlib colormap to use for plotting tissue domains
    mask_out : bool, optional (default=`True`)
        Mask out non-tissue pixels prior to showing
    ncols : int
        Number of columns for gridspec if plotting individual channels.
    save_to : str or None
        Path to image file to save results. If `None`, show figure.
    **kwargs
        Arguments to pass to `plt.imshow()` function.

    Returns
    -------
    Matplotlib object (if plotting one feature or RGB) or gridspec object (for
    multiple features). Saves plot to file if `save_to` is not `None`.
    &#34;&#34;&#34;
    # if image has multiple channels, plot them in gridspec
    if isinstance(channels, int):  # force channels into list if single integer
        channels = [channels]
    if isinstance(channels, str):  # force channels into int if single string
        channels = [self[image_index].ch.index(channels)]
    if checktype(channels):  # force channels into list of int if list of strings
        channels = [self[image_index].ch.index(x) for x in channels]
    if channels is None:  # if no channels are given, use all of them
        channels = [x for x in range(self[image_index].n_ch)]
    assert (
        len(channels) &lt;= self[image_index].n_ch
    ), &#34;Too many channels given: image has {}, expected {}&#34;.format(
        self[image_index].n_ch, len(channels)
    )
    # creating a copy of the image
    image_cp = self[image_index].copy()
    # re-scaling to set pixel value range between 0 to 1
    image_cp.scale()
    # defining cmap for discrete color bar
    cmap = plt.cm.get_cmap(cmap, self.k)
    # calculate gridspec dimensions
    if len(channels) + 1 &lt;= ncols:
        n_rows, n_cols = 1, len(channels) + 1
    else:
        n_rows, n_cols = ceil(len(channels) + 1 / ncols), ncols
    fig = plt.figure(figsize=(ncols * n_cols, ncols * n_rows))
    # arrange axes as subplots
    gs = gridspec.GridSpec(n_rows, n_cols, figure=fig)
    # plot tissue_ID first with colorbar
    ax = plt.subplot(gs[0])
    im = ax.imshow(self.tissue_IDs[image_index], cmap=cmap, **kwargs)
    ax.set_title(
        label=&#34;tissue_ID&#34;,
        loc=&#34;left&#34;,
        fontweight=&#34;bold&#34;,
        fontsize=16,
    )
    ax.tick_params(labelbottom=False, labelleft=False)
    sns.despine(bottom=True, left=True)
    # colorbar scale for tissue_IDs
    _ = plt.colorbar(im, ticks=range(self.k), shrink=0.7)
    # add plots to axes
    i = 1
    for channel in channels:
        ax = plt.subplot(gs[i])
        # make copy for alpha
        im_tmp = image_cp.img[:, :, channel].copy()
        if self[image_index].mask is not None and mask_out:
            # area outside mask NaN
            self.tissue_IDs[image_index][self[image_index].mask == 0] = np.nan
            im = ax.imshow(
                self.tissue_IDs[image_index], cmap=cmap, alpha=im_tmp, **kwargs
            )
        else:
            ax.imshow(self.tissue_IDs[image_index], alpha=im_tmp, **kwargs)
        ax.tick_params(labelbottom=False, labelleft=False)
        sns.despine(bottom=True, left=True)
        ax.set_title(
            label=self[image_index].ch[channel],
            loc=&#34;left&#34;,
            fontweight=&#34;bold&#34;,
            fontsize=16,
        )
        i = i + 1
    fig.tight_layout()
    if save_to:
        plt.savefig(fname=save_to, transparent=True, bbox_inches=&#34;tight&#34;, dpi=300)
    return fig</code></pre>
</details>
</dd>
</dl>
<h3>Inherited members</h3>
<ul class="hlist">
<li><code><b><a title="MILWRM.MILWRM.tissue_labeler" href="MILWRM.html#MILWRM.MILWRM.tissue_labeler">tissue_labeler</a></b></code>:
<ul class="hlist">
<li><code><a title="MILWRM.MILWRM.tissue_labeler.find_optimal_k" href="MILWRM.html#MILWRM.MILWRM.tissue_labeler.find_optimal_k">find_optimal_k</a></code></li>
<li><code><a title="MILWRM.MILWRM.tissue_labeler.find_tissue_regions" href="MILWRM.html#MILWRM.MILWRM.tissue_labeler.find_tissue_regions">find_tissue_regions</a></code></li>
<li><code><a title="MILWRM.MILWRM.tissue_labeler.plot_feature_loadings" href="MILWRM.html#MILWRM.MILWRM.tissue_labeler.plot_feature_loadings">plot_feature_loadings</a></code></li>
<li><code><a title="MILWRM.MILWRM.tissue_labeler.plot_feature_proportions" href="MILWRM.html#MILWRM.MILWRM.tissue_labeler.plot_feature_proportions">plot_feature_proportions</a></code></li>
</ul>
</li>
</ul>
</dd>
<dt id="MILWRM.st_labeler"><code class="flex name class">
<span>class <span class="ident">st_labeler</span></span>
<span>(</span><span>adatas)</span>
</code></dt>
<dd>
<div class="desc"><p>Tissue domain labeling class for spatial transcriptomics (ST) data</p>
<p>Initialize ST tissue labeler class</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>adatas</code></strong> :&ensp;<code>list</code> of <code>anndata.AnnData</code></dt>
<dd>Single anndata object or list of objects to label consensus tissue domains</dd>
</dl>
<h2 id="returns">Returns</h2>
<p>Does not return anything. <code>self.adatas</code> attribute is updated,
<code>self.cluster_data</code> attribute is initiated as <code>None</code>.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class st_labeler(tissue_labeler):
    &#34;&#34;&#34;
    Tissue domain labeling class for spatial transcriptomics (ST) data
    &#34;&#34;&#34;

    def __init__(self, adatas):
        &#34;&#34;&#34;
        Initialize ST tissue labeler class

        Parameters
        ----------
        adatas : list of anndata.AnnData
            Single anndata object or list of objects to label consensus tissue domains

        Returns
        -------
        Does not return anything. `self.adatas` attribute is updated,
        `self.cluster_data` attribute is initiated as `None`.
        &#34;&#34;&#34;
        tissue_labeler.__init__(self)  # initialize parent class
        if not isinstance(adatas, list):  # force single anndata object to list
            adatas = [adatas]
        print(&#34;Initiating ST labeler with {} anndata objects&#34;.format(len(adatas)))
        self.adatas = adatas
        self.raw = adatas.copy()

    def prep_cluster_data(
        self,
        use_rep,
        features=None,
        n_rings=1,
        histo=False,
        fluor_channels=None,
        spatial_graph_key=None,
        n_jobs=-1,
    ):
        &#34;&#34;&#34;
        Prepare master dataframe for tissue-level clustering

        Parameters
        ----------
        use_rep : str
            Representation from `adata.obsm` to use as clustering data (e.g. &#34;X_pca&#34;)
        features : list of int or None, optional (default=`None`)
            List of features to use from `adata.obsm[use_rep]` (e.g. [0,1,2,3,4] to
            use first 5 principal components when `use_rep`=&#34;X_pca&#34;). If `None`, use
            all features from `adata.obsm[use_rep]`
        n_rings : int, optional (default=1)
            Number of hexagonal rings around each spatial transcriptomics spot to blur
            features by for capturing regional information. Assumes 10X Genomics Visium
            platform.
        histo : bool, optional (default `False`)
            Use histology data from Visium anndata object (R,G,B brightfield features)
            in addition to `adata.obsm[use_rep]`? If fluorescent imaging data rather
            than brightfield, use `fluor_channels` argument instead.
        fluor_channels : list of int or None, optional (default `None`)
            Channels from fluorescent image to use for model training (e.g. [1,3] for
            channels 1 and 3 of Visium fluorescent imaging data). If `None`, do not
            use imaging data for training.
        spatial_graph_key : str, optional (default=`None`)
            Key in `adata.obsp` containing spatial graph connectivities (i.e.
            `&#34;spatial_connectivities&#34;`). If `None`, compute new spatial graph using
            `n_rings` in `squidpy`.
        n_jobs : int, optional (default=-1)
            Number of cores to parallelize over. Default all available cores.

        Returns
        -------
        Does not return anything. `self.adatas` are updated, adding &#34;blur_*&#34; features
        to `.obs` if `n_rings &gt; 0`.
        `self.cluster_data` becomes master `np.array` for cluster training.
        Parameters are also captured as attributes for posterity.
        &#34;&#34;&#34;
        if self.cluster_data is not None:
            print(&#34;WARNING: overwriting existing cluster data&#34;)
            self.cluster_data = None
        if features is None:
            self.features = [x for x in range(self.adatas[0].obsm[use_rep].shape[1])]
        else:
            self.features = features
        # save the hyperparams as object attributes
        self.rep = use_rep
        self.histo = histo
        self.fluor_channels = fluor_channels
        self.n_rings = n_rings
        # collect clustering data from self.adatas in parallel
        print(
            &#34;Collecting and blurring {} features from .obsm[{}]...&#34;.format(
                len(self.features),
                use_rep,
            )
        )
        cluster_data = Parallel(n_jobs=n_jobs, verbose=10)(
            delayed(prep_data_single_sample_st)(
                adata,
                adata_i,
                use_rep,
                self.features,
                histo,
                fluor_channels,
                spatial_graph_key,
                n_rings,
            )
            for adata_i, adata in enumerate(self.adatas)
        )
        batch_labels = [
            [x] * len(cluster_data[x]) for x in range(len(cluster_data))
        ]  # batch labels for umap
        self.merged_batch_labels = list(itertools.chain(*batch_labels))
        # concatenate blurred features into cluster_data df for cluster training
        subsampled_data = pd.concat(cluster_data)
        # perform z-scaling on final cluster data
        scaler = StandardScaler()
        self.scaler = scaler.fit(subsampled_data)
        scaled_data = scaler.transform(subsampled_data)
        self.cluster_data = scaled_data
        print(&#34;Collected clustering data of shape: {}&#34;.format(self.cluster_data.shape))

    def label_tissue_regions(
        self, k=None, alpha=0.05, plot_out=True, random_state=18, n_jobs=-1
    ):
        &#34;&#34;&#34;
        Perform tissue-level clustering and label pixels in the corresponding
        `anndata` objects.

        Parameters
        ----------
        k : int, optional (default=None)
            Number of tissue regions to define
        alpha: float
            Manually tuned factor on [0.0, 1.0] that penalizes the number of clusters
        plot_out : boolean, optional (default=True)
            Determines if scaled inertia plot should be output
        random_state : int, optional (default=18)
            Seed for k-means clustering model.
        n_jobs : int
            Number of cores to parallelize k-choosing across

        Returns
        -------
        Does not return anything. `self.adatas` are updated, adding &#34;tissue_ID&#34; field
        to `.obs`. `self.kmeans` contains trained `sklearn` clustering model.
        Parameters are also captured as attributes for posterity.
        &#34;&#34;&#34;
        # find optimal k with parent class
        if k is None:
            print(&#34;Determining optimal cluster number k via scaled inertia&#34;)
            self.find_optimal_k(
                plot_out=plot_out, alpha=alpha, random_state=random_state, n_jobs=n_jobs
            )
        # call k-means model from parent class
        self.find_tissue_regions(k=k, random_state=random_state)
        # loop through anndata object and add tissue labels to adata.obs dataframe
        start = 0
        print(&#34;Adding tissue_ID label to anndata objects&#34;)
        for i in range(len(self.adatas)):
            IDs = self.kmeans.labels_
            self.adatas[i].obs[&#34;tissue_ID&#34;] = IDs[start : start + self.adatas[i].n_obs]
            self.adatas[i].obs[&#34;tissue_ID&#34;] = (
                self.adatas[i].obs[&#34;tissue_ID&#34;].astype(&#34;category&#34;)
            )
            self.adatas[i].obs[&#34;tissue_ID&#34;] = (
                self.adatas[i].obs[&#34;tissue_ID&#34;].cat.set_categories(np.unique(IDs))
            )
            start += self.adatas[i].n_obs

    def confidence_score(self):
        &#34;&#34;&#34;
        estimate confidence score for each visium slide

        Parameters
        ----------

        Returns
        -------
        self.adatas[i].obs.confidence_IDs and self.confidence_score_df are added
        containing confidence score for each tissue domain assignment and mean confidence
        score for each tissue domain within each visium slide
        &#34;&#34;&#34;
        assert (
            self.kmeans is not None
        ), &#34;No cluster results found. Run \
        label_tissue_regions() first.&#34;
        i_slice = 0
        j_slice = 0
        confidence_score_df = pd.DataFrame()
        adatas = self.adatas
        cluster_data = self.cluster_data
        centroids = self.kmeans.cluster_centers_
        for i, adata in enumerate(adatas):
            j_slice = j_slice + adata.n_obs
            data = cluster_data[i_slice:j_slice]
            scores_dict = estimate_confidence_score_st(data, adata, centroids)
            df = pd.DataFrame(scores_dict.values(), columns=[i])
            confidence_score_df = pd.concat([confidence_score_df, df], axis=1)
            i_slice = i_slice + adata.n_obs
        self.confidence_score_df = confidence_score_df

    def plot_gene_loadings(
        self,
        PC_loadings,
        n_genes=10,
        ncols=None,
        titles=None,
        save_to=None,
    ):
        &#34;&#34;&#34;
        Plot MILWRM loadings in gene space specifically for MILWRM done with PCs

        Parameters
        ----------
        PC_loadings : numpy.ndarray
            numpy.ndarray containing PC loadings shape format (genes, components)
        n_genes : int, optional (default=10)
            number of genes to plot
        ncols : int, optional (default=`None`)
            Number of columns for gridspec. If `None`, uses number of tissue domains k.
        titles : list of str, optional (default=`None`)
            Titles of plots corresponding to each MILWRM domain. If `None`, titles
            will be numbers 0 through k.
        save_to : str, optional (default=`None`)
            Path to image file to save plot

        Returns
        -------
        Matplotlib object and PC loadings in gene space set as self.gene_loadings_df
        &#34;&#34;&#34;
        assert (
            self.kmeans is not None
        ), &#34;No cluster results found. Run \
        label_tissue_regions() first.&#34;
        assert (
            PC_loadings.shape[0] == self.adatas[0].n_vars
        ), f&#34;loadings matrix does not, \
        contain enough genes, there should be {self.adatas[0].n_vars} genes&#34;
        assert (
            PC_loadings.shape[1] &gt;= self.kmeans.cluster_centers_.shape[1]
        ), f&#34;loadings matrix \
        does not contain enough components, there should be atleast {self.adatas[0].n_vars} components&#34;
        if titles is None:
            titles = [&#34;tissue_ID &#34; + str(x) for x in range(self.k)]
        centroids = self.kmeans.cluster_centers_
        temp = PC_loadings.T
        loadings = temp[range(self.kmeans.cluster_centers_.shape[1])]
        gene_loadings = np.matmul(centroids, loadings)
        gene_loadings_df = pd.DataFrame(gene_loadings)
        gene_loadings_df = gene_loadings_df.T
        gene_loadings_df[&#34;genes&#34;] = self.adatas[0].var_names
        self.gene_loadings_df = gene_loadings_df
        n_panels = self.k
        if ncols is None:
            ncols = self.k
        if n_panels &lt;= ncols:
            n_rows, n_cols = 1, n_panels
        else:
            n_rows, n_cols = ceil(n_panels / ncols), ncols
        fig = plt.figure(figsize=((ncols * n_cols, ncols * n_rows)))
        left, bottom = 0.1 / n_cols, 0.1 / n_rows
        gs = gridspec.GridSpec(
            nrows=n_rows,
            ncols=n_cols,
            left=left,
            bottom=bottom,
            right=1 - (n_cols - 1) * left - 0.01 / n_cols,
            top=1 - (n_rows - 1) * bottom - 0.1 / n_rows,
        )
        for i in range(self.k):
            df = (
                gene_loadings_df[[i, &#34;genes&#34;]]
                .sort_values(i, axis=0, ascending=False)[:n_genes]
                .reset_index(drop=True)
            )
            plt.subplot(gs[i])
            df_rev = df.sort_values(i).reset_index(drop=True)
            for j, score in enumerate((df_rev[i])):
                plt.text(
                    x=score,
                    y=j + 0.1,
                    s=df_rev.loc[j, &#34;genes&#34;],
                    color=&#34;black&#34;,
                    verticalalignment=&#34;center&#34;,
                    horizontalalignment=&#34;right&#34;,
                    fontsize=&#34;medium&#34;,
                    fontstyle=&#34;italic&#34;,
                )
                plt.ylim([0, j + 1])
                plt.xlim([0, df.max().values[0] + 0.1])
                plt.tick_params(
                    axis=&#34;y&#34;,  # changes apply to the y-axis
                    which=&#34;both&#34;,  # both major and minor ticks are affected
                    left=False,
                    right=False,
                    labelleft=False,
                )
                plt.title(titles[i])
        gs.tight_layout(fig)
        if save_to is not None:
            print(&#34;Saving feature loadings to {}&#34;.format(save_to))
            plt.savefig(save_to)
        else:
            return gs

    def plot_percentage_variance_explained(
        self, fig_size=(5, 5), R_square=False, save_to=None
    ):
        &#34;&#34;&#34;
        plot percentage variance_explained or not explained by clustering

        Parameters
        ----------
        figsize : tuple of float, optional (default=(5,5))
            Size of matplotlib figure
        R_square : Boolean
            Decides if R_square is plotted or S_square
        save_to : str or None
            Path to image file to save results. If `None`, show figure.

        Returns
        -------
        Matplotlib object
        &#34;&#34;&#34;
        assert (
            self.kmeans is not None
        ), &#34;No cluster results found. Run \
        label_tissue_regions() first.&#34;
        centroids = self.kmeans.cluster_centers_
        adatas = self.adatas
        cluster_data = self.cluster_data
        S_squre_for_each_st = []
        R_squre_for_each_st = []
        i_slice = 0
        j_slice = 0
        for adata in adatas:
            j_slice = j_slice + adata.n_obs
            sub_cluster_data = cluster_data[i_slice:j_slice]
            S_square = estimate_percentage_variance_st(
                sub_cluster_data, adata, centroids
            )
            S_squre_for_each_st.append(S_square)
            R_squre_for_each_st.append(100 - S_square)
            i_slice = i_slice + adata.n_obs

        if R_square:
            fig = plt.figure(figsize=fig_size)
            plt.scatter(
                range(len(R_squre_for_each_st)), R_squre_for_each_st, color=&#34;black&#34;
            )
            plt.xlabel(&#34;images&#34;)
            plt.ylabel(&#34;percentage variance explained by Kmeans&#34;)
            plt.ylim((0, 100))
            plt.axhline(
                y=np.mean(R_squre_for_each_st),
                linestyle=&#34;dashed&#34;,
                linewidth=1,
                color=&#34;black&#34;,
            )

        else:
            fig = plt.figure(figsize=fig_size)
            fig = plt.figure(figsize=(5, 5))
            plt.scatter(
                range(len(S_squre_for_each_st)), S_squre_for_each_st, color=&#34;black&#34;
            )
            plt.xlabel(&#34;images&#34;)
            plt.ylabel(&#34;percentage variance explained by Kmeans&#34;)
            plt.ylim((0, 100))
            plt.axhline(
                y=np.mean(S_squre_for_each_st),
                linestyle=&#34;dashed&#34;,
                linewidth=1,
                color=&#34;black&#34;,
            )

        fig.tight_layout()
        if save_to:
            plt.savefig(fname=save_to, transparent=True, bbox_inches=&#34;tight&#34;, dpi=300)
        return fig

    def plot_mse_st(
        self,
        figsize=(5, 5),
        ncols=None,
        labels=None,
        titles=None,
        loc=&#34;lower right&#34;,
        bbox_coordinates=(0, 0, 1.5, 1.5),
        save_to=None,
    ):
        &#34;&#34;&#34;
        estimate mean square error within each tissue domain

        Parameters
        ----------
        fig_size : Tuple
            size for the bar plot
        ncols : int, optional (default=`None`)
            Number of columns for gridspec. If `None`, uses number of tissue domains k.
        labels : list of str, optional (default=`None`)
            Labels corresponding to each image in legend. If `None`, numeric index is
            used for each imaage
        titles : list of str, optional (default=`None`)
            Titles of plots corresponding to each MILWRM domain. If `None`, titles
            will be numbers 0 through k.
        loc : str, optional (default = &#39;lower right&#39;)
            str for legend position
        bbox_coordinates : Tuple, optional (default = (0,0,1.5,1.5))
            coordinates for the legend box
        save_to : str, optional (default=`None`)
            Path to image file to save plot

        Returns
        -------
        Matplotlib object
        &#34;&#34;&#34;
        assert (
            self.kmeans is not None
        ), &#34;No cluster results found. Run \
        label_tissue_regions() first.&#34;
        cluster_data = self.cluster_data
        adatas = self.adatas
        k = self.k
        features = self.features
        centroids = self.kmeans.cluster_centers_
        mse_id = estimate_mse_st(cluster_data, adatas, centroids, k)
        colors = plt.cm.tab20(np.linspace(0, 1, len(adatas)))
        if titles is None:
            titles = [&#34;tissue_domain &#34; + str(x) for x in range(self.k)]
        if labels is None:
            labels = range(len(adatas))
        n_panels = len(mse_id.keys())
        if ncols is None:
            ncols = len(titles)
        if n_panels &lt;= ncols:
            n_rows, n_cols = 1, n_panels
        else:
            n_rows, n_cols = ceil(n_panels / ncols), ncols
        fig = plt.figure(figsize=(n_cols * figsize[0], n_rows * figsize[1]))
        left, bottom = 0.1 / n_cols, 0.1 / n_rows
        gs = gridspec.GridSpec(
            nrows=n_rows,
            ncols=n_cols,
            left=left,
            bottom=bottom,
            right=1 - (n_cols - 1) * left - 0.01 / n_cols,
            top=1 - (n_rows - 1) * bottom - 0.1 / n_rows,
        )
        for i in mse_id.keys():
            plt.subplot(gs[i])
            df = pd.DataFrame.from_dict(mse_id[i])
            plt.boxplot(df, positions=features, showfliers=False)
            for col in df:
                for k in range(len(df[col])):
                    dots = plt.scatter(
                        col,
                        df[col][k],
                        s=k + 1,
                        color=colors[k],
                        label=labels[k] if col == 0 else &#34;&#34;,
                    )
                    offsets = dots.get_offsets()
                    jittered_offsets = offsets
                    # only jitter in the x-direction
                    jittered_offsets[:, 0] += np.random.uniform(
                        -0.3, 0.3, offsets.shape[0]
                    )
                    dots.set_offsets(jittered_offsets)
            plt.xlabel(&#34;PCs&#34;)
            plt.ylabel(&#34;mean square error&#34;)
            plt.title(titles[i])
        plt.legend(loc=loc, bbox_to_anchor=bbox_coordinates)
        gs.tight_layout(fig)
        if save_to:
            plt.savefig(fname=save_to, transparent=True, dpi=300)
        return fig

    def plot_tissue_ID_proportions_st(
        self,
        tID_labels=None,
        slide_labels=None,
        figsize=(5, 5),
        cmap=&#34;tab20&#34;,
        save_to=None,
    ):
        &#34;&#34;&#34;
        Plot proportion of each tissue domain within each slide

        Parameters
        ----------
        tID_labels : list of str, optional (default=`None`)
            List of labels corresponding to MILWRM tissue domains for plotting legend
        slide_labels : list of str, optional (default=`None`)
            List of labels for each slide batch for labeling x-axis
        figsize : tuple of float, optional (default=(5,5))
            Size of matplotlib figure
        cmap : str, optional (default = `&#34;tab20&#34;`)
            Colormap from matplotlib
        save_to : str, optional (default=`None`)
            Path to image file to save plot

        Returns
        -------
        `gridspec.GridSpec` if `save_to` is `None`, else saves plot to file
        &#34;&#34;&#34;
        df_count = pd.DataFrame()
        for adata in self.adatas:
            df = adata.obs[&#34;tissue_ID&#34;].value_counts(normalize=True, sort=False)
            df_count = pd.concat([df_count, df], axis=1)
        df_count = df_count.T.reset_index(drop=True)
        if tID_labels:
            assert (
                len(tID_labels) == df_count.shape[1]
            ), &#34;Length of given tissue domain labels does not match number of tissue domains!&#34;
            df_count.columns = tID_labels
        if slide_labels:
            assert (
                len(slide_labels) == df_count.shape[0]
            ), &#34;Length of given slide labels does not match number of slides!&#34;
            df_count.index = slide_labels
        ax = df_count.plot.bar(stacked=True, cmap=cmap, figsize=figsize)
        ax.legend(loc=&#34;best&#34;, bbox_to_anchor=(1, 1))
        ax.set_xlabel(&#34;slides&#34;)
        ax.set_ylabel(&#34;tissue domain proportion&#34;)
        ax.set_ylim((0, 1))
        plt.tight_layout()
        if save_to is not None:
            ax.figure.savefig(save_to)
        else:
            return ax

    def show_feature_overlay(
        self,
        adata_index,
        pita,
        features=None,
        histo=None,
        cmap=&#34;tab20&#34;,
        label=&#34;feature&#34;,
        ncols=4,
        save_to=None,
        **kwargs,
    ):
        &#34;&#34;&#34;
        Plot tissue_ID with individual pita features as alpha values to distinguish
        expression in identified tissue domains

        Parameters
        ----------
        adata_index : int
            Index of adata from `self.adatas` to plot overlays for (e.g. 0 for first
            adata object)
        pita : np.array
            Image of desired expression in pixel space from `.assemble_pita()`
        features : list of int, optional (default=`None`)
            List of features by index to show in plot. If `None`, use all features.
        histo : np.array or `None`, optional (default=`None`)
            Histology image to show along with pita in gridspec. If `None`, ignore.
        cmap : str, optional (default=&#34;tab20&#34;)
            Matplotlib colormap to use for plotting tissue domains
        label : str
            What to title each panel of the gridspec (i.e. &#34;PC&#34; or &#34;usage&#34;) or each
            channel in RGB image. Can also pass list of names e.g. [&#34;NeuN&#34;,&#34;GFAP&#34;,
            &#34;DAPI&#34;] corresponding to channels.
        ncols : int
            Number of columns for gridspec
        save_to : str or None
            Path to image file to save results. if `None`, show figure.
        **kwargs
            Arguments to pass to `plt.imshow()` function

        Returns
        -------
        Matplotlib object (if plotting one feature or RGB) or gridspec object (for
        multiple features). Saves plot to file if `save_to` is not `None`.
        &#34;&#34;&#34;
        assert pita.ndim &gt; 1, &#34;Pita does not have enough dimensions: {} given&#34;.format(
            pita.ndim
        )
        assert pita.ndim &lt; 4, &#34;Pita has too many dimensions: {} given&#34;.format(pita.ndim)
        # create tissue_ID pita for plotting
        tIDs = assemble_pita(
            self.adatas[adata_index],
            features=&#34;tissue_ID&#34;,
            use_rep=&#34;obs&#34;,
            plot_out=False,
            verbose=False,
        )
        # if pita has multiple features, plot them in gridspec
        if isinstance(features, int):  # force features into list if single integer
            features = [features]
        # if no features are given, use all of them
        elif features is None:
            features = [x + 1 for x in range(pita.shape[2])]
        else:
            assert (
                pita.ndim &gt; 2
            ), &#34;Not enough features in pita: shape {}, expecting 3rd dim with length {}&#34;.format(
                pita.shape, len(features)
            )
            assert (
                len(features) &lt;= pita.shape[2]
            ), &#34;Too many features given: pita has {}, expected {}&#34;.format(
                pita.shape[2], len(features)
            )
        # min-max scale each feature in pita to convert to interpretable alpha values
        mms = MinMaxScaler()
        if pita.ndim == 3:
            pita_tmp = mms.fit_transform(
                pita.reshape((pita.shape[0] * pita.shape[1], pita.shape[2]))
            )
        elif pita.ndim == 2:
            pita_tmp = mms.fit_transform(
                pita.reshape((pita.shape[0] * pita.shape[1], 1))
            )
        # reshape back to original
        pita = pita_tmp.reshape(pita.shape)
        # figure out labels for gridspec plots
        if isinstance(label, str):
            # if label is single string, name channels numerically
            labels = [&#34;{}_{}&#34;.format(label, x) for x in features]
        else:
            assert len(label) == len(
                features
            ), &#34;Please provide the same number of labels as features; {} labels given, {} features given.&#34;.format(
                len(label), len(features)
            )
            labels = label
        # calculate gridspec dimensions
        if histo is not None:
            # determine where the histo image is in anndata
            assert (
                histo
                in self.adatas[adata_index]
                .uns[&#34;spatial&#34;][
                    list(self.adatas[adata_index].uns[&#34;spatial&#34;].keys())[0]
                ][&#34;images&#34;]
                .keys()
            ), &#34;Must provide one of {} for histo&#34;.format(
                self.adatas[adata_index]
                .uns[&#34;spatial&#34;][
                    list(self.adatas[adata_index].uns[&#34;spatial&#34;].keys())[0]
                ][&#34;images&#34;]
                .keys()
            )
            histo = self.adatas[adata_index].uns[&#34;spatial&#34;][
                list(self.adatas[adata_index].uns[&#34;spatial&#34;].keys())[0]
            ][&#34;images&#34;][histo]
            if len(features) + 2 &lt;= ncols:
                n_rows, n_cols = 1, len(features) + 2
            else:
                n_rows, n_cols = ceil((len(features) + 2) / ncols), ncols
            labels = [&#34;Histology&#34;, &#34;tissue_ID&#34;] + labels  # append to front of labels
        else:
            if len(features) + 1 &lt;= ncols:
                n_rows, n_cols = 1, len(features) + 1
            else:
                n_rows, n_cols = ceil(len(features) + 1 / ncols), ncols
            labels = [&#34;tissue_ID&#34;] + labels  # append to front of labels
        fig = plt.figure(figsize=(ncols * n_cols, ncols * n_rows))
        # arrange axes as subplots
        gs = gridspec.GridSpec(n_rows, n_cols, figure=fig)
        # add plots to axes
        i = 0
        if histo is not None:
            # add histology plot to first axes
            ax = plt.subplot(gs[i])
            im = ax.imshow(histo, **kwargs)
            ax.tick_params(labelbottom=False, labelleft=False)
            sns.despine(bottom=True, left=True)
            ax.set_title(
                label=labels[i],
                loc=&#34;left&#34;,
                fontweight=&#34;bold&#34;,
                fontsize=16,
            )
            i = i + 1
        # plot tissue_ID first with colorbar
        ax = plt.subplot(gs[i])
        im = ax.imshow(tIDs, cmap=cmap, **kwargs)
        ax.tick_params(labelbottom=False, labelleft=False)
        sns.despine(bottom=True, left=True)
        ax.set_title(
            label=labels[i],
            loc=&#34;left&#34;,
            fontweight=&#34;bold&#34;,
            fontsize=16,
        )
        # colorbar scale for tissue_IDs
        _ = plt.colorbar(im, shrink=0.7)
        i = i + 1
        for feature in features:
            ax = plt.subplot(gs[i])
            im = ax.imshow(tIDs, alpha=pita[:, :, feature - 1], cmap=cmap, **kwargs)
            ax.tick_params(labelbottom=False, labelleft=False)
            sns.despine(bottom=True, left=True)
            ax.set_title(
                label=labels[i],
                loc=&#34;left&#34;,
                fontweight=&#34;bold&#34;,
                fontsize=16,
            )
            i = i + 1
        fig.tight_layout()
        if save_to:
            plt.savefig(fname=save_to, transparent=True, bbox_inches=&#34;tight&#34;, dpi=300)
        return fig</code></pre>
</details>
<h3>Ancestors</h3>
<ul class="hlist">
<li><a title="MILWRM.MILWRM.tissue_labeler" href="MILWRM.html#MILWRM.MILWRM.tissue_labeler">tissue_labeler</a></li>
</ul>
<h3>Methods</h3>
<dl>
<dt id="MILWRM.st_labeler.confidence_score"><code class="name flex">
<span>def <span class="ident">confidence_score</span></span>(<span>self)</span>
</code></dt>
<dd>
<div class="desc"><p>estimate confidence score for each visium slide</p>
<h2 id="parameters">Parameters</h2>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>self.adatas[i].obs.confidence_IDs and self.confidence_score_df are added</code></dt>
<dd>&nbsp;</dd>
<dt><code>containing confidence score for each tissue domain assignment and mean confidence</code></dt>
<dd>&nbsp;</dd>
<dt><code>score for each tissue domain within each visium slide</code></dt>
<dd>&nbsp;</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def confidence_score(self):
    &#34;&#34;&#34;
    estimate confidence score for each visium slide

    Parameters
    ----------

    Returns
    -------
    self.adatas[i].obs.confidence_IDs and self.confidence_score_df are added
    containing confidence score for each tissue domain assignment and mean confidence
    score for each tissue domain within each visium slide
    &#34;&#34;&#34;
    assert (
        self.kmeans is not None
    ), &#34;No cluster results found. Run \
    label_tissue_regions() first.&#34;
    i_slice = 0
    j_slice = 0
    confidence_score_df = pd.DataFrame()
    adatas = self.adatas
    cluster_data = self.cluster_data
    centroids = self.kmeans.cluster_centers_
    for i, adata in enumerate(adatas):
        j_slice = j_slice + adata.n_obs
        data = cluster_data[i_slice:j_slice]
        scores_dict = estimate_confidence_score_st(data, adata, centroids)
        df = pd.DataFrame(scores_dict.values(), columns=[i])
        confidence_score_df = pd.concat([confidence_score_df, df], axis=1)
        i_slice = i_slice + adata.n_obs
    self.confidence_score_df = confidence_score_df</code></pre>
</details>
</dd>
<dt id="MILWRM.st_labeler.label_tissue_regions"><code class="name flex">
<span>def <span class="ident">label_tissue_regions</span></span>(<span>self, k=None, alpha=0.05, plot_out=True, random_state=18, n_jobs=-1)</span>
</code></dt>
<dd>
<div class="desc"><p>Perform tissue-level clustering and label pixels in the corresponding
<code>anndata</code> objects.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>k</code></strong> :&ensp;<code>int</code>, optional <code>(default=None)</code></dt>
<dd>Number of tissue regions to define</dd>
<dt><strong><code>alpha</code></strong> :&ensp;<code>float</code></dt>
<dd>Manually tuned factor on [0.0, 1.0] that penalizes the number of clusters</dd>
<dt><strong><code>plot_out</code></strong> :&ensp;<code>boolean</code>, optional <code>(default=True)</code></dt>
<dd>Determines if scaled inertia plot should be output</dd>
<dt><strong><code>random_state</code></strong> :&ensp;<code>int</code>, optional <code>(default=18)</code></dt>
<dd>Seed for k-means clustering model.</dd>
<dt><strong><code>n_jobs</code></strong> :&ensp;<code>int</code></dt>
<dd>Number of cores to parallelize k-choosing across</dd>
</dl>
<h2 id="returns">Returns</h2>
<p>Does not return anything. <code>self.adatas</code> are updated, adding "tissue_ID" field
to <code>.obs</code>. <code>self.kmeans</code> contains trained <code>sklearn</code> clustering model.
Parameters are also captured as attributes for posterity.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def label_tissue_regions(
    self, k=None, alpha=0.05, plot_out=True, random_state=18, n_jobs=-1
):
    &#34;&#34;&#34;
    Perform tissue-level clustering and label pixels in the corresponding
    `anndata` objects.

    Parameters
    ----------
    k : int, optional (default=None)
        Number of tissue regions to define
    alpha: float
        Manually tuned factor on [0.0, 1.0] that penalizes the number of clusters
    plot_out : boolean, optional (default=True)
        Determines if scaled inertia plot should be output
    random_state : int, optional (default=18)
        Seed for k-means clustering model.
    n_jobs : int
        Number of cores to parallelize k-choosing across

    Returns
    -------
    Does not return anything. `self.adatas` are updated, adding &#34;tissue_ID&#34; field
    to `.obs`. `self.kmeans` contains trained `sklearn` clustering model.
    Parameters are also captured as attributes for posterity.
    &#34;&#34;&#34;
    # find optimal k with parent class
    if k is None:
        print(&#34;Determining optimal cluster number k via scaled inertia&#34;)
        self.find_optimal_k(
            plot_out=plot_out, alpha=alpha, random_state=random_state, n_jobs=n_jobs
        )
    # call k-means model from parent class
    self.find_tissue_regions(k=k, random_state=random_state)
    # loop through anndata object and add tissue labels to adata.obs dataframe
    start = 0
    print(&#34;Adding tissue_ID label to anndata objects&#34;)
    for i in range(len(self.adatas)):
        IDs = self.kmeans.labels_
        self.adatas[i].obs[&#34;tissue_ID&#34;] = IDs[start : start + self.adatas[i].n_obs]
        self.adatas[i].obs[&#34;tissue_ID&#34;] = (
            self.adatas[i].obs[&#34;tissue_ID&#34;].astype(&#34;category&#34;)
        )
        self.adatas[i].obs[&#34;tissue_ID&#34;] = (
            self.adatas[i].obs[&#34;tissue_ID&#34;].cat.set_categories(np.unique(IDs))
        )
        start += self.adatas[i].n_obs</code></pre>
</details>
</dd>
<dt id="MILWRM.st_labeler.plot_gene_loadings"><code class="name flex">
<span>def <span class="ident">plot_gene_loadings</span></span>(<span>self, PC_loadings, n_genes=10, ncols=None, titles=None, save_to=None)</span>
</code></dt>
<dd>
<div class="desc"><p>Plot MILWRM loadings in gene space specifically for MILWRM done with PCs</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>PC_loadings</code></strong> :&ensp;<code>numpy.ndarray</code></dt>
<dd>numpy.ndarray containing PC loadings shape format (genes, components)</dd>
<dt><strong><code>n_genes</code></strong> :&ensp;<code>int</code>, optional <code>(default=10)</code></dt>
<dd>number of genes to plot</dd>
<dt><strong><code>ncols</code></strong> :&ensp;<code>int</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>Number of columns for gridspec. If <code>None</code>, uses number of tissue domains k.</dd>
<dt><strong><code>titles</code></strong> :&ensp;<code>list</code> of <code>str</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>Titles of plots corresponding to each MILWRM domain. If <code>None</code>, titles
will be numbers 0 through k.</dd>
<dt><strong><code>save_to</code></strong> :&ensp;<code>str</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>Path to image file to save plot</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>Matplotlib object and PC loadings in gene space set as self.gene_loadings_df</code></dt>
<dd>&nbsp;</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def plot_gene_loadings(
    self,
    PC_loadings,
    n_genes=10,
    ncols=None,
    titles=None,
    save_to=None,
):
    &#34;&#34;&#34;
    Plot MILWRM loadings in gene space specifically for MILWRM done with PCs

    Parameters
    ----------
    PC_loadings : numpy.ndarray
        numpy.ndarray containing PC loadings shape format (genes, components)
    n_genes : int, optional (default=10)
        number of genes to plot
    ncols : int, optional (default=`None`)
        Number of columns for gridspec. If `None`, uses number of tissue domains k.
    titles : list of str, optional (default=`None`)
        Titles of plots corresponding to each MILWRM domain. If `None`, titles
        will be numbers 0 through k.
    save_to : str, optional (default=`None`)
        Path to image file to save plot

    Returns
    -------
    Matplotlib object and PC loadings in gene space set as self.gene_loadings_df
    &#34;&#34;&#34;
    assert (
        self.kmeans is not None
    ), &#34;No cluster results found. Run \
    label_tissue_regions() first.&#34;
    assert (
        PC_loadings.shape[0] == self.adatas[0].n_vars
    ), f&#34;loadings matrix does not, \
    contain enough genes, there should be {self.adatas[0].n_vars} genes&#34;
    assert (
        PC_loadings.shape[1] &gt;= self.kmeans.cluster_centers_.shape[1]
    ), f&#34;loadings matrix \
    does not contain enough components, there should be atleast {self.adatas[0].n_vars} components&#34;
    if titles is None:
        titles = [&#34;tissue_ID &#34; + str(x) for x in range(self.k)]
    centroids = self.kmeans.cluster_centers_
    temp = PC_loadings.T
    loadings = temp[range(self.kmeans.cluster_centers_.shape[1])]
    gene_loadings = np.matmul(centroids, loadings)
    gene_loadings_df = pd.DataFrame(gene_loadings)
    gene_loadings_df = gene_loadings_df.T
    gene_loadings_df[&#34;genes&#34;] = self.adatas[0].var_names
    self.gene_loadings_df = gene_loadings_df
    n_panels = self.k
    if ncols is None:
        ncols = self.k
    if n_panels &lt;= ncols:
        n_rows, n_cols = 1, n_panels
    else:
        n_rows, n_cols = ceil(n_panels / ncols), ncols
    fig = plt.figure(figsize=((ncols * n_cols, ncols * n_rows)))
    left, bottom = 0.1 / n_cols, 0.1 / n_rows
    gs = gridspec.GridSpec(
        nrows=n_rows,
        ncols=n_cols,
        left=left,
        bottom=bottom,
        right=1 - (n_cols - 1) * left - 0.01 / n_cols,
        top=1 - (n_rows - 1) * bottom - 0.1 / n_rows,
    )
    for i in range(self.k):
        df = (
            gene_loadings_df[[i, &#34;genes&#34;]]
            .sort_values(i, axis=0, ascending=False)[:n_genes]
            .reset_index(drop=True)
        )
        plt.subplot(gs[i])
        df_rev = df.sort_values(i).reset_index(drop=True)
        for j, score in enumerate((df_rev[i])):
            plt.text(
                x=score,
                y=j + 0.1,
                s=df_rev.loc[j, &#34;genes&#34;],
                color=&#34;black&#34;,
                verticalalignment=&#34;center&#34;,
                horizontalalignment=&#34;right&#34;,
                fontsize=&#34;medium&#34;,
                fontstyle=&#34;italic&#34;,
            )
            plt.ylim([0, j + 1])
            plt.xlim([0, df.max().values[0] + 0.1])
            plt.tick_params(
                axis=&#34;y&#34;,  # changes apply to the y-axis
                which=&#34;both&#34;,  # both major and minor ticks are affected
                left=False,
                right=False,
                labelleft=False,
            )
            plt.title(titles[i])
    gs.tight_layout(fig)
    if save_to is not None:
        print(&#34;Saving feature loadings to {}&#34;.format(save_to))
        plt.savefig(save_to)
    else:
        return gs</code></pre>
</details>
</dd>
<dt id="MILWRM.st_labeler.plot_mse_st"><code class="name flex">
<span>def <span class="ident">plot_mse_st</span></span>(<span>self, figsize=(5, 5), ncols=None, labels=None, titles=None, loc='lower right', bbox_coordinates=(0, 0, 1.5, 1.5), save_to=None)</span>
</code></dt>
<dd>
<div class="desc"><p>estimate mean square error within each tissue domain</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>fig_size</code></strong> :&ensp;<code>Tuple</code></dt>
<dd>size for the bar plot</dd>
<dt><strong><code>ncols</code></strong> :&ensp;<code>int</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>Number of columns for gridspec. If <code>None</code>, uses number of tissue domains k.</dd>
<dt><strong><code>labels</code></strong> :&ensp;<code>list</code> of <code>str</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>Labels corresponding to each image in legend. If <code>None</code>, numeric index is
used for each imaage</dd>
<dt><strong><code>titles</code></strong> :&ensp;<code>list</code> of <code>str</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>Titles of plots corresponding to each MILWRM domain. If <code>None</code>, titles
will be numbers 0 through k.</dd>
<dt><strong><code>loc</code></strong> :&ensp;<code>str</code>, optional <code>(default = 'lower right')</code></dt>
<dd>str for legend position</dd>
<dt><strong><code>bbox_coordinates</code></strong> :&ensp;<code>Tuple</code>, optional <code>(default = (0,0,1.5,1.5))</code></dt>
<dd>coordinates for the legend box</dd>
<dt><strong><code>save_to</code></strong> :&ensp;<code>str</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>Path to image file to save plot</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>Matplotlib object</code></dt>
<dd>&nbsp;</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def plot_mse_st(
    self,
    figsize=(5, 5),
    ncols=None,
    labels=None,
    titles=None,
    loc=&#34;lower right&#34;,
    bbox_coordinates=(0, 0, 1.5, 1.5),
    save_to=None,
):
    &#34;&#34;&#34;
    estimate mean square error within each tissue domain

    Parameters
    ----------
    fig_size : Tuple
        size for the bar plot
    ncols : int, optional (default=`None`)
        Number of columns for gridspec. If `None`, uses number of tissue domains k.
    labels : list of str, optional (default=`None`)
        Labels corresponding to each image in legend. If `None`, numeric index is
        used for each imaage
    titles : list of str, optional (default=`None`)
        Titles of plots corresponding to each MILWRM domain. If `None`, titles
        will be numbers 0 through k.
    loc : str, optional (default = &#39;lower right&#39;)
        str for legend position
    bbox_coordinates : Tuple, optional (default = (0,0,1.5,1.5))
        coordinates for the legend box
    save_to : str, optional (default=`None`)
        Path to image file to save plot

    Returns
    -------
    Matplotlib object
    &#34;&#34;&#34;
    assert (
        self.kmeans is not None
    ), &#34;No cluster results found. Run \
    label_tissue_regions() first.&#34;
    cluster_data = self.cluster_data
    adatas = self.adatas
    k = self.k
    features = self.features
    centroids = self.kmeans.cluster_centers_
    mse_id = estimate_mse_st(cluster_data, adatas, centroids, k)
    colors = plt.cm.tab20(np.linspace(0, 1, len(adatas)))
    if titles is None:
        titles = [&#34;tissue_domain &#34; + str(x) for x in range(self.k)]
    if labels is None:
        labels = range(len(adatas))
    n_panels = len(mse_id.keys())
    if ncols is None:
        ncols = len(titles)
    if n_panels &lt;= ncols:
        n_rows, n_cols = 1, n_panels
    else:
        n_rows, n_cols = ceil(n_panels / ncols), ncols
    fig = plt.figure(figsize=(n_cols * figsize[0], n_rows * figsize[1]))
    left, bottom = 0.1 / n_cols, 0.1 / n_rows
    gs = gridspec.GridSpec(
        nrows=n_rows,
        ncols=n_cols,
        left=left,
        bottom=bottom,
        right=1 - (n_cols - 1) * left - 0.01 / n_cols,
        top=1 - (n_rows - 1) * bottom - 0.1 / n_rows,
    )
    for i in mse_id.keys():
        plt.subplot(gs[i])
        df = pd.DataFrame.from_dict(mse_id[i])
        plt.boxplot(df, positions=features, showfliers=False)
        for col in df:
            for k in range(len(df[col])):
                dots = plt.scatter(
                    col,
                    df[col][k],
                    s=k + 1,
                    color=colors[k],
                    label=labels[k] if col == 0 else &#34;&#34;,
                )
                offsets = dots.get_offsets()
                jittered_offsets = offsets
                # only jitter in the x-direction
                jittered_offsets[:, 0] += np.random.uniform(
                    -0.3, 0.3, offsets.shape[0]
                )
                dots.set_offsets(jittered_offsets)
        plt.xlabel(&#34;PCs&#34;)
        plt.ylabel(&#34;mean square error&#34;)
        plt.title(titles[i])
    plt.legend(loc=loc, bbox_to_anchor=bbox_coordinates)
    gs.tight_layout(fig)
    if save_to:
        plt.savefig(fname=save_to, transparent=True, dpi=300)
    return fig</code></pre>
</details>
</dd>
<dt id="MILWRM.st_labeler.plot_percentage_variance_explained"><code class="name flex">
<span>def <span class="ident">plot_percentage_variance_explained</span></span>(<span>self, fig_size=(5, 5), R_square=False, save_to=None)</span>
</code></dt>
<dd>
<div class="desc"><p>plot percentage variance_explained or not explained by clustering</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>figsize</code></strong> :&ensp;<code>tuple</code> of <code>float</code>, optional <code>(default=(5,5))</code></dt>
<dd>Size of matplotlib figure</dd>
<dt><strong><code>R_square</code></strong> :&ensp;<code>Boolean</code></dt>
<dd>Decides if R_square is plotted or S_square</dd>
<dt><strong><code>save_to</code></strong> :&ensp;<code>str</code> or <code>None</code></dt>
<dd>Path to image file to save results. If <code>None</code>, show figure.</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>Matplotlib object</code></dt>
<dd>&nbsp;</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def plot_percentage_variance_explained(
    self, fig_size=(5, 5), R_square=False, save_to=None
):
    &#34;&#34;&#34;
    plot percentage variance_explained or not explained by clustering

    Parameters
    ----------
    figsize : tuple of float, optional (default=(5,5))
        Size of matplotlib figure
    R_square : Boolean
        Decides if R_square is plotted or S_square
    save_to : str or None
        Path to image file to save results. If `None`, show figure.

    Returns
    -------
    Matplotlib object
    &#34;&#34;&#34;
    assert (
        self.kmeans is not None
    ), &#34;No cluster results found. Run \
    label_tissue_regions() first.&#34;
    centroids = self.kmeans.cluster_centers_
    adatas = self.adatas
    cluster_data = self.cluster_data
    S_squre_for_each_st = []
    R_squre_for_each_st = []
    i_slice = 0
    j_slice = 0
    for adata in adatas:
        j_slice = j_slice + adata.n_obs
        sub_cluster_data = cluster_data[i_slice:j_slice]
        S_square = estimate_percentage_variance_st(
            sub_cluster_data, adata, centroids
        )
        S_squre_for_each_st.append(S_square)
        R_squre_for_each_st.append(100 - S_square)
        i_slice = i_slice + adata.n_obs

    if R_square:
        fig = plt.figure(figsize=fig_size)
        plt.scatter(
            range(len(R_squre_for_each_st)), R_squre_for_each_st, color=&#34;black&#34;
        )
        plt.xlabel(&#34;images&#34;)
        plt.ylabel(&#34;percentage variance explained by Kmeans&#34;)
        plt.ylim((0, 100))
        plt.axhline(
            y=np.mean(R_squre_for_each_st),
            linestyle=&#34;dashed&#34;,
            linewidth=1,
            color=&#34;black&#34;,
        )

    else:
        fig = plt.figure(figsize=fig_size)
        fig = plt.figure(figsize=(5, 5))
        plt.scatter(
            range(len(S_squre_for_each_st)), S_squre_for_each_st, color=&#34;black&#34;
        )
        plt.xlabel(&#34;images&#34;)
        plt.ylabel(&#34;percentage variance explained by Kmeans&#34;)
        plt.ylim((0, 100))
        plt.axhline(
            y=np.mean(S_squre_for_each_st),
            linestyle=&#34;dashed&#34;,
            linewidth=1,
            color=&#34;black&#34;,
        )

    fig.tight_layout()
    if save_to:
        plt.savefig(fname=save_to, transparent=True, bbox_inches=&#34;tight&#34;, dpi=300)
    return fig</code></pre>
</details>
</dd>
<dt id="MILWRM.st_labeler.plot_tissue_ID_proportions_st"><code class="name flex">
<span>def <span class="ident">plot_tissue_ID_proportions_st</span></span>(<span>self, tID_labels=None, slide_labels=None, figsize=(5, 5), cmap='tab20', save_to=None)</span>
</code></dt>
<dd>
<div class="desc"><p>Plot proportion of each tissue domain within each slide</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>tID_labels</code></strong> :&ensp;<code>list</code> of <code>str</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>List of labels corresponding to MILWRM tissue domains for plotting legend</dd>
<dt><strong><code>slide_labels</code></strong> :&ensp;<code>list</code> of <code>str</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>List of labels for each slide batch for labeling x-axis</dd>
<dt><strong><code>figsize</code></strong> :&ensp;<code>tuple</code> of <code>float</code>, optional <code>(default=(5,5))</code></dt>
<dd>Size of matplotlib figure</dd>
<dt><strong><code>cmap</code></strong> :&ensp;<code>str</code>, optional <code>(default =</code>"tab20"<code>)</code></dt>
<dd>Colormap from matplotlib</dd>
<dt><strong><code>save_to</code></strong> :&ensp;<code>str</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>Path to image file to save plot</dd>
</dl>
<h2 id="returns">Returns</h2>
<p><code>gridspec.GridSpec</code> if <code>save_to</code> is <code>None</code>, else saves plot to file</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def plot_tissue_ID_proportions_st(
    self,
    tID_labels=None,
    slide_labels=None,
    figsize=(5, 5),
    cmap=&#34;tab20&#34;,
    save_to=None,
):
    &#34;&#34;&#34;
    Plot proportion of each tissue domain within each slide

    Parameters
    ----------
    tID_labels : list of str, optional (default=`None`)
        List of labels corresponding to MILWRM tissue domains for plotting legend
    slide_labels : list of str, optional (default=`None`)
        List of labels for each slide batch for labeling x-axis
    figsize : tuple of float, optional (default=(5,5))
        Size of matplotlib figure
    cmap : str, optional (default = `&#34;tab20&#34;`)
        Colormap from matplotlib
    save_to : str, optional (default=`None`)
        Path to image file to save plot

    Returns
    -------
    `gridspec.GridSpec` if `save_to` is `None`, else saves plot to file
    &#34;&#34;&#34;
    df_count = pd.DataFrame()
    for adata in self.adatas:
        df = adata.obs[&#34;tissue_ID&#34;].value_counts(normalize=True, sort=False)
        df_count = pd.concat([df_count, df], axis=1)
    df_count = df_count.T.reset_index(drop=True)
    if tID_labels:
        assert (
            len(tID_labels) == df_count.shape[1]
        ), &#34;Length of given tissue domain labels does not match number of tissue domains!&#34;
        df_count.columns = tID_labels
    if slide_labels:
        assert (
            len(slide_labels) == df_count.shape[0]
        ), &#34;Length of given slide labels does not match number of slides!&#34;
        df_count.index = slide_labels
    ax = df_count.plot.bar(stacked=True, cmap=cmap, figsize=figsize)
    ax.legend(loc=&#34;best&#34;, bbox_to_anchor=(1, 1))
    ax.set_xlabel(&#34;slides&#34;)
    ax.set_ylabel(&#34;tissue domain proportion&#34;)
    ax.set_ylim((0, 1))
    plt.tight_layout()
    if save_to is not None:
        ax.figure.savefig(save_to)
    else:
        return ax</code></pre>
</details>
</dd>
<dt id="MILWRM.st_labeler.prep_cluster_data"><code class="name flex">
<span>def <span class="ident">prep_cluster_data</span></span>(<span>self, use_rep, features=None, n_rings=1, histo=False, fluor_channels=None, spatial_graph_key=None, n_jobs=-1)</span>
</code></dt>
<dd>
<div class="desc"><p>Prepare master dataframe for tissue-level clustering</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>use_rep</code></strong> :&ensp;<code>str</code></dt>
<dd>Representation from <code>adata.obsm</code> to use as clustering data (e.g. "X_pca")</dd>
<dt><strong><code>features</code></strong> :&ensp;<code>list</code> of <code>int</code> or <code>None</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>List of features to use from <code>adata.obsm[use_rep]</code> (e.g. [0,1,2,3,4] to
use first 5 principal components when <code>use_rep</code>="X_pca"). If <code>None</code>, use
all features from <code>adata.obsm[use_rep]</code></dd>
<dt><strong><code>n_rings</code></strong> :&ensp;<code>int</code>, optional <code>(default=1)</code></dt>
<dd>Number of hexagonal rings around each spatial transcriptomics spot to blur
features by for capturing regional information. Assumes 10X Genomics Visium
platform.</dd>
<dt><strong><code>histo</code></strong> :&ensp;<code>bool</code>, optional <code>(default </code>False<code>)</code></dt>
<dd>Use histology data from Visium anndata object (R,G,B brightfield features)
in addition to <code>adata.obsm[use_rep]</code>? If fluorescent imaging data rather
than brightfield, use <code>fluor_channels</code> argument instead.</dd>
<dt><strong><code>fluor_channels</code></strong> :&ensp;<code>list</code> of <code>int</code> or <code>None</code>, optional <code>(default </code>None<code>)</code></dt>
<dd>Channels from fluorescent image to use for model training (e.g. [1,3] for
channels 1 and 3 of Visium fluorescent imaging data). If <code>None</code>, do not
use imaging data for training.</dd>
<dt><strong><code>spatial_graph_key</code></strong> :&ensp;<code>str</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>Key in <code>adata.obsp</code> containing spatial graph connectivities (i.e.
<code>"spatial_connectivities"</code>). If <code>None</code>, compute new spatial graph using
<code>n_rings</code> in <code>squidpy</code>.</dd>
<dt><strong><code>n_jobs</code></strong> :&ensp;<code>int</code>, optional <code>(default=-1)</code></dt>
<dd>Number of cores to parallelize over. Default all available cores.</dd>
</dl>
<h2 id="returns">Returns</h2>
<p>Does not return anything. <code>self.adatas</code> are updated, adding "blur_*" features
to <code>.obs</code> if <code>n_rings &gt; 0</code>.
<code>self.cluster_data</code> becomes master <code>np.array</code> for cluster training.
Parameters are also captured as attributes for posterity.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def prep_cluster_data(
    self,
    use_rep,
    features=None,
    n_rings=1,
    histo=False,
    fluor_channels=None,
    spatial_graph_key=None,
    n_jobs=-1,
):
    &#34;&#34;&#34;
    Prepare master dataframe for tissue-level clustering

    Parameters
    ----------
    use_rep : str
        Representation from `adata.obsm` to use as clustering data (e.g. &#34;X_pca&#34;)
    features : list of int or None, optional (default=`None`)
        List of features to use from `adata.obsm[use_rep]` (e.g. [0,1,2,3,4] to
        use first 5 principal components when `use_rep`=&#34;X_pca&#34;). If `None`, use
        all features from `adata.obsm[use_rep]`
    n_rings : int, optional (default=1)
        Number of hexagonal rings around each spatial transcriptomics spot to blur
        features by for capturing regional information. Assumes 10X Genomics Visium
        platform.
    histo : bool, optional (default `False`)
        Use histology data from Visium anndata object (R,G,B brightfield features)
        in addition to `adata.obsm[use_rep]`? If fluorescent imaging data rather
        than brightfield, use `fluor_channels` argument instead.
    fluor_channels : list of int or None, optional (default `None`)
        Channels from fluorescent image to use for model training (e.g. [1,3] for
        channels 1 and 3 of Visium fluorescent imaging data). If `None`, do not
        use imaging data for training.
    spatial_graph_key : str, optional (default=`None`)
        Key in `adata.obsp` containing spatial graph connectivities (i.e.
        `&#34;spatial_connectivities&#34;`). If `None`, compute new spatial graph using
        `n_rings` in `squidpy`.
    n_jobs : int, optional (default=-1)
        Number of cores to parallelize over. Default all available cores.

    Returns
    -------
    Does not return anything. `self.adatas` are updated, adding &#34;blur_*&#34; features
    to `.obs` if `n_rings &gt; 0`.
    `self.cluster_data` becomes master `np.array` for cluster training.
    Parameters are also captured as attributes for posterity.
    &#34;&#34;&#34;
    if self.cluster_data is not None:
        print(&#34;WARNING: overwriting existing cluster data&#34;)
        self.cluster_data = None
    if features is None:
        self.features = [x for x in range(self.adatas[0].obsm[use_rep].shape[1])]
    else:
        self.features = features
    # save the hyperparams as object attributes
    self.rep = use_rep
    self.histo = histo
    self.fluor_channels = fluor_channels
    self.n_rings = n_rings
    # collect clustering data from self.adatas in parallel
    print(
        &#34;Collecting and blurring {} features from .obsm[{}]...&#34;.format(
            len(self.features),
            use_rep,
        )
    )
    cluster_data = Parallel(n_jobs=n_jobs, verbose=10)(
        delayed(prep_data_single_sample_st)(
            adata,
            adata_i,
            use_rep,
            self.features,
            histo,
            fluor_channels,
            spatial_graph_key,
            n_rings,
        )
        for adata_i, adata in enumerate(self.adatas)
    )
    batch_labels = [
        [x] * len(cluster_data[x]) for x in range(len(cluster_data))
    ]  # batch labels for umap
    self.merged_batch_labels = list(itertools.chain(*batch_labels))
    # concatenate blurred features into cluster_data df for cluster training
    subsampled_data = pd.concat(cluster_data)
    # perform z-scaling on final cluster data
    scaler = StandardScaler()
    self.scaler = scaler.fit(subsampled_data)
    scaled_data = scaler.transform(subsampled_data)
    self.cluster_data = scaled_data
    print(&#34;Collected clustering data of shape: {}&#34;.format(self.cluster_data.shape))</code></pre>
</details>
</dd>
<dt id="MILWRM.st_labeler.show_feature_overlay"><code class="name flex">
<span>def <span class="ident">show_feature_overlay</span></span>(<span>self, adata_index, pita, features=None, histo=None, cmap='tab20', label='feature', ncols=4, save_to=None, **kwargs)</span>
</code></dt>
<dd>
<div class="desc"><p>Plot tissue_ID with individual pita features as alpha values to distinguish
expression in identified tissue domains</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>adata_index</code></strong> :&ensp;<code>int</code></dt>
<dd>Index of adata from <code>self.adatas</code> to plot overlays for (e.g. 0 for first
adata object)</dd>
<dt><strong><code>pita</code></strong> :&ensp;<code>np.array</code></dt>
<dd>Image of desired expression in pixel space from <code>.<a title="MILWRM.assemble_pita" href="#MILWRM.assemble_pita">assemble_pita()</a></code></dd>
<dt><strong><code>features</code></strong> :&ensp;<code>list</code> of <code>int</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>List of features by index to show in plot. If <code>None</code>, use all features.</dd>
<dt><strong><code>histo</code></strong> :&ensp;<code>np.array</code> or <code>None</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>Histology image to show along with pita in gridspec. If <code>None</code>, ignore.</dd>
<dt><strong><code>cmap</code></strong> :&ensp;<code>str</code>, optional <code>(default="tab20")</code></dt>
<dd>Matplotlib colormap to use for plotting tissue domains</dd>
<dt><strong><code>label</code></strong> :&ensp;<code>str</code></dt>
<dd>What to title each panel of the gridspec (i.e. "PC" or "usage") or each
channel in RGB image. Can also pass list of names e.g. ["NeuN","GFAP",
"DAPI"] corresponding to channels.</dd>
<dt><strong><code>ncols</code></strong> :&ensp;<code>int</code></dt>
<dd>Number of columns for gridspec</dd>
<dt><strong><code>save_to</code></strong> :&ensp;<code>str</code> or <code>None</code></dt>
<dd>Path to image file to save results. if <code>None</code>, show figure.</dd>
<dt><strong><code>**kwargs</code></strong></dt>
<dd>Arguments to pass to <code>plt.imshow()</code> function</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>Matplotlib object (if plotting one feature</code> or <code>RGB)</code> or <code>gridspec object (for</code></dt>
<dd>&nbsp;</dd>
</dl>
<p>multiple features). Saves plot to file if <code>save_to</code> is not <code>None</code>.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def show_feature_overlay(
    self,
    adata_index,
    pita,
    features=None,
    histo=None,
    cmap=&#34;tab20&#34;,
    label=&#34;feature&#34;,
    ncols=4,
    save_to=None,
    **kwargs,
):
    &#34;&#34;&#34;
    Plot tissue_ID with individual pita features as alpha values to distinguish
    expression in identified tissue domains

    Parameters
    ----------
    adata_index : int
        Index of adata from `self.adatas` to plot overlays for (e.g. 0 for first
        adata object)
    pita : np.array
        Image of desired expression in pixel space from `.assemble_pita()`
    features : list of int, optional (default=`None`)
        List of features by index to show in plot. If `None`, use all features.
    histo : np.array or `None`, optional (default=`None`)
        Histology image to show along with pita in gridspec. If `None`, ignore.
    cmap : str, optional (default=&#34;tab20&#34;)
        Matplotlib colormap to use for plotting tissue domains
    label : str
        What to title each panel of the gridspec (i.e. &#34;PC&#34; or &#34;usage&#34;) or each
        channel in RGB image. Can also pass list of names e.g. [&#34;NeuN&#34;,&#34;GFAP&#34;,
        &#34;DAPI&#34;] corresponding to channels.
    ncols : int
        Number of columns for gridspec
    save_to : str or None
        Path to image file to save results. if `None`, show figure.
    **kwargs
        Arguments to pass to `plt.imshow()` function

    Returns
    -------
    Matplotlib object (if plotting one feature or RGB) or gridspec object (for
    multiple features). Saves plot to file if `save_to` is not `None`.
    &#34;&#34;&#34;
    assert pita.ndim &gt; 1, &#34;Pita does not have enough dimensions: {} given&#34;.format(
        pita.ndim
    )
    assert pita.ndim &lt; 4, &#34;Pita has too many dimensions: {} given&#34;.format(pita.ndim)
    # create tissue_ID pita for plotting
    tIDs = assemble_pita(
        self.adatas[adata_index],
        features=&#34;tissue_ID&#34;,
        use_rep=&#34;obs&#34;,
        plot_out=False,
        verbose=False,
    )
    # if pita has multiple features, plot them in gridspec
    if isinstance(features, int):  # force features into list if single integer
        features = [features]
    # if no features are given, use all of them
    elif features is None:
        features = [x + 1 for x in range(pita.shape[2])]
    else:
        assert (
            pita.ndim &gt; 2
        ), &#34;Not enough features in pita: shape {}, expecting 3rd dim with length {}&#34;.format(
            pita.shape, len(features)
        )
        assert (
            len(features) &lt;= pita.shape[2]
        ), &#34;Too many features given: pita has {}, expected {}&#34;.format(
            pita.shape[2], len(features)
        )
    # min-max scale each feature in pita to convert to interpretable alpha values
    mms = MinMaxScaler()
    if pita.ndim == 3:
        pita_tmp = mms.fit_transform(
            pita.reshape((pita.shape[0] * pita.shape[1], pita.shape[2]))
        )
    elif pita.ndim == 2:
        pita_tmp = mms.fit_transform(
            pita.reshape((pita.shape[0] * pita.shape[1], 1))
        )
    # reshape back to original
    pita = pita_tmp.reshape(pita.shape)
    # figure out labels for gridspec plots
    if isinstance(label, str):
        # if label is single string, name channels numerically
        labels = [&#34;{}_{}&#34;.format(label, x) for x in features]
    else:
        assert len(label) == len(
            features
        ), &#34;Please provide the same number of labels as features; {} labels given, {} features given.&#34;.format(
            len(label), len(features)
        )
        labels = label
    # calculate gridspec dimensions
    if histo is not None:
        # determine where the histo image is in anndata
        assert (
            histo
            in self.adatas[adata_index]
            .uns[&#34;spatial&#34;][
                list(self.adatas[adata_index].uns[&#34;spatial&#34;].keys())[0]
            ][&#34;images&#34;]
            .keys()
        ), &#34;Must provide one of {} for histo&#34;.format(
            self.adatas[adata_index]
            .uns[&#34;spatial&#34;][
                list(self.adatas[adata_index].uns[&#34;spatial&#34;].keys())[0]
            ][&#34;images&#34;]
            .keys()
        )
        histo = self.adatas[adata_index].uns[&#34;spatial&#34;][
            list(self.adatas[adata_index].uns[&#34;spatial&#34;].keys())[0]
        ][&#34;images&#34;][histo]
        if len(features) + 2 &lt;= ncols:
            n_rows, n_cols = 1, len(features) + 2
        else:
            n_rows, n_cols = ceil((len(features) + 2) / ncols), ncols
        labels = [&#34;Histology&#34;, &#34;tissue_ID&#34;] + labels  # append to front of labels
    else:
        if len(features) + 1 &lt;= ncols:
            n_rows, n_cols = 1, len(features) + 1
        else:
            n_rows, n_cols = ceil(len(features) + 1 / ncols), ncols
        labels = [&#34;tissue_ID&#34;] + labels  # append to front of labels
    fig = plt.figure(figsize=(ncols * n_cols, ncols * n_rows))
    # arrange axes as subplots
    gs = gridspec.GridSpec(n_rows, n_cols, figure=fig)
    # add plots to axes
    i = 0
    if histo is not None:
        # add histology plot to first axes
        ax = plt.subplot(gs[i])
        im = ax.imshow(histo, **kwargs)
        ax.tick_params(labelbottom=False, labelleft=False)
        sns.despine(bottom=True, left=True)
        ax.set_title(
            label=labels[i],
            loc=&#34;left&#34;,
            fontweight=&#34;bold&#34;,
            fontsize=16,
        )
        i = i + 1
    # plot tissue_ID first with colorbar
    ax = plt.subplot(gs[i])
    im = ax.imshow(tIDs, cmap=cmap, **kwargs)
    ax.tick_params(labelbottom=False, labelleft=False)
    sns.despine(bottom=True, left=True)
    ax.set_title(
        label=labels[i],
        loc=&#34;left&#34;,
        fontweight=&#34;bold&#34;,
        fontsize=16,
    )
    # colorbar scale for tissue_IDs
    _ = plt.colorbar(im, shrink=0.7)
    i = i + 1
    for feature in features:
        ax = plt.subplot(gs[i])
        im = ax.imshow(tIDs, alpha=pita[:, :, feature - 1], cmap=cmap, **kwargs)
        ax.tick_params(labelbottom=False, labelleft=False)
        sns.despine(bottom=True, left=True)
        ax.set_title(
            label=labels[i],
            loc=&#34;left&#34;,
            fontweight=&#34;bold&#34;,
            fontsize=16,
        )
        i = i + 1
    fig.tight_layout()
    if save_to:
        plt.savefig(fname=save_to, transparent=True, bbox_inches=&#34;tight&#34;, dpi=300)
    return fig</code></pre>
</details>
</dd>
</dl>
<h3>Inherited members</h3>
<ul class="hlist">
<li><code><b><a title="MILWRM.MILWRM.tissue_labeler" href="MILWRM.html#MILWRM.MILWRM.tissue_labeler">tissue_labeler</a></b></code>:
<ul class="hlist">
<li><code><a title="MILWRM.MILWRM.tissue_labeler.find_optimal_k" href="MILWRM.html#MILWRM.MILWRM.tissue_labeler.find_optimal_k">find_optimal_k</a></code></li>
<li><code><a title="MILWRM.MILWRM.tissue_labeler.find_tissue_regions" href="MILWRM.html#MILWRM.MILWRM.tissue_labeler.find_tissue_regions">find_tissue_regions</a></code></li>
<li><code><a title="MILWRM.MILWRM.tissue_labeler.plot_feature_loadings" href="MILWRM.html#MILWRM.MILWRM.tissue_labeler.plot_feature_loadings">plot_feature_loadings</a></code></li>
<li><code><a title="MILWRM.MILWRM.tissue_labeler.plot_feature_proportions" href="MILWRM.html#MILWRM.MILWRM.tissue_labeler.plot_feature_proportions">plot_feature_proportions</a></code></li>
</ul>
</li>
</ul>
</dd>
</dl>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3><a href="#header-submodules">Sub-modules</a></h3>
<ul>
<li><code><a title="MILWRM.MILWRM" href="MILWRM.html">MILWRM.MILWRM</a></code></li>
<li><code><a title="MILWRM.MxIF" href="MxIF.html">MILWRM.MxIF</a></code></li>
<li><code><a title="MILWRM.ST" href="ST.html">MILWRM.ST</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="">
<li><code><a title="MILWRM.assemble_pita" href="#MILWRM.assemble_pita">assemble_pita</a></code></li>
<li><code><a title="MILWRM.blur_features_st" href="#MILWRM.blur_features_st">blur_features_st</a></code></li>
<li><code><a title="MILWRM.map_pixels" href="#MILWRM.map_pixels">map_pixels</a></code></li>
<li><code><a title="MILWRM.show_pita" href="#MILWRM.show_pita">show_pita</a></code></li>
<li><code><a title="MILWRM.trim_image" href="#MILWRM.trim_image">trim_image</a></code></li>
</ul>
</li>
<li><h3><a href="#header-classes">Classes</a></h3>
<ul>
<li>
<h4><code><a title="MILWRM.img" href="#MILWRM.img">img</a></code></h4>
<ul class="">
<li><code><a title="MILWRM.img.blurring" href="#MILWRM.img.blurring">blurring</a></code></li>
<li><code><a title="MILWRM.img.calculate_non_zero_mean" href="#MILWRM.img.calculate_non_zero_mean">calculate_non_zero_mean</a></code></li>
<li><code><a title="MILWRM.img.clip" href="#MILWRM.img.clip">clip</a></code></li>
<li><code><a title="MILWRM.img.copy" href="#MILWRM.img.copy">copy</a></code></li>
<li><code><a title="MILWRM.img.create_tissue_mask" href="#MILWRM.img.create_tissue_mask">create_tissue_mask</a></code></li>
<li><code><a title="MILWRM.img.downsample" href="#MILWRM.img.downsample">downsample</a></code></li>
<li><code><a title="MILWRM.img.equalize_hist" href="#MILWRM.img.equalize_hist">equalize_hist</a></code></li>
<li><code><a title="MILWRM.img.from_npz" href="#MILWRM.img.from_npz">from_npz</a></code></li>
<li><code><a title="MILWRM.img.from_tiffs" href="#MILWRM.img.from_tiffs">from_tiffs</a></code></li>
<li><code><a title="MILWRM.img.log_normalize" href="#MILWRM.img.log_normalize">log_normalize</a></code></li>
<li><code><a title="MILWRM.img.plot_image_histogram" href="#MILWRM.img.plot_image_histogram">plot_image_histogram</a></code></li>
<li><code><a title="MILWRM.img.scale" href="#MILWRM.img.scale">scale</a></code></li>
<li><code><a title="MILWRM.img.show" href="#MILWRM.img.show">show</a></code></li>
<li><code><a title="MILWRM.img.subsample_pixels" href="#MILWRM.img.subsample_pixels">subsample_pixels</a></code></li>
<li><code><a title="MILWRM.img.to_npz" href="#MILWRM.img.to_npz">to_npz</a></code></li>
</ul>
</li>
<li>
<h4><code><a title="MILWRM.mxif_labeler" href="#MILWRM.mxif_labeler">mxif_labeler</a></code></h4>
<ul class="">
<li><code><a title="MILWRM.mxif_labeler.confidence_score_images" href="#MILWRM.mxif_labeler.confidence_score_images">confidence_score_images</a></code></li>
<li><code><a title="MILWRM.mxif_labeler.label_tissue_regions" href="#MILWRM.mxif_labeler.label_tissue_regions">label_tissue_regions</a></code></li>
<li><code><a title="MILWRM.mxif_labeler.make_umap" href="#MILWRM.mxif_labeler.make_umap">make_umap</a></code></li>
<li><code><a title="MILWRM.mxif_labeler.plot_mse_mxif" href="#MILWRM.mxif_labeler.plot_mse_mxif">plot_mse_mxif</a></code></li>
<li><code><a title="MILWRM.mxif_labeler.plot_percentage_variance_explained" href="#MILWRM.mxif_labeler.plot_percentage_variance_explained">plot_percentage_variance_explained</a></code></li>
<li><code><a title="MILWRM.mxif_labeler.plot_tissue_ID_proportions_mxif" href="#MILWRM.mxif_labeler.plot_tissue_ID_proportions_mxif">plot_tissue_ID_proportions_mxif</a></code></li>
<li><code><a title="MILWRM.mxif_labeler.prep_cluster_data" href="#MILWRM.mxif_labeler.prep_cluster_data">prep_cluster_data</a></code></li>
<li><code><a title="MILWRM.mxif_labeler.show_marker_overlay" href="#MILWRM.mxif_labeler.show_marker_overlay">show_marker_overlay</a></code></li>
</ul>
</li>
<li>
<h4><code><a title="MILWRM.st_labeler" href="#MILWRM.st_labeler">st_labeler</a></code></h4>
<ul class="">
<li><code><a title="MILWRM.st_labeler.confidence_score" href="#MILWRM.st_labeler.confidence_score">confidence_score</a></code></li>
<li><code><a title="MILWRM.st_labeler.label_tissue_regions" href="#MILWRM.st_labeler.label_tissue_regions">label_tissue_regions</a></code></li>
<li><code><a title="MILWRM.st_labeler.plot_gene_loadings" href="#MILWRM.st_labeler.plot_gene_loadings">plot_gene_loadings</a></code></li>
<li><code><a title="MILWRM.st_labeler.plot_mse_st" href="#MILWRM.st_labeler.plot_mse_st">plot_mse_st</a></code></li>
<li><code><a title="MILWRM.st_labeler.plot_percentage_variance_explained" href="#MILWRM.st_labeler.plot_percentage_variance_explained">plot_percentage_variance_explained</a></code></li>
<li><code><a title="MILWRM.st_labeler.plot_tissue_ID_proportions_st" href="#MILWRM.st_labeler.plot_tissue_ID_proportions_st">plot_tissue_ID_proportions_st</a></code></li>
<li><code><a title="MILWRM.st_labeler.prep_cluster_data" href="#MILWRM.st_labeler.prep_cluster_data">prep_cluster_data</a></code></li>
<li><code><a title="MILWRM.st_labeler.show_feature_overlay" href="#MILWRM.st_labeler.show_feature_overlay">show_feature_overlay</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc" title="pdoc: Python API documentation generator"><cite>pdoc</cite> 0.10.0</a>.</p>
</footer>
</body>
</html>